---
title: "Pacific calculus world data humann3 pahway analysis"
author: "Irina Velsko"
date: "`r format(Sys.time(), '%b %d, %Y')`"
output:
    github_document:
        fig_width: 8
        fig_height: 6
        toc: true
---


```{r load_libraries, echo = F, message = F}
library(knitr)
library(decontam)
library(data.table)
# library(vegan)
library(edgeR) # needed to run ComBat_seq
library(mixOmics)
library(janitor)
library(rstatix)
library(pander)
library(variancePartition)
library(tidyverse)
library(gplots)
library(ggrepel)
library(viridis)
library(patchwork)
library(RColorBrewer)
library(microshades)
opts_chunk$set(echo=F, warning=F, message=F, dpi=150)
```

Set the working directory of the notebook to the parent directory of the 
notebook itself.  

```{r setup}
knitr::opts_knit$set(root.dir = normalizePath(".."))
```


# humann3 tables 
```{r load_data}
## load the species and genus tables generated with humann3
humann3_path_full <- fread("04-Analysis/humann3/pathabundance_anc_params_joined_cpm.tsv")
humann3_path_full <- as_tibble(humann3_path_full)

# clean the file names
humann3_path_full <- rename(humann3_path_full, Pathway = `# Pathway`)
colnames(humann3_path_full) <- gsub("_Abundance","", colnames(humann3_path_full))
colnames(humann3_path_full) <- gsub(".SG1.1","", colnames(humann3_path_full))

# remove unmapped and ungrouped reads
humann3_path <- humann3_path_full %>% 
  filter(!str_detect(Pathway, "UNMAPPED|UNINTEGRATED")) %>%
  # remove Heping Dao calculus
  select(-matches("HPD"))


# World data
## load the species and genus tables generated with humann3
humann3_path_worldo_full <- fread("04-Analysis/humann3/world_samples/pathabundance_anc_params_world_joined_cpm.tsv")
humann3_path_worldo_full <- as_tibble(humann3_path_worldo_full)

# clean the file names
humann3_path_worldo_full <- rename(humann3_path_worldo_full, Pathway = `# Pathway`)
colnames(humann3_path_worldo_full) <- gsub("_Abundance","", colnames(humann3_path_worldo_full))
colnames(humann3_path_worldo_full) <- gsub(".SG1.1","", colnames(humann3_path_worldo_full))
colnames(humann3_path_worldo_full) <- gsub(".SG1","", colnames(humann3_path_worldo_full))

# remove unmapped and ungrouped reads
humann3_path_worldo <- humann3_path_worldo_full %>% 
  filter(!str_detect(Pathway, "UNMAPPED|UNINTEGRATED"))


# And combine the tables
humann3_path_world <- humann3_path_worldo %>%
  full_join(., humann3_path, by = "Pathway") %>%
  replace(is.na(.), 0)


```

```{r metadata}
# load the metadata file
# pacific_metadata <- fread("./05-Documentation.backup/oceania_metadata_full.tsv") %>%
#   full_join(., fread("./05-Documentation.backup/oceana_climate.tsv")) %>%
#   mutate(Island = ifelse(Type == "blank", "Blank",Island)) %>%
#   mutate(Island = ifelse(Type == "bone", "Bone", Island)) %>%
#   # mutate(Library_ID = str_replace_all(Library_ID, ".SG1.1","")) %>%
#   mutate(Island = str_replace_all(Island, "Efate_3000", "Efate 3000 BP"),
#          Island = str_replace_all(Island, "Rapa_Nui", "Rapa Nui"),
#          Island = str_replace_all(Island, "Viti_Levu", "Viti Levu"),
#          Island = str_replace_all(Island, "Taumako_Duff_Islands", "Taumako")) %>%
#   mutate(Type = str_replace_all(Type, "bone","Arch. Bone"),
#          Type = str_replace_all(Type, "calculus","Arch. calculus"))

pacific_metadata <- fread("05-Documentation.backup/oceania_metadata_full.tsv") %>%
  select(Library_ID, Lab, Type, Age_mean, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = "Pacific",
         Age_mean_log = log10(Age_mean))

pc_end_samples <- fread("./05-Documentation.backup/pc_end_samples.tsv") %>%
  pull(Species)

```

```{r}
metadata <- fread("../smoking_calculus/00-documentation.backup/full_combined_metadata.tsv") %>%
  select(Library_ID, Site_code, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = str_replace_all(Site_code, "MID","Middenbeemster"),
         Study = str_replace_all(Study, "CMB","Middenbeemster"),
         Study = str_replace_all(Study, "KIL","Kilteasheen"),
         Study = str_replace_all(Study, "RAD","Radcliffe"),
         Study = str_replace_all(Study, "ELR","Iberia"),
         Study = str_replace_all(Study, "IVE","Iberia"),
         Lab = str_replace_all(Site_code, "MID","Jena"),
         Lab = str_replace_all(Lab, "CMB","Jena"),
         Lab = str_replace_all(Lab, "ELR","Jena"),
         Lab = str_replace_all(Lab, "IVE","Jena"),
         Lab = str_replace_all(Lab, "RAD","Oxford"),
         Lab = str_replace_all(Lab, "KIL","Tuebingen")) %>%
  mutate(Age_mean = ifelse(str_detect(Site_code, "MID|CMB|RAD"), 250,
                           ifelse(Site_code == "KIL", 700,
                                  ifelse(str_detect(Site_code, "ELR|IVE"), 620, 7)))) %>%
  full_join(., fread("./05-Documentation.backup/eisenhoffer2020_ottoni2021_deepevo_mann2018w_metadata.tsv") %>%
              mutate(Run_accession = ifelse(Study == "FellowsYates2021",Sample_alias,Run_accession)) %>%
              rename(Library_ID = Run_accession)) %>%
  mutate(Sample_type = ifelse(str_detect(Study, "EXB|LIB|DEN|SOL"),"blank",Sample_type)) %>%
  mutate(Sample_type = ifelse(is.na(Sample_type),"calculus",Sample_type)) %>%
  mutate(Sample_group = ifelse(Sample_type == "blank","blank",
                               ifelse(str_detect(Library_ID, "VLC|JAE"),"modern_calculus",Sample_group))) %>%
  mutate(Sample_group = ifelse(is.na(Sample_group),"ancient_calculus",Sample_group)) %>%
  mutate(Sample_alias = ifelse(is.na(Sample_alias),Library_ID,Sample_alias)) %>%
  mutate(Age_mean_log = log10(Age_mean))

world_metadata <- metadata %>%
  # mutate(Sample_group = ifelse(is.na(Sample_group),"bone",Sample_group)) %>%
  filter(!str_detect(Sample_group, "primate|modern"))

# list the non-human primates and modern calculus to remove from the species table
nhp <- metadata %>%
  filter(str_detect(Sample_group, "primate|modern")) %>%
  select(Sample_alias) %>%
  unique() %>%
  pull()

# and combine pacific and world
full_metadata <- pacific_metadata %>%
  bind_rows(., world_metadata)


```


```{r}
# make a list of blanks b/c they don't all have the same prefixes
blanks <- full_metadata %>%
  mutate(Type = ifelse(is.na(Type),Sample_type,Type)) %>%
  filter(str_detect(Type, "blank|bone")) %>%
  filter(!str_detect(Library_ID, "ARS")) %>%
  select(Library_ID) %>%
  pull()

```


```{r cuperdec_poor_samples}
# poor_samples <- fread("./05-Documentation.backup/cfdp_fail_samples_20210413.txt") %>%
#   pull(Sample)

cfdp_fail <- fread("05-Documentation.backup/oceania_cuperdec_mpa3_anc_params_poor_samples.tsv") %>%
  rename(Library_ID = 1) %>%
  pull(Library_ID)

poor_samples <- fread("05-Documentation.backup/mpa3_anc_params_poor_samples.tsv")

world_fail <- fread("./05-Documentation.backup/cuperdec_malt_ei_ot_de_poor_samples.tsv") %>%
  bind_rows(., fread("05-Documentation.backup/oceania_cuperdec_mpa3_anc_params_poor_samples.tsv"))

```

```{r colors}
Island_colors = c("Tongatapu" = "#616161", 
                  "Rurutu" = "#9d654c", 
                  "Flores" = "#f6865e", 
                  "Rapa Nui" = "#cc79a7", 
                  "Futuna" = "#7D3560", 
                  "Efate" = "#7472af", 
                  "Vao" = "#098BD9",
                  "Efate 3000 BP" = "#7DCCFF", 
                  "Uripiv" = "#e1f2fe", 
                  "Viti Levu" = "#1ec1a7", 
                  "Raiatea" = "#148f77", 
                  "Taumako" = "#4e7705", 
                  "Watom" = "#86c73a")


# Set microshades colors
Lab_colors = c("Jena" = microshades_palette("micro_cvd_purple", 3, lightest = FALSE)[3],
                "Oklahoma" =  microshades_palette("micro_cvd_blue", 3, lightest = FALSE)[3],
                "Otago" = microshades_palette("micro_cvd_green", 3, lightest = FALSE)[3],
                "Oxford" = microshades_palette("micro_cvd_turquoise", 3, lightest = FALSE)[1],
                "Tuebingen" = microshades_palette("micro_cvd_purple", 3, lightest = FALSE)[1],
                "ACAD" = microshades_palette("micro_cvd_blue", 4, lightest = FALSE)[1],
                "Vienna" = microshades_palette("micro_cvd_green", 3, lightest = FALSE)[1])


Sample_type_colors <- c("Arch. Bone" = "#148F77", "blank" = "#8B8B8B", "Arch. calculus" = "black")

Study_colors <- c("Pacific" = "#098BD9", 
                  "Middenbeemster" = "#7472af", 
                  "Radcliffe" = "#cc79a7", 
                  "Kilteasheen" = "#f6865e", 
                  "Iberia" = "#1ec1a7",
                  "Eisenhoffer2020" = "#86c73a",
                  "Ottoni2021" = "#616161",
                  "FellowsYates2021" = "#e1f2fe",
                  "Mann2018" = "#f5c710ff") # 7D3560

```

```{r}
Island_shapes = c("Tongatapu" = 21, 
                  "Rurutu" = 21, 
                  "Flores" = 21, 
                  "Rapa Nui" = 21, 
                  "Futuna" = 21, 
                  "Efate" = 21, 
                  "Vao" = 21,
                  "Efate 3000 BP" = 21, 
                  "Uripiv" = 21, 
                  "Viti Levu" = 21, 
                  "Raiatea" = 21, 
                  "Taumako" = 21, 
                  "Watom" = 21)

Lab_shapes = c(21,21,21)
Lab_size = c(4,4,4)

Sample_type_shapes = c(21,21,21,21,21,21,21,21)
Sample_type_size = c(4,4,4,4,4,4,4,4)

```


```{r plot_pca_fxn}

# add functions for:
# making the input table for PCA?

# plotting PCA with colored dots
plot_pca <- function(df, pc1, pc2, color_group, shape_group) {
    metadata_group_colors <- get(paste(color_group, "_colors", sep = ""))
    metadata_group_shapes <- get(paste(shape_group, "_shapes", sep = ""))

    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%")
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(full_metadata, by = "Library_ID")
    
    color_group = df_X[[color_group]]
    shape_group = df_X[[shape_group]]

                       if (pc1 == 'PC1') {
                           pc1 <- df_X$PC1
                           exp_var_pc1 <- exp_var[1]
                           xaxis <- c("PC1")
                       }  else if (pc1 == 'PC2') {
                           pc1 <- df_X$PC2
                           exp_var_pc1 <- exp_var[2]
                           xaxis <- c("PC2")
                       } else if (pc1 == 'PC3') {
                           pc1 <- df_X$PC3
                           exp_var_pc1 <- exp_var[3]
                           xaxis <- c("PC3")
                       }

                       if (pc2 == 'PC1') {
                           pc2 <- df_X$PC1
                           exp_var_pc2 <- exp_var[1]
                           yaxis <- c("PC1")
                       }  else if (pc2 == 'PC2') {
                           pc2 <- df_X$PC2
                           exp_var_pc2 <- exp_var[2]
                           yaxis <- c("PC2")
                       } else if (pc2 == 'PC3') {
                           pc2 <- df_X$PC3
                           exp_var_pc2 <- exp_var[3]
                           yaxis <- c("PC3")
                       }

    pca_plot <- ggplot(df_X, aes(pc1, pc2)) +
                        geom_point(aes(fill = color_group, shape = shape_group), size = 4, stroke = 0.3, shape = 21) +
                        scale_fill_manual(values = metadata_group_colors) +
                        xlab(paste(xaxis, " - ", exp_var_pc1)) +
                        ylab(paste(yaxis, " - ", exp_var_pc2)) +
                        theme_minimal(base_size = 14) +
                        theme(legend.position = "right") +
                        theme(legend.title = element_blank())

    return(pca_plot)
}

```


```{r plot_bi}
# this function plots the PCA with biplots to look at the categories with strongest loadings
plot_pca_bi <- function(df, pc1, pc2, metadata_group, columntitle, Pathways = c("TRUE","FALSE")) {
    metadata_group_colors <- get(paste(metadata_group, "_colors", sep = ""))
    metadata_group_shapes <- get(paste(metadata_group, "_shapes", sep = ""))
   
    arrow_pc <- enquo(columntitle)
    
    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%") # explained variance for x- and y-labels
    
    # select only the PCs from the PCA and add metadata
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(full_metadata, by = "Library_ID")
    
    metadata_group = df_X[[metadata_group]]
    
    corr_lam <- df$sdev[c("PC1", "PC2", "PC3")] * sqrt(nrow(df_X))

    df_X <- df_X %>%
      mutate(PC1 = PC1 / corr_lam[1],
             PC2 = PC2 / corr_lam[2], 
             PC3 = PC3 / corr_lam[3])

    # select the correct PC column and explained variance for PC1
    if (pc1 == 'PC1') {
        Pc1 <- df_X$PC1
        exp_var_pc1 <- exp_var[1]
        xaxis <- c("PC1")
    } else if (pc1 == 'PC2') {
        Pc1 <- df_X$PC2
        exp_var_pc1 <- exp_var[2]
        xaxis <- c("PC2")
    } else if (pc1 == 'PC3') {
        Pc1 <- df_X$PC3
        exp_var_pc1 <- exp_var[3]
        xaxis <- c("PC3")
   }
    
    # select the correct PC column and explained variance for PC2
    if (pc2 == 'PC1') {
        Pc2 <- df_X$PC1
        exp_var_pc2 <- exp_var[1]
        yaxis <- c("PC1")
 } else if (pc2 == 'PC2') {
       Pc2 <- df_X$PC2
       exp_var_pc2 <- exp_var[2]
       yaxis <- c("PC2")
    } else if (pc2 == 'PC3') {
       Pc2 <- df_X$PC3
       exp_var_pc2 <- exp_var[3]
       yaxis <- c("PC3")
   }

    # Identify the 10 Pathways that have highest positive and negative loadings in the selected PC
    pws_10 <- df$loadings$X %>%
      as.data.frame(.) %>%
      rownames_to_column(var = "Function") %>%
      top_n(10, !!arrow_pc)
   if (Pathways == T) {
      pws_10 <- pws_10 %>%
        rename(Pathway = Function) %>%
        mutate(Function = sapply(Pathway, function(f) {
                                  unlist(str_split(f, ":"))[1]
                           })) 
    }

    neg_10 <- df$loadings$X %>%
      as.data.frame(.) %>%
      rownames_to_column(var = "Function") %>%
      top_n(-10, !!arrow_pc)
    if (Pathways == T) {
      neg_10 <- neg_10 %>%
      rename(Pathway = Function) %>%
       mutate(Function = sapply(Pathway, function(f) {
                                  unlist(str_split(f, ":"))[1]
                           })) 
    }


    pca_plot_bi <- ggplot(df_X, aes(x = Pc1, y = Pc2)) +
      geom_point(aes(shape = metadata_group, fill = metadata_group), size = 3.5, stroke = 0.3) +
      geom_segment(data = pws_10,
                   aes(xend = get(paste(pc1)), yend = get(paste(pc2))),
                   x = 0, y = 0, colour = "black",
                   size = 0.5,
                   arrow = arrow(length = unit(0.03, "npc"))) +
      geom_label_repel(data = pws_10,
                   aes(x = get(paste(pc1)), y = get(paste(pc2)), label = Function),
                   size = 1.5, colour = "grey20", label.padding = 0.2, max.overlaps = 12) +
      geom_segment(data = neg_10,
                   aes(xend = get(paste(pc1)), yend = get(paste(pc2))),
                   x = 0, y = 0, colour = "grey50",
                   size = 0.5,
                   arrow = arrow(length = unit(0.03, "npc"))) +
      geom_label_repel(data = neg_10,
                   aes(x = get(paste(pc1)), y = get(paste(pc2)), label = Function),
                   size = 1.5, colour = "grey20", label.padding = 0.2, max.overlaps = 12) +
      labs(x = paste(xaxis, " - ", exp_var_pc1),
           y = paste(yaxis, " - ", exp_var_pc2)) +
      scale_fill_manual(values = metadata_group_colors) +
      scale_shape_manual(values = metadata_group_shapes) +
      theme_minimal() + theme(text = element_text(size = 14)) +
      theme(legend.position = "right")

    return(pca_plot_bi)
}


```


```{r pca_plot_cont_fxn}
# for continuous data
plot_pca_cont <- function(df_pca, pc1, pc2, color_group, ncomps, title_text) {

    exp_var <- paste0(round(df_pca$prop_expl_var$X * 100, 2), "%")
    df_X <- df_pca$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(full_metadata, by = "Library_ID")

    color_group = df_X[[color_group]]

    if (pc1 == 'PC1') {
        pc1 <- df_X$PC1
        exp_var_pc1 <- exp_var[1]
        xaxis <- c("PC1")
    }  else if (pc1 == 'PC2') {
        pc1 <- df_X$PC2
        exp_var_pc1 <- exp_var[2]
        xaxis <- c("PC2")
    } else if (pc1 == 'PC3') {
        pc1 <- df_X$PC3
        exp_var_pc1 <- exp_var[3]
        xaxis <- c("PC3")
    }

    if (pc2 == 'PC1') {
        pc2 <- df_X$PC1
        exp_var_pc2 <- exp_var[1]
        yaxis <- c("PC1")
    }  else if (pc2 == 'PC2') {
        pc2 <- df_X$PC2
        exp_var_pc2 <- exp_var[2]
        yaxis <- c("PC2")
    } else if (pc2 == 'PC3') {
        pc2 <- df_X$PC3
        exp_var_pc2 <- exp_var[3]
        yaxis <- c("PC3")
    }

    pca_plot <- ggplot(df_X, aes(x = pc1, y = pc2)) +
      geom_point(aes(fill = color_group), size = 3.5, stroke = 0.3, shape = 21) +
     scale_fill_viridis_c(option = "C") +
     # scale_shape_manual(values = c(16, 16, 16, 16, 16,16,16,16)) +
     # scale_shape_manual(values = c(16, 17, 15, 18, 14,9,13,24)) +
    # stat_ellipse() +
     xlab(paste(xaxis, " - ", exp_var_pc1)) +
     ylab(paste(yaxis, " - ", exp_var_pc2)) +
     theme_minimal(base_size = 16) +
     theme(text = element_text(size=16)) +
     theme(legend.title = element_blank(),
           legend.key.size = unit(2,"mm"),
           legend.text = element_text(size = 10)) + 
     theme(legend.position = "right") +
     ggtitle(title_text) + theme(plot.title = element_text(size = 10))

    return(pca_plot)
}

```

# Pathway analysis
```{r}
# look at the pathways that are enriched in the 2 ends of the taxonomy PCA



```

```{r}
library(phyloseq)
library(ANCOMBC)

pc_ends <- fread("./05-Documentation.backup/pc_end_samples.tsv") %>%
  mutate(dir = ifelse(PC1 <0, "Negative","Positive")) %>%
  rename(Sample = Species)

# convert the GMGC table to a phyloseq object for ANCOMBC
h3_path_w_sp <- humann3_path_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  filter(!str_detect(Pathway, "\\|")) %>%
  select(Pathway, matches(pc_end_samples %>% str_c(collapse = "|"))) %>%
  # now arrange the samples alphabetically
  pivot_longer(!Pathway, names_to = "libraryID", values_to = "logs") %>%
  mutate(libraryID = str_replace_all(libraryID, ".SG1","")) %>%
  arrange(libraryID) %>%
  pivot_wider(names_from = "libraryID", values_from = "logs") %>%
  arrange(Pathway) %>%
  column_to_rownames("Pathway") %>%
  as.matrix(.)

h3_path_w_sp_tax <- h3_path_w_sp %>%
  as.data.frame(.) %>%
  rownames_to_column("Pathway") %>%
  arrange(Pathway) %>%
  select(Pathway) %>%
  mutate(Species = Pathway) %>%
  column_to_rownames("Pathway") %>%
  as.matrix(.)

metadata <- sample_data(full_metadata %>%
  inner_join(., pc_ends %>%
               select(Sample, dir) %>%
               rename(Library_ID = Sample), by = "Library_ID") %>%
  distinct() %>%
  column_to_rownames("Library_ID"))
  
 # %>%
 #  mutate(Full = as.character(Full)

OTU = otu_table(h3_path_w_sp, taxa_are_rows = TRUE)
TAX = tax_table(h3_path_w_sp_tax)
physeq = phyloseq(OTU, TAX, metadata)

physeq1 <- merge_phyloseq(physeq, metadata)

# check the phyloseq object
physeq

pacific_diff_pc_ends <- ancombc(physeq, formula = "dir", p_adj_method = "fdr", group = "dir", struc_zero = TRUE)

```

```{r}

lfc <- pacific_diff_pc_ends$res$beta 

qval <- pacific_diff_pc_ends$res$q_val

diffabund <- pacific_diff_pc_ends$res$diff_abn

alltogether <- pacific_diff_pc_ends$res$beta %>%
  rownames_to_column("Pathway") %>%
  rename(lfc_pos = 2) %>%
  full_join(., pacific_diff_pc_ends$res$q_val %>%
            rownames_to_column("Pathway") %>%
            rename(qval_pos = 2)) %>%
  full_join(., pacific_diff_pc_ends$res$diff_abn %>%
            rownames_to_column("Pathway") %>% 
            rename(da_pos = 2))


alltogether %>%
  filter(da_pos == TRUE,
         qval < 0.05) %>%
  arrange(desc(lfc_pos)) %>%
  fwrite(., "~/archgen/microbiome_calculus/pacific_calculus/05-Documentation.backup/pathway_abundance_ancombc.tsv", quote = F, sep = "\t")

```

```{r}
# 
alltogether %>%
  mutate(colorpt = ifelse(qval_pos <= 0.05 & lfc_pos >=2, "turquoise",
                          ifelse(qval_pos <= 0.05 & lfc_pos <= -2, "turquoise", "coral"))) %>%
  mutate(qval_pos = ifelse(qval_pos == 0, 7.742690e-295	, qval_pos)) %>%
  mutate(beta = -1 * log10(qval_pos)) %>%
  ggplot(., aes(x = lfc_pos, y = beta, color = colorpt)) +
     geom_point() +
     scale_fill_viridis_c(option = "C") +
     theme_minimal(base_size = 12) +
     theme(text = element_text(size=12)) +
     theme(legend.position = "right") +
     theme(plot.title = element_text(size = 10)) +
     xlab("Log-fold change") +
     ylab("-Log10(q-val)")

```


```{r all_pca}
 # all samples and blanks
humann3_path_l1_all <- humann3_path_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  filter(!str_detect(Pathway, "\\|")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  pivot_longer(!Pathway, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Pathway,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_path_pca <- tune.pca(humann3_path_l1_all, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_path_l1_all.pca <- pca(humann3_path_l1_all, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_path_l1_all.pca, "PC1", "PC2", "Study", "Sample_type")


```

Remove the poorly-preserved samples based on cuperdec. 
These were removed from the taxonomic analyses.
```{r cuperdec_poor_rem_pca}
# make a list of blanks b/c they don't all have the same prefixes

blanks <- full_metadata %>%
  mutate(Type = ifelse(is.na(Type),Sample_type,Type)) %>%
  filter(str_detect(Type, "blank|bone")) %>%
  filter(!str_detect(Library_ID, "ARS")) %>%
  select(Library_ID) %>%
  pull()

# create the input matrix
humann3_path_l1_cpdrm <- humann3_path_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  filter(!str_detect(Pathway, "\\|")) %>%
  # remove poorly preserved samples
  select(-matches(world_fail %>% pull() %>% str_c(collapse = "|"))) %>%
  select(-all_of(poor_samples %>% pull(Library_ID))) %>%
  pivot_longer(!Pathway, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Pathway,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_path_pca <- tune.pca(humann3_path_l1_cpdrm, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_path_l1_cpdrm.pca <- pca(humann3_path_l1_cpdrm, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_path_l1_cpdrm.pca, "PC1", "PC2", "Study", "Sample_type")

```



```{r, eval = F}
#  list taxonomicly well-preserved samples that plot with blanks to remove from downstream analysis
other_poor <- humann3_path_l1_cpdrm.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  filter(PC1 > 0) %>%
  select(Library_ID, Type) %>%
  filter(Type == "Arch. calculus") %>%
  pull(Library_ID)

```


```{r poor_marked}
# mark the taxonomically poorly-preserved samples to see where they fall in pathway preservation
poor_df <- poor_samples %>%
  as_tibble(.) %>%
  rename(Library_ID = 1) %>%
  # add the cuperdec poor samples
  bind_rows(., world_fail %>%
              as_tibble(.) %>%
              rename(Library_ID = 1)) %>%
  mutate(Preservation = "Poor")

exp_var <- paste0(round(humann3_path_l1_all.pca$prop_expl_var$X * 100, 2), "%")
df_X_all <- humann3_path_l1_all.pca$variates$X %>%
           as.data.frame() %>%
           rownames_to_column("Library_ID") %>%
           left_join(full_metadata, by = "Library_ID") %>%
           full_join(., poor_df, by = "Library_ID") %>%
           mutate(Preservation = ifelse(is.na(Preservation), "Good", Preservation),
                  Sample_type = ifelse(is.na(Sample_type), Type, Sample_type)) %>%
           mutate(Preservation = ifelse(Sample_type == "blank","Blank",
                                        ifelse(Sample_type == "Arch. Bone", "Arch. Bone", Preservation))) 
# %>%
#            # there's one sample with NA and no metadata, so call it poorly-preserved
#            mutate(Preservation = ifelse(is.na(Preservation), "Poor", Preservation))

# test plot
df_X_all %>%
  ggplot(., aes(PC1, PC2, color = Preservation, shape = Preservation)) +
                  geom_point(size = 3) +
                  xlab(paste("PC1 - ", exp_var[1])) +
                  ylab(paste("PC2 - ", exp_var[2])) +
                  # xlim(5, 10) +
                  # ylim(-1,1) +
                  theme_minimal(base_size = 14) +
                  theme(legend.position = "right") +
                  theme(legend.title = element_blank())

```

```{r path_poor}
# Based on the PCA above, try all samples with PC1 loadings < 7.5 as "poorly preserved"
path_poor <- humann3_path_l1_all.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(full_metadata, by = "Library_ID") %>%
  filter(PC1 > 30) %>%
  select(Library_ID, Sample_type) %>%
  filter(Sample_type != "blank") %>%
  arrange(Library_ID) %>%
  pull(Library_ID)


```

Remove the path_poor samples and see what the PCA looks like
```{r poor_removed_marked}
df_X_all %>%
  filter(!str_detect(Library_ID, path_poor %>%
                       str_c(collapse = "|"))) %>%
  ggplot(., aes(PC1, PC2, color = Preservation, shape = Preservation)) +
                  geom_point(size = 4) +
                  xlab(paste("PC1 - ", exp_var[1])) +
                  ylab(paste("PC2 - ", exp_var[2])) +
                  theme_minimal(base_size = 14) +
                  theme(legend.position = "right") +
                  theme(legend.title = element_blank())



```

What does the PCA look like if we filter low-prevalence pathways and remove
poorly-preserved samples?

```{r filter_all_pca}
# try filtering the table
# list all pathways that are present in at least 1/3 of samples, as a way to reduce the input
humann3_path.decontam_noblanks_presence_more_30 <- humann3_path_world %>%
  filter(!str_detect(Pathway, "\\|")) %>%
  # remove poorly preserved samples and blanks
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks%>% str_c(collapse = "|"))) %>%
  select(-matches(path_poor %>% str_c(collapse = "|"))) %>%
  mutate_if(is.numeric, ~1 * (. > 0)) %>%
  pivot_longer(!Pathway, names_to = "Library_ID", values_to = "Counts") %>%
  group_by(Pathway) %>%
  summarize(Percent_presence = sum(Counts)/length(Library_ID)*100) %>%
  # select the pathways that are present in at least 1/3 of samples
  ungroup() %>%
  # pathways present in > 1/3 of samples 
  filter(Percent_presence >= 33.33) %>%
  # pull out these pathways to keep
  select(Pathway)

# create the input matrix
humann3_path_l1_filt <- humann3_path_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  filter(!str_detect(Pathway, "\\|")) %>%
  # remove poorly preserved samples
  select(-matches(blanks%>% str_c(collapse = "|"))) %>%
  select(-matches(path_poor %>% str_c(collapse = "|"))) %>%
  inner_join(., humann3_path.decontam_noblanks_presence_more_30, by = "Pathway") %>%
  pivot_longer(!Pathway, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Pathway,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_path_filt_pca <- tune.pca(humann3_path_l1_filt, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_path_l1_filt.pca <- pca(humann3_path_l1_filt, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_path_l1_filt.pca, "PC1", "PC2", "Study", "Lab")

exp_var <- paste0(round(humann3_path_l1_filt.pca$prop_expl_var$X * 100, 2), "%")
df_X_filt <- humann3_path_l1_filt.pca$variates$X %>%
           as.data.frame() %>%
           rownames_to_column("Library_ID") %>%
           left_join(full_metadata, by = "Library_ID") %>%
           # full_join(., poor_df, by = "Library_ID") %>%
           mutate(Preservation = "Good") %>%
           mutate(Preservation = ifelse(Type == "blank","Blank",
                                        ifelse(Type == "Arch. Bone", "Arch. Bone", Preservation)))

# test plot
df_X_filt %>%
  ggplot(., aes(PC1, PC2, color = Preservation, shape = Preservation)) +
                  geom_point(size = 4) +
                  xlab(paste("PC1 - ", exp_var[1])) +
                  ylab(paste("PC2 - ", exp_var[2])) +
                  # xlim(5, 10) +
                  # ylim(-1,1) +
                  theme_minimal(base_size = 14) +
                  theme(legend.position = "right") +
                  theme(legend.title = element_blank())


```


Plot a PCA with only the well-preserved samples
```{r pca_paths_noblanks}

humann3_path_l1_noblanks <- humann3_path_world %>%
  filter(!str_detect(Pathway, "\\|")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly prserved samples
  select(-matches(path_poor %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks%>% str_c(collapse = "|"))) %>%
  # select(-all_of(poor_samples)) %>%
  # select(-all_of(other_poor)) %>%
  # inner_join(., humann3_path.decontam_noblanks_presence_more_30, by = "Pathway") %>%
  pivot_longer(!Pathway, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Pathway,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_path_pca <- tune.pca(humann3_path_l1_noblanks, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_path_l1_noblanks.pca <- pca(humann3_path_l1_noblanks, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_path_l1_noblanks.pca, "PC1", "PC2", "Study", "Sample_type")
plot_pca(humann3_path_l1_noblanks.pca, "PC1", "PC2", "Lab", "Sample_type")

```

Plot a PCA with only the well-preserved samples and filter the pathways by prevalence
```{r pca_paths_filt_noblanks}
humann3_path_l1_noblanks_filt <- humann3_path_world %>%
  filter(!str_detect(Pathway, "\\|")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(path_poor %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks%>% str_c(collapse = "|"))) %>%
  # select(-all_of(poor_samples)) %>%
  # select(-all_of(other_poor)) %>%
  inner_join(., humann3_path.decontam_noblanks_presence_more_30, by = "Pathway") %>%
  pivot_longer(!Pathway, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Pathway,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_path_pca <- tune.pca(humann3_path_l1_noblanks_filt, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_path_l1_noblanks_filt.pca <- pca(humann3_path_l1_noblanks_filt, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_path_l1_noblanks_filt.pca, "PC1", "PC2", "Study", "Sample_type")
plot_pca(humann3_path_l1_noblanks_filt.pca, "PC1", "PC2", "Lab", "Lab")

```

Run a PERMANOVA with adonis2 on the abundance-filtered clr-transformed pathway matrix.
```{r adonis_pe_all_pres33}

humann3_path_l1_noblanks_filt <- humann3_path_world %>%
  filter(!str_detect(Pathway, "\\|")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(path_poor %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  # and remove these 2 b/c they have no age data for the adonis test
  select(-ERR5729638, -ERR5729643) %>%
  # select(-all_of(poor_samples)) %>%
  # select(-all_of(other_poor)) %>%
  inner_join(., humann3_path.decontam_noblanks_presence_more_30, by = "Pathway") %>%
  pivot_longer(!Pathway, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Pathway,Counts) %>%
  column_to_rownames("Library_ID")

# perform a PCA to see how the data cluster
humann3_path_l1_noblanks_filt.pca <- pca(humann3_path_l1_noblanks_filt, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
humann3_path_l1_noblanks_filt.euc <- vegan::vegdist(humann3_path_l1_noblanks_filt.pca$X, method = "euclidean", na.rm = F)
humann3_path_l1_noblanks_filt.euc.pcoa <- ape::pcoa(humann3_path_l1_noblanks_filt.euc)

humann3_path_l1_noblanks_filt.euc.pcoavectors <- as.data.frame(humann3_path_l1_noblanks_filt.euc.pcoa$vectors)
humann3_path_l1_noblanks_filt.euc.pcoavectors <- humann3_path_l1_noblanks_filt.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

humann3_path_l1_noblanks_filt.euc.vecmet <- left_join(humann3_path_l1_noblanks_filt.euc.pcoavectors, full_metadata, by = "Library_ID")

humann3_path_l1_noblanks_filt.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Lab)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Lab_colors) +
  # scale_shape_manual(values = Lab_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Lab")

metadata_all <- full_metadata %>%
    right_join(., humann3_path_l1_noblanks_filt.euc.pcoavectors %>% select(Library_ID), by = "Library_ID") %>%
  drop_na(Study)
  
metadata_all <- humann3_path_l1_noblanks_filt.euc.pcoavectors %>% 
  select(Library_ID) %>%
  left_join(., full_metadata, by = "Library_ID") %>%
  drop_na(Study)


# test for difference between different metadata categories 
vegan::adonis2(humann3_path_l1_noblanks_filt.euc ~ Lab, metadata_all, permutations = 999, by = "margin")

all_permanova <- vegan::adonis2(humann3_path_l1_noblanks_filt.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(vegan::adonis2(humann3_path_l1_noblanks_filt.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(vegan::adonis2(humann3_path_l1_noblanks_filt.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(vegan::adonis2(humann3_path_l1_noblanks_filt.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(vegan::adonis2(humann3_path_l1_noblanks_filt.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","**","**","*"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)


# all together
vegan::adonis2(humann3_path_l1_noblanks_filt.euc ~ Lab  + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
vegan::adonis2(humann3_path_l1_noblanks_filt.euc ~ Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)


```

## Run hierarchical clustering on the samples 
```{r}

humann3_path_l1_noblanks_filt.euc.mat <- as.matrix(humann3_path_l1_noblanks_filt.euc)

selected_metadata <- world_metadata %>%
  full_join(., pacific_metadata) %>%
  select(Library_ID, Study, Lab, Age_mean) %>%
  inner_join(., humann3_path_l1_noblanks_filt.euc.mat %>%
               as.data.frame(.) %>%
               rownames_to_column("Library_ID") %>%
               select(Library_ID), by = "Library_ID") %>%
  column_to_rownames("Library_ID")

col_age = circlize::colorRamp2(c(0, 3000), c("#098BD9", "#7D3560"))
col_fun = circlize::colorRamp2(c(0, 25, 50), c("#ffffff", "#77c8f9", "#098BD9"))


row_ha <- ComplexHeatmap::HeatmapAnnotation(df = selected_metadata, col = list(Study = Study_colors,
                                                                               Lab = Lab_colors,
                                                                               Age_mean = col_age))

ComplexHeatmap::Heatmap(humann3_path_l1_noblanks_filt.euc.mat, name = "EucDist", col = col_fun, clustering_method_rows = "complete", clustering_method_columns = "complete", top_annotation = row_ha)



```










