---
title: "Pacific and world calculus humann3 gene family analysis"
author: "Irina Velsko"
date: "`r format(Sys.time(), '%b %d, %Y')`"
output:
    github_document:
        fig_width: 8
        fig_height: 6
        toc: true
---


```{r load_libraries, echo = F, message = F}
library(knitr)
library(decontam)
library(data.table)
# library(vegan)
library(edgeR) # needed to run ComBat_seq
library(mixOmics)
library(janitor)
library(rstatix)
library(pander)
library(variancePartition)
library(vegan)
library(tidyverse)
library(gplots)
library(ggrepel)
library(pals)
library(viridis)
library(patchwork)
library(RColorBrewer)
library(microshades)
opts_chunk$set(echo=F, warning=F, message=F, dpi=150)
```

Set the working directory of the notebook to the parent directory of the 
notebook itself.  

```{r setup}
knitr::opts_knit$set(root.dir = normalizePath(".."))
```


# humann3 tables 
```{r load_UR90_data, eval = F}
## load the species and genus tables generated with humann3
humann3_GFs_full <- fread("./04-Analysis/humann3/genefamilies_anc_params_joined_cpm_ur90rxn.tsv")
humann3_GFs_full <- as_tibble(humann3_GFs_full)

# clean the file names
humann3_GFs_full <- rename(humann3_GFs_full, GeneFamily = `# Gene Family`)
colnames(humann3_GFs_full) <- gsub("_Abundance-RPKs","", colnames(humann3_GFs_full))
colnames(humann3_GFs_full) <- gsub(".SG1.1","", colnames(humann3_GFs_full))

# remove unmapped and ungrouped reads
humann3_GFs <- humann3_GFs_full %>% filter(!str_detect(GeneFamily, "UNMAPPED|UNGROUPED"))

# remove all species-level-assignments
humann3_GFs_top <- humann3_GFs %>%
  filter(!str_detect(GeneFamily, "\\|")) %>%
  # remove Heping Dao calculus
  select(-c("HPD005.B0101","HPD006.B0101","HPD007.B0101"))
  

```

```{r load_kegg_data}

## load the species and genus tables generated with humann3
humann3_GFs_full <- fread("./04-Analysis/humann3/genefamilies_anc_params_joined_cpm_ko_names.tsv")
humann3_GFs_full <- as_tibble(humann3_GFs_full)

# clean the file names
humann3_GFs_full <- rename(humann3_GFs_full, GeneFamily = `# Gene Family`)
colnames(humann3_GFs_full) <- gsub("_Abundance-RPKs","", colnames(humann3_GFs_full))
colnames(humann3_GFs_full) <- gsub(".SG1.1","", colnames(humann3_GFs_full))

# remove unmapped and ungrouped reads
humann3_GFs <- humann3_GFs_full %>% 
  filter(!str_detect(GeneFamily, "UNMAPPED|UNGROUPED")) %>%
  # remove Heping Dao calculus
  select(-c("HPD005.B0101","HPD006.B0101","HPD007.B0101"))

# remove all species-level-assignments
humann3_GFs_top <- humann3_GFs %>%
  filter(!str_detect(GeneFamily, "\\|"))


# World data
## load the species and genus tables generated with humann3
humann3_KO_worldo_full <- fread("04-Analysis/humann3/world_samples/genefamilies_anc_params_world_joined_cpm_ko_names.tsv")
humann3_KO_worldo_full <- as_tibble(humann3_KO_worldo_full)

# clean the file names
humann3_KO_worldo_full <- rename(humann3_KO_worldo_full, GeneFamily = `# Gene Family`)
colnames(humann3_KO_worldo_full) <- gsub("_Abundance-RPKs","", colnames(humann3_KO_worldo_full))
colnames(humann3_KO_worldo_full) <- gsub(".SG1.1","", colnames(humann3_KO_worldo_full))
colnames(humann3_KO_worldo_full) <- gsub(".SG1","", colnames(humann3_KO_worldo_full))

# remove unmapped and ungrouped reads
humann3_KO_worldo <- humann3_KO_worldo_full %>% 
  filter(!str_detect(GeneFamily, "UNMAPPED|UNGROUPED"))


# And combine the tables
humann3_KO_world <- humann3_KO_worldo %>%
  filter(!str_detect(GeneFamily, "\\|")) %>%
  full_join(., humann3_GFs_top, by = "GeneFamily") %>%
  replace(is.na(.), 0)

humann3_KO_world %>%
  fwrite(., "./05-Documentation.backup/h3_world_GFs_no_species.tsv", quote = F, sep = "\t")

```

```{r}

humann3_KO_world <- fread("./05-Documentation.backup/h3_world_GFs_no_species.tsv")

```


```{r metadata}
# load the metadata file
# metadata <- fread("./05-Documentation.backup/oceania_metadata_full.tsv") %>%
#   full_join(., fread("./05-Documentation.backup/oceana_climate.tsv")) %>%
#   mutate(Island = ifelse(Type == "blank", "Blank",Island)) %>%
#   mutate(Island = ifelse(Type == "bone", "Bone", Island)) %>%
#   mutate(Library_ID = str_replace_all(Library_ID, ".SG1.1","")) %>%
#   mutate(Island = str_replace_all(Island, "Efate_3000", "Efate 3000 BP"),
#          Island = str_replace_all(Island, "Rapa_Nui", "Rapa Nui"),
#          Island = str_replace_all(Island, "Viti_Levu", "Viti Levu"),
#          Island = str_replace_all(Island, "Taumako_Duff_Islands", "Taumako")) %>%
#   mutate(Type = str_replace_all(Type, "bone","Arch. Bone"),
#          Type = str_replace_all(Type, "calculus","Arch. calculus")) %>%
#   mutate(log_total_reads = log10(No_reads_no_human))

pacific_metadata <- fread("05-Documentation.backup/oceania_metadata_full.tsv") %>%
  select(Library_ID, Lab, Type, Age_mean, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = "Pacific",
          Continent = "Pacific",
         Age_mean_log = log10(Age_mean))

pc_end_samples <- fread("./05-Documentation.backup/pc_end_samples.tsv") %>%
  pull(Species)


     
```

```{r}
metadata <- fread("../smoking_calculus/00-documentation.backup/full_combined_metadata.tsv") %>%
  select(Library_ID, Site_code, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = str_replace_all(Site_code, "MID","Middenbeemster"),
         Study = str_replace_all(Study, "CMB","Middenbeemster"),
         Study = str_replace_all(Study, "KIL","Kilteasheen"),
         Study = str_replace_all(Study, "RAD","Radcliffe"),
         Study = str_replace_all(Study, "ELR","Iberia"),
         Study = str_replace_all(Study, "IVE","Iberia"),
         Lab = str_replace_all(Site_code, "MID","Jena"),
         Lab = str_replace_all(Lab, "CMB","Jena"),
         Lab = str_replace_all(Lab, "ELR","Jena"),
         Lab = str_replace_all(Lab, "IVE","Jena"),
         Lab = str_replace_all(Lab, "RAD","Oxford"),
         Lab = str_replace_all(Lab, "KIL","Tuebingen")) %>%
  mutate(Age_mean = ifelse(str_detect(Site_code, "MID|CMB|RAD"), 250,
                           ifelse(Site_code == "KIL", 700,
                                  ifelse(str_detect(Site_code, "ELR|IVE"), 620, 7)))) %>%
  mutate(Continent = "Europe") %>%
  full_join(., fread("./05-Documentation.backup/eisenhoffer2020_ottoni2021_deepevo_mann2018w_metadata.tsv") %>%
              mutate(Run_accession = ifelse(Study == "FellowsYates2021",Sample_alias,Run_accession)) %>%
              rename(Library_ID = Run_accession)) %>%
  mutate(Sample_type = ifelse(str_detect(Study, "EXB|LIB|DEN|SOL"),"blank",Sample_type)) %>%
  mutate(Sample_type = ifelse(is.na(Sample_type),"calculus",Sample_type)) %>%
  mutate(Sample_group = ifelse(Sample_type == "blank","blank",
                               ifelse(str_detect(Library_ID, "VLC|JAE"),"modern_calculus",Sample_group))) %>%
  mutate(Sample_group = ifelse(is.na(Sample_group),"ancient_calculus",Sample_group)) %>%
  mutate(Sample_alias = ifelse(is.na(Sample_alias),Library_ID,Sample_alias)) %>%
  mutate(Age_mean_log = log10(Age_mean))

world_metadata <- metadata %>%
  # mutate(Sample_group = ifelse(is.na(Sample_group),"bone",Sample_group)) %>%
  filter(!str_detect(Sample_group, "primate|modern"))

# list the non-human primates and modern calculus to remove from the species table
nhp <- metadata %>%
  filter(str_detect(Sample_group, "primate|modern")) %>%
  select(Sample_alias) %>%
  unique() %>%
  pull()

# and combine pacific and world
full_metadata <- pacific_metadata %>%
  bind_rows(., world_metadata)


```


```{r}
# make a list of blanks b/c they don't all have the same prefixes
blanks <- full_metadata %>%
  mutate(Type = ifelse(is.na(Type),Sample_type,Type)) %>%
  filter(str_detect(Type, "blank|bone")) %>%
  filter(!str_detect(Library_ID, "ARS")) %>%
  select(Library_ID) %>%
  pull()

```


```{r cuperdec_poor_samples}
poor_samples <- fread("./05-Documentation.backup/cfdp_fail_samples_20210413.txt") %>%
  pull(Sample)

world_fail <- fread("./05-Documentation.backup/cuperdec_malt_ei_ot_de_poor_samples.tsv") %>%
  bind_rows(., fread("05-Documentation.backup/oceania_cuperdec_mpa3_anc_params_poor_samples.tsv"))

```

```{r colors}
Island_colors = c("Tongatapu" = "#616161", 
                  "Rurutu" = "#9d654c", 
                  "Flores" = "#f6865e", 
                  "Rapa Nui" = "#cc79a7", 
                  "Futuna" = "#7D3560", 
                  "Efate" = "#7472af", 
                  "Vao" = "#098BD9",
                  "Efate 3000 BP" = "#7DCCFF", 
                  "Uripiv" = "#e1f2fe", 
                  "Viti Levu" = "#1ec1a7", 
                  "Raiatea" = "#148f77", 
                  "Taumako" = "#4e7705", 
                  "Watom" = "#86c73a")


# Set microshades colors
# Set microshades colors
Lab_colors = c("Jena" = microshades_palette("micro_cvd_purple", 3, lightest = FALSE)[3],
                "Oklahoma" =  microshades_palette("micro_cvd_blue", 3, lightest = FALSE)[3],
                "Otago" = microshades_palette("micro_cvd_green", 3, lightest = FALSE)[3],
                "Oxford" = microshades_palette("micro_cvd_turquoise", 3, lightest = FALSE)[1],
                "Tuebingen" = microshades_palette("micro_cvd_purple", 3, lightest = FALSE)[1],
                "ACAD" = microshades_palette("micro_cvd_blue", 4, lightest = FALSE)[1],
                "Vienna" = microshades_palette("micro_cvd_green", 3, lightest = FALSE)[1])


Sample_type_colors <- c("Arch. Bone" = "#148F77", "blank" = "#8B8B8B", "Arch. calculus" = "black")

Study_colors <- c("Pacific" = "#098BD9", 
                  "Middenbeemster" = "#7472af", 
                  "Radcliffe" = "#cc79a7", 
                  "Kilteasheen" = "#f6865e", 
                  "Iberia" = "#1ec1a7",
                  "Eisenhoffer2020" = "#86c73a",
                  "Ottoni2021" = "#616161",
                  "FellowsYates2021" = "#e1f2fe",
                  "Mann2018" = "#f5c710ff") # 7D3560


Continent_colors <- c("Pacific" = "#098BD9", 
                  "Europe" = "#7472af", 
                  "Africa" = "#cc79a7", 
                  "Caribbean" = "#f6865e", 
                  "Asia" = "#86c73a",
                  "North America" = "#1ec1a7")


```

```{r}
Island_shapes = c("Tongatapu" = 21, 
                  "Rurutu" = 21, 
                  "Flores" = 21, 
                  "Rapa Nui" = 21, 
                  "Futuna" = 21, 
                  "Efate" = 21, 
                  "Vao" = 21,
                  "Efate 3000 BP" = 21, 
                  "Uripiv" = 21, 
                  "Viti Levu" = 21, 
                  "Raiatea" = 21, 
                  "Taumako" = 21, 
                  "Watom" = 21)

Lab_shapes = c(21,21,21,21,21,21,21,21)
Lab_size = c(4,4,4,4,4,4,4)

Study_shapes = c(21,21,21,21,21,21,21,21,21,21,21)

Type_shapes = c(21,21,21,21,21,21,21,21)
Type_size = c(4,4,4,4,4,4,4,4)

```


```{r plot_pca_fxn}

# add functions for:
# making the input table for PCA?

# plotting PCA with colored dots
plot_pca <- function(df, pc1, pc2, color_group, shape_group) {
    metadata_group_colors <- get(paste(color_group, "_colors", sep = ""))
    metadata_group_shapes <- get(paste(shape_group, "_shapes", sep = ""))

    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%")
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(full_metadata, by = "Library_ID")
    
    color_group = df_X[[color_group]]
    shape_group = df_X[[shape_group]]

                       if (pc1 == 'PC1') {
                           pc1 <- df_X$PC1
                           exp_var_pc1 <- exp_var[1]
                           xaxis <- c("PC1")
                       }  else if (pc1 == 'PC2') {
                           pc1 <- df_X$PC2
                           exp_var_pc1 <- exp_var[2]
                           xaxis <- c("PC2")
                       } else if (pc1 == 'PC3') {
                           pc1 <- df_X$PC3
                           exp_var_pc1 <- exp_var[3]
                           xaxis <- c("PC3")
                       }

                       if (pc2 == 'PC1') {
                           pc2 <- df_X$PC1
                           exp_var_pc2 <- exp_var[1]
                           yaxis <- c("PC1")
                       }  else if (pc2 == 'PC2') {
                           pc2 <- df_X$PC2
                           exp_var_pc2 <- exp_var[2]
                           yaxis <- c("PC2")
                       } else if (pc2 == 'PC3') {
                           pc2 <- df_X$PC3
                           exp_var_pc2 <- exp_var[3]
                           yaxis <- c("PC3")
                       }

    pca_plot <- ggplot(df_X, aes(pc1, pc2)) +
                        geom_point(aes(fill = color_group, shape = shape_group), size = 3, stroke = 0.3, shape = 21) +
                        scale_fill_manual(values = metadata_group_colors) +
                        xlab(paste(xaxis, " - ", exp_var_pc1)) +
                        ylab(paste(yaxis, " - ", exp_var_pc2)) +
                        theme_minimal(base_size = 14) +
                        theme(legend.position = "right") +
                        theme(legend.title = element_blank())

    return(pca_plot)
}

```


```{r plot_bi}
# this function plots the PCA with biplots to look at the categories with strongest loadings
plot_pca_bi <- function(df, pc1, pc2, metadata_group, columntitle, GeneFamilys = c("TRUE","FALSE")) {
    metadata_group_colors <- get(paste(metadata_group, "_colors", sep = ""))
    metadata_group_shapes <- get(paste(metadata_group, "_shapes", sep = ""))
   
    arrow_pc <- enquo(columntitle)
    
    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%") # explained variance for x- and y-labels
    
    # select only the PCs from the PCA and add metadata
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(full_metadata, by = "Library_ID")
    
    metadata_group = df_X[[metadata_group]]
    
    corr_lam <- df$sdev[c("PC1", "PC2", "PC3")] * sqrt(nrow(df_X))

    df_X <- df_X %>%
      mutate(PC1 = PC1 / corr_lam[1],
             PC2 = PC2 / corr_lam[2], 
             PC3 = PC3 / corr_lam[3])

    # select the correct PC column and explained variance for PC1
    if (pc1 == 'PC1') {
        Pc1 <- df_X$PC1
        exp_var_pc1 <- exp_var[1]
        xaxis <- c("PC1")
    } else if (pc1 == 'PC2') {
        Pc1 <- df_X$PC2
        exp_var_pc1 <- exp_var[2]
        xaxis <- c("PC2")
    } else if (pc1 == 'PC3') {
        Pc1 <- df_X$PC3
        exp_var_pc1 <- exp_var[3]
        xaxis <- c("PC3")
   }
    
    # select the correct PC column and explained variance for PC2
    if (pc2 == 'PC1') {
        Pc2 <- df_X$PC1
        exp_var_pc2 <- exp_var[1]
        yaxis <- c("PC1")
 } else if (pc2 == 'PC2') {
       Pc2 <- df_X$PC2
       exp_var_pc2 <- exp_var[2]
       yaxis <- c("PC2")
    } else if (pc2 == 'PC3') {
       Pc2 <- df_X$PC3
       exp_var_pc2 <- exp_var[3]
       yaxis <- c("PC3")
   }

    # Identify the 10 GeneFamilys that have highest positive and negative loadings in the selected PC
    pws_10 <- df$loadings$X %>%
      as.data.frame(.) %>%
      rownames_to_column(var = "Function") %>%
      top_n(10, !!arrow_pc)
   if (GeneFamilys == T) {
      pws_10 <- pws_10 %>%
        rename(GeneFamily = Function) %>%
        mutate(Function = sapply(GeneFamily, function(f) {
                                  unlist(str_split(f, ":"))[1]
                           })) 
    }

    neg_10 <- df$loadings$X %>%
      as.data.frame(.) %>%
      rownames_to_column(var = "Function") %>%
      top_n(-10, !!arrow_pc)
    if (GeneFamilys == T) {
      neg_10 <- neg_10 %>%
      rename(GeneFamily = Function) %>%
       mutate(Function = sapply(GeneFamily, function(f) {
                                  unlist(str_split(f, ":"))[1]
                           })) 
    }


    pca_plot_bi <- ggplot(df_X, aes(x = Pc1, y = Pc2)) +
      geom_point(aes(shape = metadata_group, fill = metadata_group), size = 3.5, stroke = 0.3) +
      geom_segment(data = pws_10,
                   aes(xend = get(paste(pc1)), yend = get(paste(pc2))),
                   x = 0, y = 0, colour = "black",
                   size = 0.5,
                   arrow = arrow(length = unit(0.03, "npc"))) +
      geom_label_repel(data = pws_10,
                   aes(x = get(paste(pc1)), y = get(paste(pc2)), label = Function),
                   size = 1.5, colour = "grey20", label.padding = 0.2, max.overlaps = 12) +
      geom_segment(data = neg_10,
                   aes(xend = get(paste(pc1)), yend = get(paste(pc2))),
                   x = 0, y = 0, colour = "grey50",
                   size = 0.5,
                   arrow = arrow(length = unit(0.03, "npc"))) +
      geom_label_repel(data = neg_10,
                   aes(x = get(paste(pc1)), y = get(paste(pc2)), label = Function),
                   size = 1.5, colour = "grey20", label.padding = 0.2, max.overlaps = 12) +
      labs(x = paste(xaxis, " - ", exp_var_pc1),
           y = paste(yaxis, " - ", exp_var_pc2)) +
      scale_fill_manual(values = metadata_group_colors) +
      scale_shape_manual(values = metadata_group_shapes) +
      theme_minimal() + theme(text = element_text(size = 14)) +
      theme(legend.position = "right")

    return(pca_plot_bi)
}


```


```{r pca_plot_cont_fxn}
# for continuous data
plot_pca_cont <- function(df_pca, pc1, pc2, color_group, ncomps, title_text) {

    exp_var <- paste0(round(df_pca$prop_expl_var$X * 100, 2), "%")
    df_X <- df_pca$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(full_metadata, by = "Library_ID") %>%
              left_join(., h3_GF_counts, by = "Library_ID")

    color_group = df_X[[color_group]]

    if (pc1 == 'PC1') {
        pc1 <- df_X$PC1
        exp_var_pc1 <- exp_var[1]
        xaxis <- c("PC1")
    }  else if (pc1 == 'PC2') {
        pc1 <- df_X$PC2
        exp_var_pc1 <- exp_var[2]
        xaxis <- c("PC2")
    } else if (pc1 == 'PC3') {
        pc1 <- df_X$PC3
        exp_var_pc1 <- exp_var[3]
        xaxis <- c("PC3")
    }

    if (pc2 == 'PC1') {
        pc2 <- df_X$PC1
        exp_var_pc2 <- exp_var[1]
        yaxis <- c("PC1")
    }  else if (pc2 == 'PC2') {
        pc2 <- df_X$PC2
        exp_var_pc2 <- exp_var[2]
        yaxis <- c("PC2")
    } else if (pc2 == 'PC3') {
        pc2 <- df_X$PC3
        exp_var_pc2 <- exp_var[3]
        yaxis <- c("PC3")
    }

    pca_plot <- ggplot(df_X, aes(x = pc1, y = pc2)) +
      geom_point(aes(fill = color_group), size = 3.5, stroke = 0.3, shape = 21) +
     scale_fill_viridis_c(option = "C") +
     # scale_shape_manual(values = c(16, 16, 16, 16, 16,16,16,16)) +
     # scale_shape_manual(values = c(16, 17, 15, 18, 14,9,13,24)) +
    # stat_ellipse() +
     xlab(paste(xaxis, " - ", exp_var_pc1)) +
     ylab(paste(yaxis, " - ", exp_var_pc2)) +
     theme_minimal(base_size = 16) +
     theme(text = element_text(size=16)) +
     theme(legend.title = element_blank(),
           legend.key.size = unit(2,"mm"),
           legend.text = element_text(size = 10)) + 
     theme(legend.position = "right") +
     ggtitle(title_text) + theme(plot.title = element_text(size = 10))

    return(pca_plot)
}

```

# GeneFamily analysis

## Run hierarchical clustering on the samples 
```{r get_GF_totals}
# how many pathways were identified per cluster?

h3_GF_counts_prep <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  # and remove these 2 b/c they have no age data for the adonis test
  select(-ERR5729638, -ERR5729643) %>%
  # inner_join(., humann3_GF.decontam_noblanks_presence_more_30, by = "GeneFamily") %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  as_tibble()

# list samples with less than 750 GeneFamilies identified (All samples with less fall into Cluster 2, which is probably biasing the comparison of the 2 clusters)
under750 <- h3_GF_counts_prep %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts > 0,1,0)) %>%
  group_by(Library_ID) %>%
  summarize(total_GFs = sum(Counts)) %>%
  ungroup() %>%
  filter(total_GFs < 750) %>%
  pull(Library_ID)

```

```{r}

# also read in the number of species in the final cleaned species table for comparison between the # of species detected and the # of GFs in each sample

sp_counts <- fread("./05-Documentation.backup/pe_all_table_species_counts.tsv")

```


Plot a PCA with only the well-preserved samples and filter the pathways by abundance
```{r pca_GFs_filt_noblanks}
humann3_gf_l1_noblanks_filt <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(GeneFamily,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_gf_pca <- tune.pca(humann3_gf_l1_noblanks_filt, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_gf_l1_noblanks_filt.pca <- pca(humann3_gf_l1_noblanks_filt, ncomp = 3, logratio = 'CLR')

# plot the PCA
h3_gf_pc12 <- plot_pca(humann3_gf_l1_noblanks_filt.pca, "PC1", "PC2", "Study", "Lab")
h3_gf_pc12
plot_pca(humann3_gf_l1_noblanks_filt.pca, "PC1", "PC2", "Lab", "Lab")

# ggsave("./06-publication/prelim_figs/h3_GFs_pc12.pdf", plot = h3_gf_pc12, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"), dpi = 300)


```

```{r}

# does making the heatmap with non-CLR-transformed data look different?
# Yup it does, maybe stick to the CLR-transformed data
humann3_gf_l1_noblanks_filt_mod <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  # pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  # spread(GeneFamily,Counts) %>%
  separate(GeneFamily, into = "KEGG", extra = "drop", sep = ":") %>%
  column_to_rownames("KEGG") %>%
  as.matrix(.)


humann3_gf_l1_noblanks_filt.clr <- humann3_gf_l1_noblanks_filt.pca$X %>%
  as.data.frame.matrix() %>%
  rownames_to_column("Library_ID") %>%
  pivot_longer(!Library_ID, names_to = "kegg", values_to = "counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "counts") %>%
  separate(kegg, into = "KEGG", extra = "drop", sep = ":") %>%
  column_to_rownames("KEGG")

humann3_gf_l1_noblanks_filt.clr.mat <- as.matrix(humann3_gf_l1_noblanks_filt.clr)

# humann3_gf_l1_noblanks_filt.euc.mat <- as.matrix(humann3_gf_l1_noblanks_filt.euc)

selected_metadata <- world_metadata %>%
  full_join(., pacific_metadata) %>%
  select(Library_ID, Study, Lab, Age_mean, Continent) %>%
  inner_join(., humann3_gf_l1_noblanks_filt.pca$X %>%
               as.data.frame.matrix(.) %>%
               as.data.frame(.) %>%
               rownames_to_column("Library_ID") %>%
               select(Library_ID), by = "Library_ID") %>%
  arrange(Library_ID) %>%
  column_to_rownames("Library_ID")

col_age = circlize::colorRamp2(c(0, 3000), c("#098BD9", "#7D3560"))
col_fun = circlize::colorRamp2(c(-5, 0, 5), c("#098BD9", "#ffffff", "#f5c710ff")) # 75, 130


row_ha <- ComplexHeatmap::HeatmapAnnotation(df = selected_metadata, col = list(Study = Study_colors,
                                                                               Lab = Lab_colors,
                                                                               Age_mean = col_age,
                                                                               Continent = Continent_colors))


set.seed(1)
ht1 = ComplexHeatmap::Heatmap(humann3_gf_l1_noblanks_filt.clr.mat, col = col_fun, name = "CPM", clustering_method_rows = "complete", clustering_method_columns = "complete",column_km = 2, row_km = 3, top_annotation = row_ha, border = TRUE, show_row_names = F, show_column_names = T, column_names_gp = grid::gpar(fontsize = 4))

ht1 = ComplexHeatmap::draw(ht1)

roworder <- ComplexHeatmap::row_order(ht1)
colorder <- ComplexHeatmap::column_order(ht1)


# and save the heatmap as an svg
# svg(file = "./06-publication/supplemental_figures/Sup_fig_6/KEGG_KO_heatmap.svg", width = 14, height = 12)
# 
# set.seed(1)
# ht1 = ComplexHeatmap::Heatmap(humann3_gf_l1_noblanks_filt.clr.mat, col = col_fun, name = "CPM", clustering_method_rows = "complete", clustering_method_columns = "complete",column_km = 2, row_km = 3, top_annotation = row_ha, border = TRUE, show_row_names = F, show_column_names = F)
# 
# ComplexHeatmap::draw(ht1)
# 
# dev.off()


```


```{r row_order}
# turn the row numbers into a data table to combine with the matrix to get the KEGG entries
row_clusters <- roworder$`3` %>%
  as.data.frame(.) %>%
  rename(line_number = 1) %>%
  mutate(Cluster = 3) %>%
  bind_rows(., roworder$`2` %>%
            as.data.frame(.) %>%
            rename(line_number = 1) %>%
            mutate(Cluster = 2)) %>%
  bind_rows(., roworder$`1` %>%
            as.data.frame(.) %>%
            rename(line_number = 1) %>%
            mutate(Cluster = 1)) %>%
  mutate(dendro_order = 1:n())
  
# add row numbers to the KEGG orthologs of the input matrix
kegg_cluster_info <- humann3_gf_l1_noblanks_filt.clr.mat %>%
  as.data.frame(.) %>%
  rownames_to_column("KEGG")  %>%
  select(KEGG) %>%
  mutate(line_number = 1:n()) %>%
  full_join(., row_clusters, by  = "line_number") %>%
  # now get the full entry with description from the original table
  full_join(., humann3_gf_l1_noblanks_filt %>%
  rownames_to_column("Library_ID")  %>%
  pivot_longer(!Library_ID, names_to = "pathway", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  select(pathway) %>%
  mutate(double = pathway) %>%
  separate(double, into = "KEGG", extra = "drop", sep = ":"), by = "KEGG") %>%
  arrange(desc(Cluster), dendro_order)

# fwrite(kegg_cluster_info, "./05-Documentation.backup/humann3_GFs_world_kegg_clustering.tsv", quote = F, sep = "\t")


# col = col_fun, 

# humann3_gf_l1_noblanks_filt.clr %>%
#   as.data.frame() %>%
#   rownames_to_column("Library_ID") %>%
#   pivot_longer(!Library_ID, names_to = "kegg", values_to = "counts") %>%
#   filter(counts >  0) %>%
#   slice_max(counts)
  
```


```{r col_order}
# turn the row numbers into a data table to combine with the matrix to get the KEGG entries
column_clusters <- colorder$`2` %>%
  as.data.frame(.) %>%
  rename(line_number = 1) %>%
  mutate(Cluster = 2) %>%
  bind_rows(., colorder$`1` %>%
            as.data.frame(.) %>%
            rename(line_number = 1) %>%
            mutate(Cluster = 1)) %>%
  mutate(dendro_order = 1:n())
  
# add row numbers to the KEGG orthologs of the input matrix
sample_cluster_info <- humann3_gf_l1_noblanks_filt.clr.mat %>%
  as.data.frame(.) %>%
  rownames_to_column("KEGG")  %>%
  pivot_longer(!KEGG, names_to = "Library_ID", values_to = "counts") %>%
  pivot_wider(names_from = "KEGG", values_from = "counts") %>%
  select(Library_ID) %>%
  mutate(line_number = 1:n()) %>%
  full_join(., column_clusters, by  = "line_number") %>%
  arrange(desc(Cluster), dendro_order)

# fwrite(sample_cluster_info, "./05-Documentation.backup/humann3_GFs_world_sample_clustering.tsv", quote = F, sep = "\t")


```


```{r}
# put the KOs of groups 2 and 3 into https://www.genome.jp/kegg/mapper/color.html to get an idea of the pathways they're in
# read the file from copying the output of that search
# also read in the table that matches the map# and the high-level pathway information
paths_details <- fread("./05-Documentation.backup/KEGG_map_pathway_info.tsv") %>%
  mutate(map = str_replace_all(map, "map","")) %>%
  separate(pathway, into = c("Path","path"), sep = " ", extra = "merge") %>%
  select(-Path) %>%
  rename(pathway = path)

keggtable <- fread("./05-Documentation.backup/KEGG_color_humann3_GFs_world_list.tsv") 

```

```{r}
# read in the json file with all KOs per pathway

result <- jsonlite::fromJSON("~/Downloads/ko00001.json", flatten = T)

json_data_frame <- tibble(as.data.frame(result))

json_df <- unnest(json_data_frame)
json_df <- json_df %>%
  rename(nextone = children)

json_df2 <- unnest(json_df)
json_df2 <- json_df2 %>%
  rename(nowone = children)

json_df3 <- unnest(json_df2)

json_df3 <- json_df3 %>%
  separate(name3, into = c("kegg","name"), sep = "; ", extra = "merge") %>%
  separate(kegg, into = c("KO","gene"), sep = "  ", extra = "merge")


json_df3 %>%
  select(name1) %>%
  unique()
  
```

```{r}
# now combine the json file and the kegg_cluster_info table to get the pathways for each KO in each cluster
# # remove pathways that are eukaryote or human-specific
# # from 09140 Cellular Processes
# 09141 Transport and catabolism
# 09143 Cell growth and death
# 09144 Cellular community - eukaryotes
# # from 09150 Organismal Systems - all
# # from 09160 Human Diseases 
# 09161 Cancer: overview
# 09162 Cancer: specific types
# 09172 Infectious disease: viral
# 09171 Infectious disease: bacterial
# 09163 Immune disease
# 09164 Neurodegenerative disease
# 09166 Cardiovascular disease
# 09167 Endocrine and metabolic disease
# 09176 Drug resistance: antineoplastic

cluster_ko_paths <- kegg_cluster_info %>%
  rename(KO = KEGG) %>%
  select(-line_number, -dendro_order) %>%
  left_join(., json_df3 %>%
              filter(!str_detect(`children.name`,"09150")) %>%
              select(name1, name2, KO), by = "KO") %>%
  # remove the unwanted pathways
  filter(!str_detect(name1, "09141|09143|09144|09161|09162|09171|09172|09163|09164|09166|09167|09176"))



# order the pathways by decreasing % in Cluster3
cl1_order_n1 <- cluster_ko_paths %>%
  group_by(Cluster, name1) %>%
  count() %>%
  arrange(Cluster, desc(n)) %>%
  ungroup() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  pivot_wider(names_from = "Cluster", values_from = "n", values_fill = 0) %>%
  pivot_longer(!name1, names_to = "Cluster", values_to = "n") %>%
  group_by(Cluster) %>%
  mutate(Percent = n / sum(n) * 100) %>%
  ungroup() %>%
  arrange(Cluster, Percent) %>%
  filter(Cluster == 1) %>%
  separate(name1, into = c("number","path"), sep = " ", extra = "merge") %>%
  pull(path)

  
fig <- cluster_ko_paths %>%
  group_by(Cluster, name1) %>%
  count() %>%
  arrange(Cluster, desc(n)) %>%
  ungroup() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  pivot_wider(names_from = "Cluster", values_from = "n", values_fill = 0) %>%
  pivot_longer(!name1, names_to = "Cluster", values_to = "n") %>%
  group_by(Cluster) %>%
  mutate(Percent = n / sum(n) * 100) %>%
  ungroup() %>%
  filter(Cluster != 2) %>%
  mutate(Cluster = as.character(Cluster)) %>%
  separate(name1, into = c("number","name1"), sep = " ", extra = "merge") %>%
  select(-number) %>%
  mutate(name1 = fct_relevel(name1, cl1_order_n1)) %>%
  ggplot(., aes(x = name1, y = Percent, fill = Cluster)) +
         geom_bar(stat = "identity", position = position_dodge2(width = 0.9, preserve = "single"), color = "black") +
         theme_minimal(base_size = 11) +
         theme(axis.text.x = element_text(angle = 0, hjust = 1.0, vjust = 0.5),
               # axis.text.y = element_text(vjust = -0.1),
               axis.text = element_text(size = 10)) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=8)) +
         scale_fill_manual(values = c("#f5c710ff","#098BD9")) +
         coord_flip() +
         ylab("% of KOs per cluster") +
         xlab("Pathway")
fig

# ggsave("./06-publication/prelim_figs/KOs_diffs_clusters.pdf", plot = fig, device = "pdf",
#        scale = 1, width = 7, height = 7, units = c("in"), dpi = 300)
      
```

```{r}

# significant differences between clusters for any of the pathways?
pathlist <- cluster_ko_paths %>%
  group_by(Cluster, name1) %>%
  count() %>%
  ungroup() %>%
  filter(Cluster != 2) %>%
  mutate(Cluster = as.character(Cluster),
         Cluster = ifelse(Cluster == "1","cluster1","cluster2")) %>%
  pivot_wider(names_from = "Cluster", values_from = "n") %>%
  separate(name1, into = c("map","pathway"), sep = " ", extra = "merge") %>%
  pull(pathway) 

cl1list <- cluster_ko_paths %>%
  group_by(Cluster, name1) %>%
  count() %>%
  ungroup() %>%
  filter(Cluster != 2) %>%
  mutate(Cluster = as.character(Cluster),
         Cluster = ifelse(Cluster == "1","cluster1","cluster3")) %>%
  pivot_wider(names_from = "Cluster", values_from = "n") %>%
  replace(is.na(.), 0) %>%
  pull(cluster1) 

cl3list <- cluster_ko_paths %>%
  group_by(Cluster, name1) %>%
  count() %>%
  ungroup() %>%
  filter(Cluster != 2) %>%
  mutate(Cluster = as.character(Cluster),
         Cluster = ifelse(Cluster == "1","cluster1","cluster3")) %>%
  pivot_wider(names_from = "Cluster", values_from = "n") %>%
  replace(is.na(.), 0) %>%
  pull(cluster3) 

intab <- as.table(cbind(
  cl1list, cl3list
))

dimnames(intab) <- list(
  pathway = pathlist,
  Cluster = c("Cluster1","Cluster3")
)

intab %>%
  row_wise_prop_test(., alternative = "greater", p.adjust.method = "fdr")

intab %>%
  row_wise_fisher_test(., alternative = "greater", p.adjust.method = "fdr")



```

```{r}

# since only the Signal transduction (09132) and Protein families: genetic information processing (09182) pathways are 
# significantly different, let's look at the species that contributed those KOs in each cluster

# order the Signal transduction pathways
cl1_order <- cluster_ko_paths %>%
  filter(str_detect(name1, "09132")) %>%
  group_by(Cluster, name2) %>%
  count() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  pivot_wider(names_from = "Cluster", values_from = "n", values_fill = 0) %>%
  pivot_longer(!name2, names_to = "Cluster", values_to = "n") %>%
  filter(Cluster == 1) %>%
  arrange(Cluster, n) %>%
  ungroup() %>%
  separate(name2, into = "carb", sep = "\\[", extra = "drop") %>%
  separate(carb, into = c("number","path"), sep = " ", extra = "merge") %>%
  select(path, number, everything()) %>%
  mutate(number = str_replace_all(number, "^0","- ko0")) %>%
  unite(name2, path:number, sep = " ") %>%
  pull(name2)


# Signal transduction pathway KOs enriched in sample cluster 2 are in KEGG cluster 1?
fig2 <- cluster_ko_paths %>%
  filter(str_detect(name1, "09132")) %>%
  filter(Cluster != 2) %>%
  select(Cluster, name2) %>%
  group_by(Cluster, name2) %>%
  count() %>%
  # arrange(Cluster, desc(n)) %>%
  ungroup() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  separate(name2, into = "carb", sep = "\\[", extra = "drop") %>%
  separate(carb, into = c("number","path"), sep = " ", extra = "merge") %>%
  select(path, number, everything()) %>%
  mutate(number = str_replace_all(number, "^0","- ko0")) %>%
  unite(name2, path:number, sep = " ") %>%
  mutate(name2 = fct_relevel(name2, cl1_order)) %>%
  arrange(name2) %>%
  ggplot(., aes(x = name2, y = n, fill = Cluster)) +
         geom_bar(stat = "identity", position = position_dodge2(width = 0.9, preserve = "single"), color = "black") +
         theme_minimal(base_size = 11) +
         theme(axis.text.x = element_text(angle = 0, hjust = 1.0, vjust = 0.5),
               axis.text = element_text(size = 10)) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         scale_fill_manual(values = c("#f5c710ff","#098BD9")) +
         ylab("No. KOs") +
         xlab("Signal transduction pathway") +
         coord_flip()
fig2

# ggsave("./06-publication/prelim_figs/KOs_carbs_clusters.pdf", plot = fig2, device = "pdf",
#        scale = 1, width = 8, height = 4, units = c("in"), dpi = 300)

# order the Protein families: genetic information processing pathways
cl1_order <- cluster_ko_paths %>%
  filter(str_detect(name1, "09182")) %>%
  group_by(Cluster, name2) %>%
  count() %>%
  ungroup() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  pivot_wider(names_from = "Cluster", values_from = "n", values_fill = 0) %>%
  pivot_longer(!name2, names_to = "Cluster", values_to = "n") %>%
  filter(Cluster == 1) %>%
  arrange(Cluster, n) %>%
  separate(name2, into = "carb", sep = "\\[", extra = "drop") %>%
  separate(carb, into = c("number","path"), sep = " ", extra = "merge") %>%
  select(path, number, everything()) %>%
  mutate(number = str_replace_all(number, "^0","- ko0")) %>%
  unite(name2, path:number, sep = " ") %>%
  pull(name2)


# Protein families: genetic information processing pathway KOs enriched in sample cluster 1 are in KEGG cluster 3?
fig2 <- cluster_ko_paths %>%
  filter(str_detect(name1, "09182")) %>%
  filter(Cluster != 2) %>%
  select(Cluster, name2) %>%
  group_by(Cluster, name2) %>%
  count() %>%
  # arrange(Cluster, desc(n)) %>%
  ungroup() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  separate(name2, into = "carb", sep = "\\[", extra = "drop") %>%
  separate(carb, into = c("number","path"), sep = " ", extra = "merge") %>%
  select(path, number, everything()) %>%
  mutate(number = str_replace_all(number, "^0","- ko0")) %>%
  unite(name2, path:number, sep = " ") %>%
  mutate(name2 = fct_relevel(name2, cl1_order)) %>%
  arrange(name2) %>%
  ggplot(., aes(x = name2, y = n, fill = Cluster)) +
         geom_bar(stat = "identity", position = position_dodge2(width = 0.9, preserve = "single"), color = "black") +
         theme_minimal(base_size = 11) +
         theme(axis.text.x = element_text(angle = 0, hjust = 1.0, vjust = 0.5),
               axis.text = element_text(size = 10)) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         scale_fill_manual(values = c("#f5c710ff","#098BD9")) +
         ylab("No. KOs") +
         xlab("Protein families: genetic information processing pathway") +
         coord_flip()
fig2

```

```{r}

# maybe make a heatmap of all the KOs in the Two-component system pathway in Cluster 1 with all samples in the order of the dendrogram?
two_comp_ko <- json_df3 %>% 
  filter(str_detect(name2, "Two-component") & str_detect(name1, "09132")) %>%
  select(KO) %>%
  unique() %>%
  inner_join(., kegg_cluster_info %>%
              filter(Cluster != 2) %>%
              select(KEGG) %>%
              rename(KO = KEGG), by = "KO") %>%
  pull(KO)

sample_dendro_order <- sample_cluster_info %>%
  arrange(dendro_order) %>%
  pull(Library_ID)

ko_dendro_order <- kegg_cluster_info %>%
  arrange(desc(dendro_order)) %>%
  pull(KEGG)

vector_color = sample_cluster_info %>%
  arrange(dendro_order) %>%
  mutate(Color = ifelse(Cluster == 1,"#098BD9","#f5c710ff")) %>%
  pull(Color) %>%
  as.character()

humann3_gf_l1_noblanks_filt %>%
  rownames_to_column("Library_ID") %>%
  pivot_longer(!Library_ID, names_to = "ortholog", values_to = "CPM") %>%
  pivot_wider(names_from = "Library_ID", values_from = "CPM") %>%
  separate(ortholog, into = c("KO","Ortholog"), sep = ": ", extra = "merge") %>%
  filter(str_detect(KO, two_comp_ko %>% str_c(collapse = "|"))) %>%
  select(-Ortholog) %>%
  pivot_longer(!KO, names_to = "Library_ID", values_to = "CPM") %>%
  mutate(log_CPM = log10(CPM)) %>%
  # now add the cluster information adn the dendrogram order
  left_join(., sample_cluster_info, by = "Library_ID") %>%
  arrange(dendro_order) %>%
  mutate(Library_ID = fct_relevel(Library_ID, sample_dendro_order),
         KO = fct_relevel(KO, ko_dendro_order)) %>%
  arrange(Library_ID, KO) %>%
  ggplot(., aes(Library_ID, KO, fill = log_CPM)) +
    geom_tile() +
    scale_fill_viridis_c(option = "A") +
    theme_minimal(base_size = 7) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, color = vector_color),
          plot.margin = margin(0, 0, 0, 0, "pt"),
          axis.title.y = element_blank(),
          legend.position = "left")

```



```{r, eval = F}
#  now check for statistical differences between the clusters for each pathway

# significant differences between clusters for any of the pathways?
carblist <- cluster_ko_paths %>%
  filter(str_detect(name1, "09101")) %>%
  group_by(Cluster, name2) %>%
  count() %>%
  # arrange(Cluster, desc(n)) %>%
  ungroup() %>%
  filter(Cluster == 1) %>%
  separate(name2, into = "carb", sep = "\\[", extra = "drop") %>%
  separate(carb, into = c("number","name2"), sep = " ", extra = "merge") %>%
  pull(name2)

cl1list <- cluster_ko_paths %>%
  filter(str_detect(name1, "09101")) %>%
  group_by(Cluster, name2) %>%
  count() %>%
  ungroup() %>%
  filter(Cluster != 2) %>%
  mutate(Cluster = as.character(Cluster),
         Cluster = ifelse(Cluster == "1","cluster1","cluster3")) %>%
  pivot_wider(names_from = "Cluster", values_from = "n") %>%
  replace(is.na(.), 0) %>%
  pull(cluster1) 

cl3list <- cluster_ko_paths %>%
  filter(str_detect(name1, "09101")) %>%
  group_by(Cluster, name2) %>%
  count() %>%
  ungroup() %>%
  filter(Cluster != 2) %>%
  mutate(Cluster = as.character(Cluster),
         Cluster = ifelse(Cluster == "1","cluster1","cluster3")) %>%
  pivot_wider(names_from = "Cluster", values_from = "n") %>%
  replace(is.na(.), 0) %>%
  pull(cluster3) 

intab <- as.table(cbind(
  cl1list, cl3list
))

dimnames(intab) <- list(
  pathway = carblist,
  Cluster = c("Cluster1","Cluster3")
)

intab

intab %>%
  row_wise_fisher_test(., alternative = "greater", p.adjust.method = "holm") %>%
  arrange(`p.adj`)


```

```{r ko_species}
# Which species contribute to the pathways in SampleCluster1/KOCluster1, SampleCluster1/KOCluster3, SampleCluster2/KOCluster1, SampleCluster2/KOCluster3?
# get unique KOs in the TCS paths (remember no overlap in the KOs of cluster1 and cluster3)
# tcs_kos <- cluster_ko_paths %>%
#   filter(str_detect(name1, "09132"), str_detect(name2, "Two-component")) %>%
#   filter(Cluster != 2) %>%
#   select(Cluster, KO, name2) %>%
#   arrange(Cluster, name2, KO) %>%
#   mutate(Cluster = as.character(Cluster),
#          Cluster = ifelse(Cluster == "1","cluster1","cluster3")) %>%
#   select(KO) %>%
#   unique() %>%
#   pull(KO)
  

# now get the species of these KOs from the original table
# modify this section, copied from below
plot_table <- humann3_KO_worldo %>%
  full_join(., humann3_GFs, by = "GeneFamily") %>%
  replace(is.na(.), 0) %>%
  # remove non-human primates and modern samples
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  # get only the carbohydrate KOs
  filter(str_detect(GeneFamily, two_comp_ko %>% str_c(collapse = "|"))) %>%
  # select only the levels with species
  filter(str_detect(GeneFamily, "\\|")) %>%
  # now break up the GeneFamily long name
  separate(GeneFamily, into = c("KO","other"), sep = ": ") %>%
  separate(other, into = c("Gene_name","Species"), sep = "\\|") %>%
  separate(Species, into = c("Genus","Species"), sep = "\\.s__") %>%
  mutate(Genus = str_replace_all(Genus, "g__","")) %>%
  # convert to long table to add clusters
  pivot_longer(!KO:Species, names_to = "Library_ID", values_to = "CPM") %>%
  left_join(., sample_cluster_info %>% 
              select(Library_ID, Cluster), by = "Library_ID") %>%
  distinct()  %>%
  left_join(., cluster_ko_paths %>%
              select(KO, name1, name2) %>%
              filter(str_detect(name2, "02020")), by = "KO")

```

```{r}
# calculate the % for each ortholog contributed by each genus         
sample_cluster1_stats <- plot_table %>%
  select(-Gene_name, -Species) %>%
  filter(Cluster == 1) %>%
  distinct() %>%
  group_by(KO, Genus) %>%
  summarize(Sum = sum(CPM)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  ungroup(.)

# use the two_comp_ko list

# create the list of TCS KOs
# humann3_path_biplot_pc <- cluster_ko_paths %>%
#   filter(str_detect(name1, "09132")) %>%
#   filter(Cluster == 1) %>%
#   select(Cluster, KO, name2) %>%
#   # filter(str_detect(name2, "Glycolysis")) %>%
#   unique() %>%
#   pull(KO)

# calculate the total % of all genera that contribute < X% to each ortholog
humann3_path_pc1pws_stats_extra <- lapply(two_comp_ko, function(eclass) {
 high_percent <- sample_cluster1_stats %>%
   filter(KO == eclass) %>%
   filter(Percent < 5) %>%
   summarise(Remaining = sum(Percent)) %>%
   mutate(KO = eclass,
          Genus = "Other")
}) %>%
 bind_rows(.)

# add this additional % to the main table
humann3_path_pcbi_bar_df <- humann3_path_pc1pws_stats_extra %>%
  rename(Percent = Remaining) %>%
  bind_rows(., sample_cluster1_stats %>%
              select(-Sum)) %>% 
  select(KO, Genus, Percent) %>%
  distinct() %>%
  left_join(., cluster_ko_paths %>%
              filter(str_detect(KO, two_comp_ko  %>% str_c(collapse = "|"))) %>%
              select(KO, name1, name2), by = "KO") %>%
  filter(str_detect(name2, "02020")) %>%
  separate(name2, into = "pathway", sep = "\\[", extra = "drop") %>%
  # filter(str_detect(pathway, "02020")) %>%
  arrange(KO)

```

```{r}

ko_cl1_tcs <- kegg_cluster_info %>%
  filter(str_detect(KEGG, two_comp_ko %>% str_c(collapse = "|"))) %>%
  filter(Cluster == 1) %>%
  arrange(dendro_order) %>%
  pull(KEGG)

kegg_cluster_info %>%
  filter(str_detect(KEGG, two_comp_ko %>% str_c(collapse = "|"))) %>%
  filter(Cluster == 1) %>%
  arrange(dendro_order)


fig3 <- humann3_path_pcbi_bar_df %>%
  # filter for only the KOs in KO cluster 1
  filter(str_detect(KO, ko_cl1_tcs %>% str_c(collapse = "|"))) %>%
  filter(Percent >= 5 | (Percent <= 5 & Genus == "Other")) %>% # filter out the genera with % < 5, but keep Other < 5
  mutate(Genus = fct_relevel(Genus, "Other","unclassified"),
         KO = fct_relevel(KO, ko_cl1_tcs)) %>%
  # mutate(Genus = fct_relevel(Genus, "Other","unclassified","Aggregatibacter","Capnocytophaga","Cardiobacterium","Eikenella","Haemophilus","Kingella","Lautropia","Neisseria","Ottowia","Streptococcus"),
  #        Path = fct_relevel(Path, humann3_pathway_biplot_list %>%
  #                             filter(Direction == "PC1+") %>%
  #                             pull(Path))) %>%
  ggplot(., aes(x=KO, y=Percent, fill = Genus)) +
    geom_bar(stat = "identity") +
    theme_minimal() +
    scale_fill_manual(values = c("#616161","#8B8B8B", # microshades_cvd_palette
                                 "#7D3560","#CC79A7","#E794C1", # "#A1527F",
                                 "#098BD9","#7DCCFF","#BCE1FF", # "#56B4E9",
                                 "#4E7705","#97CE2F","#BDEC6F", # "#6D9F06",
                                 "#148F77","#43BA8F","#48C9B0","#9D654C")) + #. "#009E73",
     theme(text = element_text(size=9),
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
    ylab("Percent") +
    xlab("KEGG Ortholog") +
    theme(title = element_text(size=10)) +
    facet_grid(~pathway, scales = "free_x", space = 'free')
fig3

# ggsave("./06-publication/prelim_figs/KOs_carbs_clusters_species.pdf", plot = fig3, device = "pdf",
#        scale = 1, width = 12, height = 5, units = c("in"), dpi = 300)

```

```{r ottowia}

ottowia_sp <- fread("./05-Documentation.backup/ottowia_sp_counts.tsv")

ott_table <- humann3_KO_worldo %>%
  full_join(., humann3_GFs, by = "GeneFamily") %>%
  replace(is.na(.), 0) %>%
  # remove non-human primates and modern samples
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  # get only Ottowia
  filter(str_detect(GeneFamily, "Ottowia")) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "CPM") %>%
  separate(GeneFamily, into = c("KO","other"), sep = ": ") %>%
  separate(other, into = c("Gene_name","Species"), sep = "\\|") %>%
  separate(Species, into = c("Genus","Species"), sep = "\\.s__") %>%
  select(-Genus, -Species, -Gene_name) 

ott_table %>%
  group_by(Library_ID) %>%
  summarize(mean_cpm = mean(CPM)) %>%
  ungroup() %>%
  full_join(., ottowia_sp) %>%
  drop_na(mean_cpm) %>%
  drop_na(CLR) %>%
  left_join(., metadata %>%
              select(Library_ID, Continent)) %>%
  left_join(., sample_cluster_info, by = "Library_ID") %>%
  mutate(Continent = ifelse(is.na(Continent),"Pacific",Continent)) %>%
  mutate(Cluster = as.character(Cluster)) %>%
  ggplot(., aes(x = Counts, y = CLR)) + # Species
         # geom_point(aes(fill = Continent, shape = Cluster), size = 2.5, stroke = 0.2) +
         ggstar::geom_star(aes(fill = Continent, starshape = Cluster), size = 1.5) + # , shape = 21
         # geom_smooth(method = glm, se = F, col = "black", linewidth = 0.7) +
         scale_fill_manual(values = Continent_colors) +
         scale_color_manual(values = Continent_colors) +
         ggstar::scale_starshape_manual(values = c(15,9)) +
         theme_classic(base_size = 12) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         # ggpubr::stat_regline_equation(label.y = 1, aes(label = ..rr.label..)) +
         # scale_y_log10() +
         ylab("Ottowia Counts (CLR)") +
         xlab("Ottowia mean CPM")

# make it the # of KOs contributed by Ottowia instead of mean CPM

```

```{r}

ott_table %>%
  filter(CPM > 0) %>%
  group_by(Library_ID) %>%
  count(name = "no_KOs") %>%
  ungroup() %>%
  full_join(., ottowia_sp) %>%
  drop_na(no_KOs) %>%
  drop_na(CLR) %>%
  left_join(., metadata %>%
              select(Library_ID, Continent)) %>%
  left_join(., sample_cluster_info, by = "Library_ID") %>%
  mutate(Continent = ifelse(is.na(Continent),"Pacific",Continent)) %>%
  mutate(Cluster = as.character(Cluster)) %>%
  ggplot(., aes(x = no_KOs, y = log_counts)) + # Species
         # geom_point(aes(fill = Continent, shape = Cluster), size = 2.5, stroke = 0.2) +
         ggstar::geom_star(aes(fill = Continent, starshape = Cluster), size = 1.5) + # , shape = 21
         geom_smooth(method = glm, se = F, col = "black", linewidth = 0.7) +
         scale_fill_manual(values = Continent_colors) +
         scale_color_manual(values = Continent_colors) +
         ggstar::scale_starshape_manual(values = c(15,9)) +
         # scale_shape_manual(values = c(21,24)) +
         theme_classic(base_size = 12) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         ggpubr::stat_regline_equation(label.y = 6, aes(label = ..rr.label..)) +
         ylab("Ottowia log10(Counts)") +
         xlab("Ottowia no. of KOs")

```

```{r}
options(scipen = 999)
library(scales)
ottowia_sp %>%
  left_join(., metadata %>%
              select(Library_ID, Continent)) %>%
  left_join(., sample_cluster_info, by = "Library_ID") %>%
  mutate(Continent = ifelse(is.na(Continent),"Pacific",Continent)) %>%
  mutate(Cluster = as.character(Cluster)) %>%
  drop_na(Cluster) %>%
  # filter(Counts > 1) %>%
  ggplot(., aes(x = Cluster, y = Counts, fill = Cluster)) +
         geom_boxplot(outlier.color = NA) +
         geom_point(position = position_jitterdodge(jitter.width = 0.5), alpha = 0.5) +
         scale_fill_manual(values = c("#7D3560","#098BD9")) +
         theme_minimal() +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         scale_y_log10() +
         # scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         ylab("Ottowia counts") +
         xlab("Cluster") +
         ggtitle("")

ottowia_sp %>%
  left_join(., metadata %>%
              select(Library_ID, Continent)) %>%
  left_join(., sample_cluster_info, by = "Library_ID") %>%
  mutate(Continent = ifelse(is.na(Continent),"Pacific",Continent)) %>%
  mutate(Cluster = as.character(Cluster)) %>%
  drop_na(Cluster) %>%
  kruskal_test(Counts ~ Cluster)
  
```

```{r}
# make the Library_ID dendrogram order a vector to factorize by
sample_den_order <- sample_cluster_info %>%
  arrange(dendro_order) %>%
  pull(Library_ID)

# add missing information from the species table, adding +1 like the others have for log-transformation
# Somehow GOY006.A0101 and LIS002.A0101 are missing from the species tables
missing_ot <- rbind(c("FUM003.A0101",1314), c("GOY006.A0102",1),c("ZAF001.A0101",1), c("LIS002.A0101",1)) %>%
  as_tibble() %>%
  rename(Library_ID = 1,
         Counts = 2) %>%
  mutate(Counts = as.numeric(Counts),
       log_counts = log10(Counts)) %>%
  mutate(log_counts = ifelse(str_detect(Library_ID, "GOY"),NA,log_counts)) %>%
  mutate(log_counts = ifelse(str_detect(Library_ID, "LIS"),NA,log_counts))

ott_abund <- ottowia_sp %>%
  full_join(., missing_ot) %>%
  mutate(Counts = Counts / 100000) %>%
  mutate(Library_ID = fct_relevel(Library_ID, sample_den_order)) %>%
  arrange(Library_ID) %>%
  ggplot(., aes(x=Library_ID, y=Counts)) +
    geom_bar(stat = "identity") +
    theme_classic() +
    theme(text = element_text(size=9),
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
    theme(axis.text.x = element_blank()) +
    # geom_hline(yintercept = 25000, linetype = "dotted") +
    theme(panel.grid.major.y = element_line()) +
    # scale_y_log10() +
    # ylim(NA,500000) +
    ylab("Ottowia counts. (x10^5)") +
    xlab("Library") +
    theme(title = element_text(size=10))
ott_abund

# ggsave("./06-publication/main_figures/Figure_7/ottowia_abund.pdf", plot = ott_abund, device = "pdf",
#        scale = 1, width = 17, height = 2, units = c("in"), dpi = 300)


# which continents have > 25000 reads per sample?
ottowia_sp %>%
  full_join(., missing_ot) %>%
  left_join(., metadata %>%
              select(Library_ID, Continent)) %>%
  left_join(., sample_cluster_info, by = "Library_ID") %>%
  mutate(Continent = ifelse(is.na(Continent),"Pacific",Continent)) %>%
  mutate(Cluster = as.character(Cluster)) %>%
  filter(Counts >= 25000) %>%
  arrange(Cluster, desc(Counts))
  # arrange(Cluster,  Continent, Library_ID)
  
  
```


```{r, eval = F}

carb_kos <- cluster_ko_paths %>%
  filter(str_detect(name1, "09101")) %>%
  filter(Cluster != 2) %>%
  select(KO) %>%
  distinct() %>%
  arrange(KO) %>%
  pull(KO)


carb_ko_count_table <- humann3_KO_worldo %>%
  full_join(., humann3_GFs, by = "GeneFamily") %>%
  replace(is.na(.), 0) %>%
  # remove non-human primates and modern samples
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  # get only the carbohydrate KOs
  filter(str_detect(GeneFamily, carb_kos %>% str_c(collapse = "|"))) %>%
  # seelct only the levels with species
  filter(str_detect(GeneFamily, "\\|")) %>%
  # now break up the GeneFamily long name
  separate(GeneFamily, into = c("KO","other"), sep = ": ") %>%
  separate(other, into = c("Gene_name","Species"), sep = "\\|") %>%
  separate(Species, into = c("Genus","Species"), sep = "\\.s__") %>%
  mutate(Genus = str_replace_all(Genus, "g__","")) %>%
  # convert to long table to add clusters
  pivot_longer(!KO:Species, names_to = "Library_ID", values_to = "CPM") %>%
  left_join(., sample_cluster_info %>%
              select(Library_ID, Cluster), by = "Library_ID") %>%
  # figure out which KOs generally come from which species
  filter(CPM > 0) %>%
  group_by(Cluster, Species) %>%
  count() %>%
  mutate(Cluster = ifelse(Cluster == 1,"cluster1","cluster2")) %>%
  pivot_wider(names_from = "Cluster", values_from = "n") %>%
  mutate(Species = replace_na(Species, "Unknown")) %>%
  replace(is.na(.), 0) %>%
  arrange(Species)

  # test for significant differences between the clusters for each species?
ko_species <- carb_ko_count_table %>%
  pull(Species) 

ko_cl1 <- carb_ko_count_table %>%
  pull(cluster1) 

ko_cl2 <- carb_ko_count_table %>%
  pull(cluster2) 

intab <- as.table(cbind(
  ko_cl1, ko_cl2
))

dimnames(intab) <- list(
  Species = ko_species,
  Cluster = c("Cluster1","Cluster2")
)

intab %>%
  row_wise_prop_test(., alternative = "greater", p.adjust.method = "fdr") %>%
  arrange(p.adj) %>%
  rename(Species = group) %>%
  inner_join(., carb_ko_count_table, by = "Species") %>%
  mutate(Higher = ifelse(cluster1 > cluster2,"one","two"))


  
```


```{r, eval = F}
# set sample order for plotting
path_order <- keggtable %>%
  group_by(Cluster) %>%
  mutate(Percent = Counts / sum(Counts) *  100) %>%
  select(-Counts) %>%
  ungroup() %>%
  pivot_wider(names_from = "Cluster", values_from = "Percent") %>%
  mutate(Cluster2 = replace_na(Cluster2, 0),
         Cluster3 = replace_na(Cluster3, 0)) %>%
  mutate(greater = ifelse(Cluster1 > Cluster3, "Cluster1",
                          ifelse(Cluster3 >  Cluster1, "Cluster3", "Same"))) %>%
  mutate(path = Pathway) %>%
  separate(path, into = "map", extra = "drop", sep = " ") %>%
  mutate(map = str_replace_all(map, "map0","")) %>%
  left_join(., paths_details, by = "map") %>%
  mutate(map = str_replace_all(map, "^","map")) %>%
  group_by(greater, detail_group)  %>%
  count() %>%
  ungroup() %>%
  arrange(greater, desc(n)) %>%
  filter(greater ==  "Cluster1") %>%
  pull(detail_group)

# plot
keggtable %>%
  mutate(path = Pathway) %>%
  separate(path, into = "map", extra = "drop", sep = " ") %>%
  mutate(map = str_replace_all(map, "map0","")) %>%
  left_join(., paths_details, by = "map") %>%
  mutate(map = str_replace_all(map, "^","map")) %>%
  separate(pathway, into = "path", extra = "drop", sep = ":") %>%
  filter(detail_group != "Global and overview maps") %>%
  select(-Pathway) %>%
  select(path, everything()) %>%
  group_by(Cluster) %>%
  mutate(Percent = Counts / sum(Counts) *  100) %>%
  select(-Counts) %>%
  ungroup() %>%
  distinct() %>%
  # group_by(Cluster, detail_group) %>%
  # summarize(Totals = sum(Percent)) %>%
  pivot_wider(names_from = "Cluster", values_from = "Percent") %>%
  mutate(Cluster2 = replace_na(Cluster2, 0),
         Cluster3 = replace_na(Cluster3, 0)) %>%
  mutate(greater = ifelse(Cluster2 > Cluster3, "Cluster2",
                          ifelse(Cluster3 >  Cluster2, "Cluster3", "Same"))) %>%
  group_by(greater, detail_group)  %>%
  count() %>%
  ungroup() %>%
  mutate(detail_group = fct_relevel(detail_group, path_order)) %>%
  arrange(detail_group) %>%
  ggplot(., aes(x = detail_group, y = n, fill = greater)) +
         geom_bar(stat = "identity", position = position_dodge2(width = 0.9, preserve = "single"), color = "black") +
         theme_minimal(base_size = 11) +
         theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0.5),
               axis.text = element_text(size = 10)) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=8)) +
         ylab("number of pathways")


```


```{r, eval = F}
path_order <- keggtable %>%
  pivot_wider(names_from = "Cluster", values_from = "Counts") %>%
  mutate(path = Pathway) %>%
  separate(path, into = "map", extra = "drop", sep = " ") %>%
  mutate(map = str_replace_all(map, "map","")) %>%
  left_join(., paths_details, by = "map") %>%
  mutate(map = str_replace_all(map, "^","map")) %>%
  separate(pathway, into = "path", extra = "drop", sep = ":") %>%
  mutate(Cluster1 = replace_na(Cluster1, 0),
         Cluster3 = replace_na(Cluster3, 0)) %>%
  filter(detail_group != "Global and overview maps") %>%
  select(-Pathway, -map, -major_group) %>%
  select(path, detail_group, Cluster1, Cluster3) %>%
  pivot_longer(!path:detail_group, names_to = "Cluster", values_to = "Counts") %>%
  group_by(Cluster, detail_group) %>%
  summarize(total_counts = sum(Counts)) %>%
  mutate(percent = total_counts / sum(total_counts) * 100) %>%
  filter(Cluster == "Cluster3") %>%
  arrange(desc(percent)) %>%
  pull(detail_group)

fig <- keggtable %>%
  pivot_wider(names_from = "Cluster", values_from = "Counts") %>%
  mutate(path = Pathway) %>%
  separate(path, into = "map", extra = "drop", sep = " ") %>%
  mutate(map = str_replace_all(map, "map","")) %>%
  left_join(., paths_details, by = "map") %>%
  mutate(map = str_replace_all(map, "^","map")) %>%
  separate(pathway, into = "path", extra = "drop", sep = ":") %>%
  mutate(Cluster1 = replace_na(Cluster1, 0),
         Cluster3 = replace_na(Cluster3, 0)) %>%
  filter(detail_group != "Global and overview maps") %>%
  select(-Pathway, -map, -major_group) %>%
  select(path, detail_group, Cluster1, Cluster3) %>%
  pivot_longer(!path:detail_group, names_to = "Cluster", values_to = "Counts") %>%
  group_by(Cluster, detail_group) %>%
  summarize(total_counts = sum(Counts)) %>%
  mutate(percent = total_counts / sum(total_counts) * 100) %>%
  mutate(detail_group = fct_relevel(detail_group, path_order)) %>%
  ggplot(., aes(x = detail_group, y = percent, fill = Cluster)) +
         geom_bar(stat = "identity", position = position_dodge2(width = 0.9, preserve = "single"), color = "black") +
         theme_minimal(base_size = 11) +
         theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0.5),
               axis.text = element_text(size = 10)) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=8)) +
         scale_fill_manual(values = c("#f5c710ff","#098BD9"))
         ylab("% of KOs in each pathway")

fig
# ggsave("./06-publication/prelim_figs/KOs_diffs_clusters.pdf", plot = fig, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"), dpi = 300)


```

```{r}

exp_var <- paste0(round(humann3_gf_l1_noblanks_filt.pca$prop_expl_var$X * 100, 2), "%")
df_X <- humann3_gf_l1_noblanks_filt.pca$variates$X %>%
          as.data.frame() %>%
          rownames_to_column("Library_ID") %>%
          inner_join(full_metadata, by = "Library_ID") %>%
          full_join(., sample_cluster_info %>%
                      select(Library_ID, Cluster), by = "Library_ID")
    
pca_kegg <- df_X %>%
 mutate(Cluster = as.character(Cluster)) %>% 
  ggplot(., aes(PC1, PC2)) +
    # geom_point(aes(fill = Study, shape = Cluster), size = 4, stroke = 0.3) +
    ggstar::geom_star(aes(fill = Continent, starshape = Cluster), size = 1.5) + # , shape = 21
    scale_fill_manual(values = Continent_colors) +
    ggstar::scale_starshape_manual(values = c(15,9)) +
    xlab(paste("PC1 - ", exp_var[1])) +
    ylab(paste("PC2 - ", exp_var[2])) +
    theme_minimal(base_size = 14) +
    theme(legend.position = "right") +
    theme(legend.title = element_blank())
 pca_kegg

ggsave("./06-publication/supplemental_figures/Sup_fig_6/KOs_pca_clusters.pdf", plot = pca_kegg, device = "pdf",
       scale = 1, width = 8.5, height = 5, units = c("in"), dpi = 300)
   
```

```{r}

ko_counts <- h3_GF_counts_prep %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  left_join(., sample_cluster_info %>%
              select(Library_ID, Cluster), by  = "Library_ID") %>%
  mutate(Counts = ifelse(Counts > 0,1,0)) %>%
  group_by(Cluster, Library_ID) %>%
  summarize(total_GFs = sum(Counts)) %>%
  ungroup() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  left_join(., metadata %>%
              select(Library_ID, Study), by = "Library_ID") %>%
  mutate(Study = str_replace_all(Study, "Kilteasheen","Mann2018"),
         Study = str_replace_all(Study, "Middenbeemster","Velsko2022"),
         Study = str_replace_all(Study, "Iberia","Velsko2022"),
         Study = str_replace_all(Study, "Radcliffe","Velsko2019"),
         Study = replace_na(Study, "Pacific")) %>%
  drop_na(Cluster)



exp_var <- paste0(round(humann3_gf_l1_noblanks_filt.pca$prop_expl_var$X * 100, 2), "%")
df_X <- humann3_gf_l1_noblanks_filt.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(full_metadata, by = "Library_ID") %>%
  full_join(., sample_cluster_info %>%
            select(Library_ID, Cluster), by = "Library_ID") %>%
  left_join(., ko_counts %>%
              select(Library_ID, total_GFs), by = "Library_ID")
          
    
df_X %>%
  ggplot(., aes(PC1, PC2)) +
    geom_point(aes(fill = total_GFs), size = 3.5, stroke = 0.3, shape = 21) +
    scale_fill_viridis_c(option = "C") +
    xlab(paste("PC1 - ", exp_var[1])) +
    ylab(paste("PC2 - ", exp_var[2])) +
    theme_minimal(base_size = 14) +
    theme(legend.position = "right")


```



```{r}
raincloud_kegg <- h3_GF_counts_prep %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  left_join(., sample_cluster_info %>%
              select(Library_ID, Cluster), by  = "Library_ID") %>%
  mutate(Counts = ifelse(Counts > 0,1,0)) %>%
  group_by(Cluster, Library_ID) %>%
  summarize(total_GFs = sum(Counts)) %>%
  ungroup() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  left_join(., metadata %>%
              select(Library_ID, Study), by = "Library_ID") %>%
  mutate(Study = str_replace_all(Study, "Kilteasheen","Mann2018"),
         Study = str_replace_all(Study, "Middenbeemster","Velsko2022"),
         Study = str_replace_all(Study, "Iberia","Velsko2022"),
         Study = str_replace_all(Study, "Radcliffe","Velsko2019"),
         Study = replace_na(Study, "Pacific")) %>%
  drop_na(Cluster) %>%
  # mutate(Study = fct_relevel(Study, "Pacific","Eisenhoffer2020","FellowsYates2021","Mann2018","Mann2018_world","Ottoni2021","Velsko2019","Velsko2022")) %>%
  ggplot(., aes(x = Cluster, y = total_GFs, fill = Cluster)) + 
  ## add half-violin from {ggdist} package
  ggdist::stat_halfeye(
    ## custom bandwidth
    adjust = .5, 
    ## adjust height
    width = .6, 
    ## move geom to the right
    justification = -.2, 
    ## remove slab interval
    .width = 0, 
    point_colour = NA
  ) + 
  geom_boxplot(
    width = .12, 
    ## remove outliers
    outlier.color = NA, ## `outlier.shape = NA` works as well
    color = "grey30"
  ) +
  ## add dot plots from {ggdist} package
  ggdist::stat_dots(
    ## orientation to the left
    side = "left", 
    ## move geom to the left
    justification = 1.1, 
    ## adjust grouping (binning) of observations 
    binwidth = 15
  ) + 
  ## remove white space on the left
  coord_cartesian(xlim = c(1.2, NA)) +
  theme_minimal(base_size = 10) +
  scale_fill_manual(values = c("#7D3560","#098BD9")) +
  theme(axis.text.y = element_text(size = 10)) +
  theme(legend.position = "none") +
  xlab("Cluster") +
  ylab("Total Gene Families") +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 1)) 

raincloud_kegg

# ggsave("./06-publication/prelim_figs/KOs_raincloud_clusters.pdf", plot = raincloud_kegg, device = "pdf",
#        scale = 1, width = 8.5, height = 5, units = c("in"), dpi = 300)

h3_GF_counts_prep %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  left_join(., sample_cluster_info %>%
              select(Library_ID, Cluster), by  = "Library_ID") %>%
  mutate(Counts = ifelse(Counts > 0,1,0)) %>%
  group_by(Cluster, Library_ID) %>%
  summarize(total_GFs = sum(Counts)) %>%
  ungroup() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  drop_na(Cluster) %>%
  t_test(., total_GFs ~ Cluster) # kruskal_test

```

```{r}
sp_rain <- sp_counts %>%
  left_join(., sample_cluster_info %>%
              select(Library_ID, Cluster), by  = "Library_ID") %>%
  drop_na(Cluster) %>%
  mutate(Cluster = as.character(Cluster)) %>%
  ggplot(., aes(x = Cluster, y = no_species, fill = Cluster)) + 
  ## add half-violin from {ggdist} package
  ggdist::stat_halfeye(
    ## custom bandwidth
    adjust = .5, 
    ## adjust height
    width = .6, 
    ## move geom to the right
    justification = -.2, 
    ## remove slab interval
    .width = 0, 
    point_colour = NA
  ) + 
  geom_boxplot(
    width = .12, 
    ## remove outliers
    outlier.color = NA, ## `outlier.shape = NA` works as well
    color = "grey30"
  ) +
  ## add dot plots from {ggdist} package
  ggdist::stat_dots(
    ## orientation to the left
    side = "left", 
    ## move geom to the left
    justification = 1.1, 
    ## adjust grouping (binning) of observations 
    binwidth = 1.0
  ) + 
  ## remove white space on the left
  coord_cartesian(xlim = c(1.2, NA)) +
  theme_minimal(base_size = 10) +
  scale_fill_manual(values = c("#7D3560","#098BD9")) +
  theme(axis.text.y = element_text(size = 10)) +
  theme(legend.position = "none") +
  xlab("Cluster") +
  ylab("No. species") +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) 

sp_rain

# ggsave("./06-publication/prelim_figs/KOs_raincloud_clusters.pdf", plot = raincloud_kegg, device = "pdf",
#        scale = 1, width = 8.5, height = 5, units = c("in"), dpi = 300)

sp_counts %>%
  left_join(., sample_cluster_info %>%
              select(Library_ID, Cluster), by  = "Library_ID") %>%
  drop_na(Cluster) %>%
  mutate(Cluster = as.character(Cluster)) %>%
  t_test(., no_species ~ Cluster) # kruskal_test
  
```

```{r}
library(ggstar)

sp_v_KO <- h3_GF_counts_prep %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  left_join(., sample_cluster_info %>%
              select(Library_ID, Cluster), by  = "Library_ID") %>%
  mutate(Counts = ifelse(Counts > 0,1,0)) %>%
  group_by(Cluster, Library_ID) %>%
  summarize(total_GFs = sum(Counts)) %>%
  ungroup() %>%
  mutate(Cluster = as.character(Cluster)) %>%
  full_join(., sp_counts, by = "Library_ID") %>%
  drop_na(Cluster) %>%
  drop_na(no_species) %>%
  left_join(., full_metadata %>%
              select(Library_ID, Study, Continent)) %>%
  ggplot(., aes(x = total_GFs, y = no_species)) + # Species
         # geom_point(aes(fill = Continent, shape = Cluster), size = 2.5, stroke = 0.2) +
         geom_star(aes(fill = Continent, starshape = Cluster), size = 1.5) + # , shape = 21
         geom_smooth(method = glm, se = F, col = "black", linewidth = 0.7) +
         scale_fill_manual(values = Continent_colors) +
         scale_color_manual(values = Continent_colors) +
         scale_starshape_manual(values = c(15,9)) +
         # scale_shape_manual(values = c(21,24)) +
         theme_classic(base_size = 12) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         ggpubr::stat_regline_equation(label.y = 175, aes(label = ..rr.label..)) +
         ylab("Number of species") +
         xlab("Number of KEGG orthologs")

sp_v_KO

```

```{r}

# KO_sp <- ((pca_kegg / sp_v_KO + plot_layout(guides = 'collect')) | (raincloud_kegg + sp_rain + plot_layout(widths = c(1,1))))  +
#   plot_layout(widths = c(1,1.2))

KO_sp_dots <- pca_kegg / sp_v_KO + plot_layout(heights = c(1.5,1))
KO_sp_dots

# ggsave("./06-publication/supplemental_figures/Sup_fig_7/clusters_sp_KOs.pdf", plot = KO_sp_dots, device = "pdf",
#        scale = 1, width = 7, height = 6, units = c("in"), dpi = 300)

KO_sp_rain <- raincloud_kegg | sp_rain
KO_sp_rain

# ggsave("./06-publication/supplemental_figures/Sup_fig_7/clusters_sp_KOs_rain.pdf", plot = KO_sp_rain, device = "pdf",
#        scale = 1, width = 12, height = 6, units = c("in"), dpi = 300)



```


```{r get_GF_totals}
# how many KOs were identified per cluster?

h3_GF_counts <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  # and remove these 2 b/c they have no age data for the adonis test
  select(-ERR5729638, -ERR5729643) %>%
  # inner_join(., humann3_GF.decontam_noblanks_presence_more_30, by = "GeneFamily") %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  as_tibble() %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts > 0,1,0)) %>%
  group_by(Library_ID) %>%
  summarize(total_GFs = sum(Counts)) %>%
  ungroup()

```

```{r kegg_carb_list, eval = F}

kegg_carbs <- fread("./05-Documentation.backup/KEGG_KO_pathway_carbohydrates.tsv") %>%
  separate(KO_gene, into = c("KO","gene"), sep = "  ")

```

```{r kegg_carbs, eval = F}
# the cluster that's enriched in carbohydrate-utilization enzymes is not the expected one. Let's see what species those KOs come from
ko_carbs_samples <- humann3_GFs_top %>%
  # mutate(GeneFamily = str_trim(GeneFamily, side = "both")) %>%
  select(GeneFamily) %>%
  separate(GeneFamily, into = c("KO"), sep = ":", extra = "drop") %>%
  inner_join(., kegg_carbs %>%
               select(KO) %>% 
               unique(), by = "KO") %>%
  pull(KO)



humann3_KO_worldo %>%
  full_join(., humann3_GFs, by = "GeneFamily") %>%
  replace(is.na(.), 0) %>%
  # remove non-human primates and modern samples
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  # get only the carbohydrate KOs
  filter(str_detect(GeneFamily, ko_carbs_samples %>% str_c(collapse = "|"))) %>%
  # seelct only the levels with species
  filter(str_detect(GeneFamily, "\\|")) %>%
  # now break up the GeneFamily long name
  separate(GeneFamily, into = c("KO","other"), sep = ": ") %>%
  separate(other, into = c("Gene_name","Species"), sep = "\\|") %>%
  separate(Species, into = c("Genus","Species"), sep = "\\.s__") %>%
  mutate(Genus = str_replace_all(Genus, "g__","")) %>%
  # convert to long table to add clusters
  pivot_longer(!KO:Species, names_to = "Library_ID", values_to = "CPM") %>%
  left_join(., sample_cluster_info %>%
              select(Library_ID, Cluster), by = "Library_ID") %>%
  # figure out which KOs generally come from which species
  filter(CPM > 0) %>%
  group_by(Cluster, Species) %>%
  count() %>%
  mutate(Cluster = ifelse(Cluster == 1,"cluster1","cluster2")) %>%
  pivot_wider(names_from = "Cluster", values_from = "n") %>%
  mutate(Species = replace_na(Species, "Unknown"))%>%
  replace(is.na(.), 0) %>%
  arrange(Species)
  adorn_totals(where = "row") %>%
  as_tibble() %>%
  arrange(desc(cluster1))
  

```


```{r}
# how many species are contributing to the KO clusters in each sample?
filetable <- humann3_GFs_full %>%
  full_join(., humann3_KO_worldo_full, by = "GeneFamily") %>%
  select(-matches("HPD")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  filter(!str_detect(GeneFamily, "UNMAPPED|UNGROUPED")) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  # slice_head(n = 30) %>%
  separate(GeneFamily, into = c("KO","Taxon"), sep = "\\|", extra = "merge") %>%
  drop_na(Taxon) %>%
  separate(Taxon, into = c("Genus","Species"), sep = "\\.", extra = "merge") %>%
  pivot_longer(!KO:Species, names_to = "Library_ID", values_to = "CPM") %>%
  filter(CPM > 0) %>%
  select(Library_ID, Species) %>%
  distinct() %>%
  group_by(Library_ID) %>%
  count(name = "Species") %>%
  ungroup() %>%
  left_join(., full_metadata %>%
              select(Library_ID, Study, Continent)) 

s6 <- filetable %>%
  mutate(Continent = fct_relevel(Continent, "Pacific","Asia","Africa","Europe","Caribbean","North America")) %>%
   ggplot(., aes(x = Species, fill = Continent)) +
         geom_bar(color = "black", size = 0.25) +
         # geom_point(position = position_jitterdodge(jitter.width = 0.5), alpha = 0.5) +
         scale_fill_manual(values = Continent_colors) +
         theme_minimal() +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         scale_x_continuous(breaks=scales::pretty_breaks(n=10)) +
         ylab("No. samples") +
         xlab("No. species contributing named KOs") +
         ggtitle("") +
         facet_wrap(~Continent)

s6

# ggsave("./06-publication/supplemental_figures/Sup_fig_6/no_sp_cont_KOs.pdf", plot = s6, device = "pdf",
#        scale = 1, width = 12, height = 6, units = c("in"), dpi = 300)

humann3_GFs_full %>%
  full_join(., humann3_KO_worldo_full %>%
              select(-matches(nhp %>% str_c(collapse = "|"))), by = "GeneFamily") %>%
  select(-matches("HPD|EXB|LIB")) %>%
  # filter(!str_detect(GeneFamily, "UNMAPPED|UNGROUPED")) %>%
  # slice_head(n = 30) %>%
  separate(GeneFamily, into = c("KO","Taxon"), sep = "\\|", extra = "merge") %>%
  drop_na(Taxon) %>%
  separate(Taxon, into = c("Genus","Species"), sep = "\\.", extra = "merge") %>%
  pivot_longer(!KO:Species, names_to = "Library_ID", values_to = "CPM") %>%
  filter(CPM > 0) %>%
  select(KO, Library_ID, Species) %>%
  drop_na(Species) %>%
  distinct() %>%
  group_by(Library_ID, KO) %>%
  count() %>%
  pivot_wider(names_from = "Library_ID", values_from = "n") %>%
  replace(is.na(.), 0)

filetable %>%
  mutate(Continent = fct_relevel(Continent, "Pacific","Asia","Africa","Europe","Caribbean","North America")) %>%
  group_by(Continent) %>%
  count(Species, name = "No_samples") %>%
   ggplot(., aes(x = No_samples, y = Species, fill = Continent)) +
         geom_bar(stat = "identity", color = "black", stroke = 0.5) +
         # geom_point(position = position_jitterdodge(jitter.width = 0.5), alpha = 0.5) +
         scale_fill_manual(values = Continent_colors) +
         theme_minimal() +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         scale_x_continuous(breaks=scales::pretty_breaks(n=10)) +
         ylab("No. species contributing named KO") +
         xlab("No. samples") +
         ggtitle("") +
         facet_wrap(~Continent)

```

```{r}

filetable <- humann3_GFs_full %>%
  full_join(., humann3_KO_worldo_full, by = "GeneFamily") %>%
  select(-matches("HPD")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  filter(str_detect(GeneFamily, "UNGROUPED")) %>%
  # adorn_totals(where = "col") %>%
  # mutate(Percent = Total / sum(Total) * 100) %>%
  # filter(Percent >= 0.005) %>%
  # select(-Total, -Percent) %>%
  # slice_head(n = 30) %>%
  separate(GeneFamily, into = c("KO","Taxon"), sep = "\\|", extra = "merge") %>%
  drop_na(Taxon) %>%
  separate(Taxon, into = c("Genus","Species"), sep = "\\.", extra = "merge") %>%
  pivot_longer(!KO:Species, names_to = "Library_ID", values_to = "CPM") %>%
  filter(CPM > 0) %>%
  select(Library_ID, Species) %>%
  distinct() %>%
  group_by(Library_ID) %>%
  count(name = "Species") %>%
  ungroup() %>%
  left_join(., full_metadata %>%
              select(Library_ID, Study, Continent))

filetable %>%
  mutate(Continent = fct_relevel(Continent, "Pacific","Asia","Africa","Europe","Caribbean","North America")) %>%
   ggplot(., aes(x = Species, fill = Continent)) +
         geom_bar(color = "black", size = 0.25) +
         # geom_point(position = position_jitterdodge(jitter.width = 0.5), alpha = 0.5) +
         scale_fill_manual(values = Continent_colors) +
         theme_minimal() +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         theme(panel.grid.minor.y = element_blank()) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         scale_x_continuous(breaks=scales::pretty_breaks(n=10)) +
         ylab("No. samples") +
         xlab("No. species contributing ungrouped GFs") +
         ggtitle("") +
  facet_wrap(~Continent)

```

```{r unclassified_cpm}

humann3_GFs_full %>%
  full_join(., humann3_KO_worldo_full, by = "GeneFamily") %>%
  select(-matches("HPD")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  filter(!str_detect(GeneFamily, "UNMAPPED|UNGROUPED")) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  # slice_head(n = 30) %>%
  separate(GeneFamily, into = c("KO","Taxon"), sep = "\\|", extra = "merge") %>%
  drop_na(Taxon) %>%
  separate(Taxon, into = c("Genus","Species"), sep = "\\.", extra = "merge") %>%
  filter(Genus == "unclassified") %>%
  pivot_longer(!KO:Species, names_to = "Library_ID", values_to = "CPM") %>%
  filter(CPM > 0) %>%
  group_by(Library_ID) %>%
  summarize(sum_cpm = sum(CPM)) %>%
  ungroup() %>%
  left_join(., full_metadata %>%
              select(Library_ID, Study, Continent)) %>%
  ggplot(., aes(x = Continent, y = sum_cpm, fill = Continent)) +
         geom_boxplot(outlier.color = NA) +
         geom_point(position = position_jitterdodge(jitter.width = 0.5), alpha = 0.5) +
         scale_fill_manual(values = Continent_colors) +
         theme_minimal() +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         # scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         ylab("Unclassified CPM") +
         xlab("Continent") +
         ggtitle("")

```


```{r ungrouped_cpm}

humann3_GFs_full %>%
  full_join(., humann3_KO_worldo_full, by = "GeneFamily") %>%
  select(-matches("HPD")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  filter(str_detect(GeneFamily, "UNGROUPED")) %>%
  filter(!str_detect(GeneFamily, "\\|")) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "CPM") %>%
  left_join(., full_metadata %>%
              select(Library_ID, Study, Continent)) %>%
  ggplot(., aes(x = Continent, y = CPM, fill = Continent)) +
         geom_boxplot(outlier.color = NA) +
         geom_point(position = position_jitterdodge(jitter.width = 0.5), alpha = 0.5) +
         scale_fill_manual(values = Continent_colors) +
         theme_minimal() +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         # scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         ylab("UNGROUPPED CPM") +
         xlab("Continent") +
         ggtitle("")

```

```{r unmapped_cpm}

humann3_GFs_full %>%
  full_join(., humann3_KO_worldo_full, by = "GeneFamily") %>%
  select(-matches("HPD")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  filter(str_detect(GeneFamily, "UNMAPPED")) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "CPM") %>%
  left_join(., full_metadata %>%
              select(Library_ID, Study, Continent)) %>%
  ggplot(., aes(x = Continent, y = CPM, fill = Continent)) +
         geom_boxplot(outlier.color = NA) +
         geom_point(position = position_jitterdodge(jitter.width = 0.5), alpha = 0.5) +
         scale_fill_manual(values = Continent_colors) +
         theme_minimal() +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         # scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         ylab("UNMAPPED CPM") +
         xlab("Continent") +
         ggtitle("")

```

```{r total_IDd_cpm}

humann3_GFs_full %>%
  full_join(., humann3_KO_worldo_full, by = "GeneFamily") %>%
  select(-matches("HPD")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  filter(!str_detect(GeneFamily, "UNMAPPED|UNGROUPED")) %>%
  filter(!str_detect(GeneFamily, "\\|")) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "CPM") %>%
  filter(CPM > 0) %>%
  group_by(Library_ID) %>%
  summarize(sum_cpm = sum(CPM)) %>%
  ungroup() %>%
  left_join(., full_metadata %>%
              select(Library_ID, Study, Continent)) %>%
  ggplot(., aes(x = Continent, y = sum_cpm, fill = Continent)) +
         geom_boxplot(outlier.color = NA) +
         geom_point(position = position_jitterdodge(jitter.width = 0.5), alpha = 0.5) +
         scale_fill_manual(values = Continent_colors) +
         theme_minimal() +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         # scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         ylab("Total KO CPM") +
         xlab("Continent") +
         ggtitle("")

```


```{r number_genefamilies}

humann3_GFs_full %>%
  full_join(., humann3_KO_worldo_full, by = "GeneFamily") %>%
  select(-matches("HPD")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  filter(!str_detect(GeneFamily, "UNMAPPED|UNGROUPED")) %>%
  filter(!str_detect(GeneFamily, "\\|")) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "CPM") %>%
  filter(CPM > 0) %>%
  group_by(Library_ID) %>%
  count(name = "no_KOs") %>%
  ungroup() %>%
  left_join(., full_metadata %>%
              select(Library_ID, Study, Continent)) %>%
  ggplot(., aes(x = Continent, y = no_KOs, fill = Continent)) +
         geom_boxplot(outlier.color = NA) +
         geom_point(position = position_jitterdodge(jitter.width = 0.5), alpha = 0.5) +
         scale_fill_manual(values = Continent_colors) +
         theme_minimal() +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         # scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         ylab("No. KEGG Orthologs") +
         xlab("Continent") +
         ggtitle("")

```

```{r number_uniq_KO_per_continent}

humann3_GFs_full %>%
  full_join(., humann3_KO_worldo_full, by = "GeneFamily") %>%
  select(-matches("HPD")) %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  filter(!str_detect(GeneFamily, "UNMAPPED|UNGROUPED")) %>%
  filter(!str_detect(GeneFamily, "\\|")) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "CPM") %>%
  # filter(CPM > 0) %>%
  mutate(CPM = ifelse(CPM > 0, 1, 0)) %>%
  # pivot_wider(names_from = "Library_ID", values_from = "CPM", values_fill = 0)
  left_join(., full_metadata %>%
              select(Library_ID, Continent)) %>%
  distinct() %>%
  group_by(GeneFamily, Continent) %>%
  summarize(no_libraries = sum(CPM)) %>%
  ungroup() %>%
  replace(is.na(.), 0) %>%
  pivot_wider(names_from = "Continent", values_from = "no_libraries") %>%
  mutate(Continent_pres = ifelse(Pacific > 0 & Africa == 0 & Asia == 0 & Caribbean == 0 & Europe == 0 & `North America` == 0, "Pacific","Other")) %>%
  filter(Continent_pres == "Pacific") %>%
  arrange(desc(Pacific))

```




```{r, eval = F}
library(KEGGREST)
keggpull <- kegg_cluster_info %>% filter(Cluster == 3) %>% select(KEGG) %>% unique() %>% pull(KEGG)
query <- keggGet(keggpull)


```



```{r, eval = F}
library(phyloseq)
library(ANCOMBC)

pc_ends <- fread("./05-Documentation.backup/pc_end_samples.tsv") %>%
  mutate(dir = ifelse(PC1 <0, "Negative","Positive")) %>%
  rename(Sample = Species)

# convert the KO table to a phyloseq object for ANCOMBC
h3_KO_w_sp <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  select(GeneFamily, matches(pc_end_samples %>% str_c(collapse = "|"))) %>%
  # also remove samples that have very few (<100) gene families identified
  select(-matches("TAF017.E0101|DLV002.A0101|TAF016.B0101|TAF017.A0101|HCLVMBCX2-3505-21-00-01_S21")) %>%
  # now arrange the samples alphabetically
  pivot_longer(!GeneFamily, names_to = "libraryID", values_to = "logs") %>%
  mutate(libraryID = str_replace_all(libraryID, ".SG1","")) %>%
  arrange(libraryID) %>%
  pivot_wider(names_from = "libraryID", values_from = "logs") %>%
  arrange(GeneFamily) %>%
  column_to_rownames("GeneFamily") %>%
  as.matrix(.)

h3_KO_w_sp_tax <- h3_KO_w_sp %>%
  as.data.frame(.) %>%
  rownames_to_column("GeneFamily") %>%
  arrange(GeneFamily) %>%
  select(GeneFamily) %>%
  mutate(Species = GeneFamily) %>%
  column_to_rownames("GeneFamily") %>%
  as.matrix(.)

metadat <- sample_data(metadata %>%
  inner_join(., pc_ends %>%
               select(Sample, dir) %>%
               rename(Library_ID = Sample), by = "Library_ID") %>%
  distinct() %>%
  column_to_rownames("Library_ID"))
  
 # %>%
 #  mutate(Full = as.character(Full)

OTU = otu_table(h3_KO_w_sp, taxa_are_rows = TRUE)
TAX = tax_table(h3_KO_w_sp_tax)
physeq = phyloseq(OTU, TAX, metadat)

physeq1 <- merge_phyloseq(physeq, metadat)

# check the phyloseq object
physeq

pacific_diff_pc_ends <- ancombc(physeq, formula = "dir", p_adj_method = "fdr", group = "dir", struc_zero = TRUE)

```

```{r, eval = F}

lfc <- pacific_diff_pc_ends$res$beta 

qval <- pacific_diff_pc_ends$res$q_val

diffabund <- pacific_diff_pc_ends$res$diff_abn

alltogether <- pacific_diff_pc_ends$res$beta %>%
  rownames_to_column("Pathway") %>%
  rename(lfc_pos = 2) %>%
  full_join(., pacific_diff_pc_ends$res$q_val %>%
            rownames_to_column("Pathway") %>%
            rename(qval_pos = 2)) %>%
  full_join(., pacific_diff_pc_ends$res$diff_abn %>%
            rownames_to_column("Pathway") %>% 
            rename(da_pos = 2))


alltogether %>%
  filter(da_pos == TRUE,
         qval < 0.05,
         lfc_pos >=2.5 | lfc_pos <= -2.5) %>%
  arrange(desc(lfc_pos)) 
# %>%
#   fwrite(., "~/archgen/microbiome_calculus/pacific_calculus/05-Documentation.backup/pathway_abundance_ancombc.tsv", quote = F, sep = "\t")

```

```{r, eval = F}
# 
alltogether %>%
  mutate(colorpt = ifelse(qval_pos <= 0.05 & lfc_pos >=2.5, "turquoise",
                          ifelse(qval_pos <= 0.05 & lfc_pos <= -2.5, "turquoise", "coral"))) %>%
  mutate(qval_pos = ifelse(qval_pos == 0, 7.742690e-295	, qval_pos)) %>%
  mutate(beta = -1 * log10(qval_pos)) %>%
  ggplot(., aes(x = lfc_pos, y = beta, color = colorpt)) +
     geom_point() +
     scale_fill_viridis_c(option = "C") +
     theme_minimal(base_size = 12) +
     theme(text = element_text(size=12)) +
     theme(legend.position = "right") +
     theme(plot.title = element_text(size = 10)) +
     xlab("Log-fold change") +
     ylab("-Log10(q-val)")

```


## Non-corrected data again
Now let's look at which pathways are driving separation of the sample groups.
Which of the pathways have the strongest loadings in PC1 and PC2?
```{r pca_gf_biplot}
# which paths distinguish sample types when HMP and JAE are included?
h3_gf_cor_biplot_pc1 <- plot_pca_bi(humann3_gf_l1_noblanks_filt.pca, "PC1", "PC2", "Study", PC1, GeneFamily = T)
h3_gf_cor_biplot_pc1

# ggsave("./05-results.backup/presentation_pdfs/pca_EG_bi_noyanomami_GFs.pdf", plot = h3_gf_cor_biplot_pc1, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"), dpi = 300)

# what are those pathways?
pandoc.table(h3_gf_cor_biplot_pc1$plot_env$pws_10 %>% select(-Function) %>% arrange(desc(PC1)),
             digits = 5,
             keep.trailing.zeros = T,
             justify = "lccr",
             split.tables = Inf,
             caption = "PC1 positive loadings")

pandoc.table(h3_gf_cor_biplot_pc1$plot_env$neg_10 %>% select(-Function) %>% arrange(PC1),
             digits = 5,
             keep.trailing.zeros = T,
             justify = "lccr",
             split.tables = Inf,
             caption = "PC1 negative loadings")

# make a table of the pathways to save, to use again later in another R notebook
humann3_GFway_biplot_list <- h3_gf_cor_biplot_pc1$plot_env$pws_10 %>% select(-Function) %>% arrange(desc(PC1)) %>% select(GeneFamily) %>% mutate(Direction = "PC1+")
humann3_GFway_biplot_list <- humann3_GFway_biplot_list %>%
  bind_rows(h3_gf_cor_biplot_pc1$plot_env$neg_10 %>% select(-Function) %>% arrange(desc(PC1)) %>% select(GeneFamily)%>% mutate(Direction = "PC1-"))

# which paths distinguish sample types when HMP and JAE are included?
h3_gf_cor_biplot_pc2 <- plot_pca_bi(humann3_gf_l1_noblanks_filt.pca, "PC1", "PC2", "Study", PC2, GeneFamily = T)
h3_gf_cor_biplot_pc2

# what are those pathways?
pandoc.table(h3_gf_cor_biplot_pc2$plot_env$pws_10 %>% select(-Function) %>% arrange(desc(PC2)),
             digits = 5,
             keep.trailing.zeros = T,
             justify = "lccr",
             split.tables = Inf,
             caption = "PC2 positive loadings")

pandoc.table(h3_gf_cor_biplot_pc2$plot_env$neg_10 %>% select(-Function) %>% arrange(PC2),
             digits = 5,
             keep.trailing.zeros = T,
             justify = "lccr",
             split.tables = Inf,
             caption = "PC2 negative loadings")

humann3_GFway_biplot_list <- humann3_GFway_biplot_list %>%
  bind_rows(h3_gf_cor_biplot_pc2$plot_env$pws_10 %>% select(-Function) %>% arrange(desc(PC2)) %>% select(GeneFamily) %>% mutate(Direction = "PC2+"))
humann3_GFway_biplot_list <- humann3_GFway_biplot_list %>%
  bind_rows(h3_gf_cor_biplot_pc2$plot_env$neg_10 %>% select(-Function) %>% arrange(desc(PC2)) %>% select(GeneFamily) %>% mutate(Direction = "PC2-"))


fwrite(humann3_GFway_biplot_list, file = "./04-Analysis/humann3/humann3_GFs_biplot_list.tsv", quote = F, sep = "\t")

```

## Other factors
```{r GF_totals}
# Total # of gene families identified
plot_pca_cont(humann3_gf_l1_noblanks_filt.pca, "PC1", "PC2", "total_GFs", 3, "No. Gene Families")

```


```{r mean_age}
# mean age (BP)
plot_pca_cont(humann3_gf_l1_noblanks_filt.pca, "PC1", "PC2", "Age_mean", 3, "Mean age (BP)")

```

```{r gc_avg}
# mean age (BP)
plot_pca_cont(humann3_gf_l1_noblanks_filt.pca, "PC1", "PC2", "gc_avg", 3, "GC (%)")

```

```{r no_reads}
# mean age (BP)
plot_pca_cont(humann3_gf_l1_noblanks_filt.pca, "PC1", "PC2", "No_reads_no_human", 3, "Total reads")

```

```{r adonis_gf}


# species present at <0.005% per lab
humann3_gf_l1_noblanks_filt <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(under750 %>% str_c(collapse = "|"))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(GeneFamily,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_gf_l1_noblanks_filt_pca <- tune.pca(humann3_gf_l1_noblanks_filt, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_gf_l1_noblanks_filt.pca <- pca(humann3_gf_l1_noblanks_filt, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
humann3_gf_l1_noblanks_filt.euc <- vegdist(humann3_gf_l1_noblanks_filt.pca$X, method = "euclidean", na.rm = F)
humann3_gf_l1_noblanks_filt.euc.pcoa <- ape::pcoa(humann3_gf_l1_noblanks_filt.euc)

humann3_gf_l1_noblanks_filt.euc.pcoavectors <- as.data.frame(humann3_gf_l1_noblanks_filt.euc.pcoa$vectors)
humann3_gf_l1_noblanks_filt.euc.pcoavectors <- humann3_gf_l1_noblanks_filt.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

humann3_gf_l1_noblanks_filt.euc.vecmet <- left_join(humann3_gf_l1_noblanks_filt.euc.pcoavectors, full_metadata, by = "Library_ID")

humann3_gf_l1_noblanks_filt.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Study, shape = Lab)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Study_colors) +
  scale_shape_manual(values = Lab_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Site_code")

# humann3_gf_l1_noblanks_filt %>%
#   rownames_to_column("Library_ID") %>%
#   select(Library_ID) %>%
#   anti_join(., metadata %>% select(Library_ID, Anginosus), by  = "Library_ID")


# test for difference between different metadata categories 
# all together
adonis2(humann3_gf_l1_noblanks_filt.euc ~ Lab + Study + Age_mean + gc_avg + seq_len_avg + No_reads_no_human, humann3_gf_l1_noblanks_filt.euc.vecmet, permutations = 999, by = "margin")

# controlled for Lab
adonis2(humann3_gf_l1_noblanks_filt.euc ~ Study + Age_mean + gc_avg + seq_len_avg + No_reads_no_human, humann3_gf_l1_noblanks_filt.euc.vecmet, permutations = 999, by = "margin", strata = humann3_gf_l1_noblanks_filt.euc.vecmet$Lab)

```


## Canonical Correlation Analysis

```{r pca_species_all}
poor_df <- poor_samples %>%
  as_tibble(.) %>%
  rename(Library_ID = 1) %>%
  mutate(Preservation = "Poor")

exp_var_all <- paste0(round(humann3_gf_l1_noblanks_filt.pca$prop_expl_var$X * 100, 2), "%")
all_pca_values <- humann3_gf_l1_noblanks_filt.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  # and remove these 2 b/c they have no age data for the adonis test
  filter(!str_detect(Library_ID, "ERR5729638|ERR5729643")) %>%
  full_join(full_metadata, by = "Library_ID") %>%
  full_join(., poor_df, by = "Library_ID") %>%
  mutate(Preservation = ifelse(is.na(Preservation), "Good", Preservation)) %>%
  mutate(Preservation = ifelse(Type == "blank","Blank",
                               ifelse(Type == "Arch. Bone", "Arch. Bone", Preservation))) %>%
  drop_na(PC1) %>%
  left_join(., full_metadata %>%
    select(Study) %>%
    unique() %>%
    mutate(Study_n = 1:n()), by = "Study") %>%
  mutate(Lab_n = ifelse(Lab == "Jena",0,
                        ifelse(Lab == "Oklahoma",1,
                               ifelse(Lab == "Otago",2,
                                      ifelse(Lab == "Oxford",3,
                                             ifelse(Lab == "Tuebingen",4,
                                                    ifelse(Lab == "Vienna",5,6))))))) 

```

```{r all_corrs}
# rename variables to look nice in figures
all_pca_values_select <- all_pca_values %>%
  select(Library_ID,PC1, PC2, PC3, seq_len_avg, gc_avg, No_reads_no_human,  Age_mean, Lab_n, Study_n) %>%
  rename(`Seq. length` = seq_len_avg,
         GC = gc_avg,
         `Total reads` = No_reads_no_human,
         Lab = Lab_n,
         Study = Study_n,
         `Mean age (BP)` = Age_mean)

form <- ~ PC1 + PC2 + PC3 + `Seq. length` + GC + `Total reads` + `Mean age (BP)` + Lab + Study
C_all = canCorPairs(form, all_pca_values_select)
plotCorrMatrix( C_all )

```


```{r pearson_corr_mid}
# test for significant differences in diversity between the metadata categories

pearson_all <- all_pca_values %>%
  # remove columns to avoid duplicates in renaming
  select(-Study, -Lab)  %>%
  # rename to match the chunk above
  rename(`Seq. length` = seq_len_avg,
         GC = gc_avg,
         `Total reads` = No_reads_no_human,
         Lab = Lab_n,
         Study = Study_n,
         `Mean age (BP)` = Age_mean) %>%
  rstatix::cor_test(., PC1, PC2, PC3, "Seq. length", GC, "Total reads", "Mean age (BP)", Lab, Study, method="pearson")

pearson_all

pearson_all %>%
  mutate(same = var1 == var2) %>%
  filter(same == FALSE) %>%
  select(-same) %>%
  filter(p <= 0.001) %>%
  filter(cor > 0.40 | cor < -0.40) %>%
  distinct(statistic, .keep_all = TRUE) %>%
  arrange(p)
  
```

```{r}

C_all %>%
  as_data_frame(.) %>%
  mutate(names = rownames(C_all)) %>%
  select(names, everything()) %>%
  pivot_longer(!names, names_to = "Corrs", values_to = "values") %>%
  filter(values != 1) %>%
  filter(values >= 0.4) %>%
  distinct(values, .keep_all = TRUE)

```


```{r}
# colorset <- rev(colorspace::sequential_hcl(10, palette = "BuPu"))
# maybe get colors from here: https://colordesigner.io/

colorset <- c("#ffffff","#e4f4fe","#c8e9fd","#addefc","#92d3fa","#77c8f9","#5bbdf8","#40b2f7","#25a8f6","#0a9cf4","#098BD9") # ,"#087dc3","#076fae","#066198","#055382"

 pvals_all <- all_pca_values_select %>%
   drop_na(Library_ID) %>%
  column_to_rownames("Library_ID") %>%
  as.matrix(.) %>%
  corrplot::cor.mtest(., method="pearson")

C_all %>%
  corrplot::corrplot(., p.mat = pvals_all$p, order = "original",  type = "upper", col = colorset, is.corr = FALSE, cl.cex = 1.0, diag = FALSE, tl.cex = 1.0,
                     tl.col = 'black', sig.level = c(0.001), pch.cex = 2.2, insig = 'label_sig', pch.col = 'grey20')

#  order = "hclust", hclust.method = "complete"

```

```{r, eval = F}
svg(file = "./06-publication/prelim_figs/genefamilies_corr_plot.svg") 

C_all %>%
  corrplot::corrplot(., p.mat = pvals_all$p, order = "hclust", type = "upper", col = colorset, is.corr = FALSE, cl.cex = 1.2, diag = FALSE, tl.cex = 1.4,
                     tl.col = 'black', sig.level = c(0.001), pch.cex = 2.2, insig = 'label_sig', pch.col = 'grey20')

dev.off()

```



## Not used
```{r all_pca, eval = F}
 # all samples and blanks
humann3_gf_l1_all <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  select(-TAF017.E0101,-DLV002.A0101,-TAF016.B0101,-TAF017.A0101,-`HCLVMBCX2-3505-21-00-01_S21`) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(GeneFamily,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_gf_pca <- tune.pca(humann3_gf_l1_all, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_gf_l1_all.pca <- pca(humann3_gf_l1_all, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_gf_l1_all.pca, "PC1", "PC2", "Study", "Type")

# humann3_gf_l1_all.pca$variates$X %>%
#   as_tibble()

```

```{r all_pca_noblanks, eval = F}
 # all samples no blanks
humann3_gf_l1_all_noblanks <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-TAF017.E0101,-DLV002.A0101,-TAF016.B0101,-TAF017.A0101,-`HCLVMBCX2-3505-21-00-01_S21`) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(GeneFamily,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_gf_pca <- tune.pca(humann3_gf_l1_all_noblanks, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_gf_l1_all_noblanks.pca <- pca(humann3_gf_l1_all_noblanks, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_gf_l1_all_noblanks.pca, "PC1", "PC2", "Study", "Type")

# humann3_gf_l1_all_noblanks.pca$variates$X %>%
#   as_tibble()

```

Remove the poorly-preserved samples based on cuperdec. 
These were removed from the taxonomic analyses.
```{r cuperdec_poor_rem_pca, eval = F}
# make a list of blanks b/c they don't all have the same prefixes
blanks <- full_metadata %>%
  mutate(Type = ifelse(is.na(Type),Sample_type,Type)) %>%
  filter(str_detect(Type, "blank|bone")) %>%
  filter(!str_detect(Library_ID, "ARS")) %>%
  select(Library_ID) %>%
  pull()

# create the input matrix
humann3_gf_l1_cpdrm <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-TAF017.E0101,-DLV002.A0101,-TAF016.B0101,-TAF017.A0101,-`HCLVMBCX2-3505-21-00-01_S21`) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(GeneFamily,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_gf_pca <- tune.pca(humann3_gf_l1_cpdrm, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_gf_l1_cpdrm.pca <- pca(humann3_gf_l1_cpdrm, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_gf_l1_cpdrm.pca, "PC1", "PC2", "Study", "Type")

```

Clearly estimating preservation based on functional profile gives a different
result than based on taxonomic profile. There are several samples with
apparently blank/bone-like functional profiles, but good taxonomic profiles.
Let's see what a PCA looks like with all samples, colored/shaped by
cuperdec-determined preservation.

```{r, eval = F}
#  list taxonomicly well-preserved samples that plot with blanks to remove from downstream analysis
other_poor <- humann3_gf_l1_cpdrm.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  filter(PC1 > 0) %>%
  select(Library_ID, Type) %>%
  filter(Type == "Arch. calculus") %>%
  pull(Library_ID)

```


```{r poor_marked, eval = F}
# mark the taxonomically poorly-preserved samples to see where they fall in pathway preservation
poor_df <- poor_samples %>%
  as_tibble(.) %>%
  rename(Library_ID = 1) %>%
  mutate(Preservation = "Poor")

exp_var <- paste0(round(humann3_gf_l1_all.pca$prop_expl_var$X * 100, 2), "%")
df_X_all <- humann3_gf_l1_all.pca$variates$X %>%
           as.data.frame() %>%
           rownames_to_column("Library_ID") %>%
           left_join(full_metadata, by = "Library_ID") %>%
           full_join(., poor_df, by = "Library_ID") %>%
           mutate(Preservation = ifelse(is.na(Preservation), "Good", Preservation)) %>%
           mutate(Preservation = ifelse(Type == "blank","Blank",
                                        ifelse(Type == "Arch. Bone", "Arch. Bone", Preservation))) %>%
           # there's one sample with NA and no metadata, so call it poorly-preserved
           mutate(Preservation = ifelse(is.na(Preservation), "Poor", Preservation))

# test plot
df_X_all %>%
  ggplot(., aes(PC1, PC2, color = Preservation, shape = Preservation)) +
                  geom_point(size = 4) +
                  xlab(paste("PC1 - ", exp_var[1])) +
                  ylab(paste("PC2 - ", exp_var[2])) +
                  # xlim(5, 10) +
                  # ylim(-1,1) +
                  theme_minimal(base_size = 14) +
                  theme(legend.position = "right") +
                  theme(legend.title = element_blank())



```

```{r gf_poor, eval = F}
# Based on the PCA above, try all samples with PC1 loadings < -5 as "poorly preserved"
gf_poor <- humann3_gf_l1_all.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(full_metadata, by = "Library_ID") %>%
  filter(PC1 < 0) %>%
  select(Library_ID, Type) %>%
  filter(Type == "Arch. calculus") %>%
  arrange(Library_ID) %>%
  pull(Library_ID)

humann3_gf_l1_all.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(full_metadata, by = "Library_ID") %>%
  filter(PC1 < 0) %>%
  select(Library_ID, Type) %>%
  filter(Type == "Arch. calculus") %>%
  select(Library_ID) %>%
  arrange(Library_ID) %>%
  fwrite(., "../strep_clades/gf_poor_samples.tsv", quote = F, sep = "\t")
  
```

Remove the gf_poor samples and see what the PCA looks like
```{r poor_removed_marked, eval = F}
df_X_all %>%
  filter(!str_detect(Library_ID, gf_poor %>%
                       str_c(collapse = "|"))) %>%
  ggplot(., aes(PC1, PC2, color = Preservation, shape = Preservation)) +
                  geom_point(size = 4) +
                  xlab(paste("PC1 - ", exp_var[1])) +
                  ylab(paste("PC2 - ", exp_var[2])) +
                  theme_minimal(base_size = 14) +
                  theme(legend.position = "right") +
                  theme(legend.title = element_blank())



```

What does the PCA look like if we filter low-prevalence gene families and remove
poorly-preserved samples?

```{r filter_all_pca, eval = F}
# try filtering the table
# list all pathways that are present in at least 1/3 of samples, as a way to reduce the input
humann3_GF.decontam_noblanks_presence_more_30 <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples and blanks
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-TAF017.E0101,-DLV002.A0101,-TAF016.B0101,-TAF017.A0101,-`HCLVMBCX2-3505-21-00-01_S21`) %>%
  mutate_if(is.numeric, ~1 * (. > 0)) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  group_by(GeneFamily) %>%
  summarize(Percent_presence = sum(Counts)/length(Library_ID)*100) %>%
  # select the pathways that are present in at least 1/3 of at least 2 groups
  ungroup() %>%
  # pathways present in > 1/3 of samples 
  filter(Percent_presence >= 33.33) %>%
  # pull out these pathways to keep
  select(GeneFamily)

# create the input matrix
humann3_gf_l1_filt <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-TAF017.E0101,-DLV002.A0101,-TAF016.B0101,-TAF017.A0101,-`HCLVMBCX2-3505-21-00-01_S21`) %>%
  inner_join(., humann3_GF.decontam_noblanks_presence_more_30, by = "GeneFamily") %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(GeneFamily,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_gf_filt_pca <- tune.pca(humann3_gf_l1_filt, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_gf_l1_filt.pca <- pca(humann3_gf_l1_filt, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_gf_l1_filt.pca, "PC1", "PC2", "Study", "Type")

exp_var <- paste0(round(humann3_gf_l1_filt.pca$prop_expl_var$X * 100, 2), "%")
df_X_filt <- humann3_gf_l1_filt.pca$variates$X %>%
           as.data.frame() %>%
           rownames_to_column("Library_ID") %>%
           left_join(full_metadata, by = "Library_ID") %>%
           full_join(., poor_df, by = "Library_ID") %>%
           mutate(Preservation = ifelse(is.na(Preservation), "Good", Preservation)) %>%
           mutate(Preservation = ifelse(Type == "blank","Blank",
                                        ifelse(Type == "Arch. Bone", "Arch. Bone", Preservation)))

# test plot
df_X_filt %>%
  ggplot(., aes(PC1, PC2, color = Preservation, shape = Preservation)) +
                  geom_point(size = 4) +
                  xlab(paste("PC1 - ", exp_var[1])) +
                  ylab(paste("PC2 - ", exp_var[2])) +
                  # xlim(5, 10) +
                  # ylim(-1,1) +
                  theme_minimal(base_size = 14) +
                  theme(legend.position = "right") +
                  theme(legend.title = element_blank())


```


Plot a PCA with only the well-preserved samples
```{r pca_GFs_noblanks, eval = F}

humann3_gf_l1_noblanks <- humann3_KO_world %>%
  select(-matches(nhp %>% str_c(collapse = "|"))) %>%
  # remove poorly preserved samples
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  select(-matches(blanks %>% str_c(collapse = "|"))) %>%
  select(-TAF017.E0101,-DLV002.A0101,-TAF016.B0101,-TAF017.A0101,-`HCLVMBCX2-3505-21-00-01_S21`) %>%
  # select(-matches(poor_samples %>% str_c(collapse = "|")) %>%
  # select(-all_of(other_poor)) %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(GeneFamily,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.humann3_gf_pca <- tune.pca(humann3_gf_l1_noblanks, logratio = 'CLR')

# perform a PCA to see how the data cluster
humann3_gf_l1_noblanks.pca <- pca(humann3_gf_l1_noblanks, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(humann3_gf_l1_noblanks.pca, "PC1", "PC2", "Study", "Lab")
plot_pca(humann3_gf_l1_noblanks.pca, "PC1", "PC2", "Lab", "Lab")

```



## Batch correction with ComBat
```{r, eval = F}
source("./02-Scripts.backup/ComBat-seq/ComBat_seq.R")
source("./02-Scripts.backup/ComBat-seq/helper_seq.R")

```

Use the filtered table for batch-correction
```{r, eval = F}
# transpose dataframe for ComBat
# this table has the poorly-preserved samples removed
humann3_gf_l1_noblanks_filt_t <- humann3_gf_l1_noblanks_filt %>%
  rownames_to_column("Library_ID") %>%
  pivot_longer(!Library_ID, names_to = "GeneFamily", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  column_to_rownames("GeneFamily")

h3_gf_mat <- as.matrix(humann3_gf_l1_noblanks_filt_t)

lab_batch <- humann3_gf_l1_noblanks_filt %>%
  rownames_to_column("Library_ID") %>%
  left_join(., full_metadata) %>%
  select(Library_ID, Lab, everything()) %>%
  mutate(Lab = ifelse(is.na(Lab),"Otago",Lab)) %>%
  select(Lab) %>%
  pull()

```

Or use the non-filtered tale
```{r eval = F}
# transpose dataframe for ComBat
# this table has the poorly-preserved samples removed
humann3_gf_l1_noblanks_t <- humann3_gf_l1_noblanks %>%
  rownames_to_column("Library_ID") %>%
  pivot_longer(!Library_ID, names_to = "GeneFamily", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  column_to_rownames("GeneFamily")

h3_gf_mat <- as.matrix(humann3_gf_l1_noblanks_t)

lab_batch <- humann3_gf_l1_noblanks %>%
  rownames_to_column("Library_ID") %>%
  left_join(., full_metadata) %>%
  select(Library_ID, Lab, everything()) %>%
  mutate(Lab = ifelse(is.na(Lab),"Otago",Lab)) %>%
  select(Lab) %>%
  pull()

```

```{r, eval = F}
h3_gf_mat_adjusted <- ComBat_seq(h3_gf_mat, batch=lab_batch)

```

```{r, eval = F}
h3_gf_mat_cor <- h3_gf_mat_adjusted %>%
  as.data.frame(.) %>%
  rownames_to_column("GeneFamily")

# fwrite(h3_gf_mat_cor, "./04-Analysis/humann3/humann3_GFs_batch_corrected.tsv")

```


```{r, eval = F}
# now check the PCA after batch correction
h3_gf_cor <- h3_gf_mat_cor %>%
  pivot_longer(!GeneFamily, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(GeneFamily,Counts) %>%
  column_to_rownames("Library_ID")


# check the number of components to retain by tuning the PCA
tune.pca(h3_gf_cor, logratio = 'CLR')

# perform a PCA to see how the data cluster
h3_gf_cor.pca <- pca(h3_gf_cor, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(h3_gf_cor.pca, "PC1", "PC2", "Study", "Lab")
plot_pca(h3_gf_cor.pca, "PC1", "PC2", "Lab", "Lab")


```








