---
title: "Pacific calculus humann3 pahway analysis"
author: "Irina Velsko"
date: "`r format(Sys.time(), '%b %d, %Y')`"
output:
    github_document:
        fig_width: 8
        fig_height: 6
        toc: true
---


```{r load_libraries, echo = F, message = F}
library(knitr)
library(data.table)
# library(edgeR) # needed to run ComBat_seq
library(mixOmics)
library(janitor)
library(rstatix)
library(vegan)
library(pander)
library(variancePartition)
library(tidyverse)
library(ggstar)
library(ggrepel)
library(viridis)
library(patchwork)
library(RColorBrewer)
library(microshades)
opts_chunk$set(echo=F, warning=F, message=F, dpi=150)
```

Set the working directory of the notebook to the parent directory of the 
notebook itself.  

```{r setup}
knitr::opts_knit$set(root.dir = normalizePath(".."))
```


# contaminant tables 
```{r load_contam_lists}
# genus lists
cont_otago_gn <- fread("./05-Documentation.backup/contaminant_otago_gn.tsv")
cont_oklahoma_gn <- fread("./05-Documentation.backup/contaminant_oklahoma_gn.tsv")
cont_jena_gn <- fread("./05-Documentation.backup/contaminant_jena_gn.tsv")

cont_all_gn <- fread("./05-Documentation.backup/contaminant_pacific_ooj_gn.tsv")

# make sure the same contaminants are in the individual and combined tables (expect 874 - yes, same total)
cont_otago_gn %>%
  bind_rows(., cont_oklahoma_gn) %>%
  bind_rows(., cont_jena_gn) %>%
  distinct() %>%
  count()


# species lists
cont_otago_sp <- fread("./05-Documentation.backup/contaminant_otago_sp.tsv", sep = "\t")
cont_oklahoma_sp <- fread("./05-Documentation.backup/contaminant_oklahoma_sp.tsv", sep = "\t")
cont_jena_sp <- fread("./05-Documentation.backup/contaminant_jena_sp.tsv", sep = "\t")

cont_all_sp <- fread("./05-Documentation.backup/contaminant_pacific_ooj_sp.tsv", sep = "\t")

```

Load metadata
```{r load_metadata, eval = F}
#only need to run this once, then read in the file with the DNA concentration added
metadata <- fread("./05-Documentation.backup/oceana_metadata.tsv") %>%
  rename(Library_ID = Library_ID) %>%
  mutate(Library_ID = str_replace_all(Library_ID, ".SG1.1","")) %>%
  full_join(., fread("./05-Documentation.backup/oceana_meta_decontam.tsv") %>%
              select(-Lab) %>%
              rename(Library_ID = SampleID), by = "Library_ID") %>%
  mutate(Sample_or_Control = ifelse(SourceSink == "sink", "sample","control")) %>%
  mutate(Sample_or_Control = ifelse(Env == "blank", "control",Sample_or_Control)) %>%
  mutate(Sample_or_Control = ifelse(is.na(Sample_or_Control) & Type == "blank", "control",Sample_or_Control)) %>%
  mutate(Sample_or_Control = ifelse(is.na(Sample_or_Control) & Type == "bone", "control",Sample_or_Control)) %>%
  mutate(Sample_or_Control = ifelse(is.na(Sample_or_Control) & Type == "calculus","sample",Sample_or_Control)) %>%
  mutate(Type = ifelse(str_detect(Library_ID, "ARS"),"bone",Type))

# can't write into the Pacific folder, so write it here and move it to 05-Documentation.backup
# fwrite(metadata, "../strep_clades/00-documentation/oceana_metadata_merged.tsv", quote = F, sep = "\t", na = "NA")

```

```{r}
metadata <- fread("05-Documentation.backup/oceania_metadata_full.tsv") %>%
  # make the island names match the colors/shapes
  mutate(Island = str_replace_all(Island, "_"," "),
         Island = str_replace_all(Island, " Duff Islands",""),
         Island = str_replace_all(Island, "3000","3000 BP")) %>%
  mutate_all(na_if,"")

```


```{r poorly_preserved}
cfdp_fail <- fread("05-Documentation.backup/oceania_cuperdec_species_poor_samples.tsv") %>%
  rename(Library_ID = 1)

```

```{r}
# list all bones that are part of this study
# the ARS bones are a little different in the PCA, so we won't use them
bones <- metadata %>%
  filter(Type == "bone") %>%
  select(Library_ID, Lab, Type) %>%
  filter(!str_detect(Library_ID, "ARS|HPD"))

# list all samples and blanks from Otago
otago <- metadata %>%
  anti_join(., cfdp_fail) %>%
  filter(Lab == "Otago") %>%
  select(Library_ID, Lab, Type) %>%
  filter(Library_ID != "HCLVMBCX2-3505-13-00-01_S13") %>%
  arrange(Library_ID) %>%
  filter(Type != "blank") %>%
  pull(Library_ID)

# list all samples and blanks from Oklahoma
oklahoma <- metadata %>%
  anti_join(., cfdp_fail) %>%
  filter(Lab == "Oklahoma") %>%
  select(Library_ID, Lab, Type) %>%
  filter(Type != "bone") %>%
  filter(Type != "blank") %>%
  pull(Library_ID)

# list all samples and blanks from Jena
jena <- metadata %>%
  anti_join(., cfdp_fail) %>%
  filter(Lab == "Jena") %>%
  select(Library_ID, Lab, Type) %>%
  filter(Type != "bone") %>%
  filter(Type != "blank") %>%
  filter(!str_detect(Library_ID, "HPD")) %>%
  pull(Library_ID)

# list all samples to include
sample_list <- c(otago,oklahoma,jena)

```

```{r colors}
Island_colors = c("Tongatapu" = "#616161", #  
                  "Rurutu" = "#9d654c", 
                  "Flores" = "#f6865e", 
                  "Rapa Nui" = "#cc79a7", 
                  "Futuna" = "#7D3560", 
                  "Efate" = "#7472af", 
                  "Vao" = "#098BD9",
                  "Efate 3000 BP" = "#7DCCFF", 
                  "Uripiv" = "#e1f2fe", 
                  "Viti Levu" = "#1ec1a7", 
                  "Raiatea" = "#148f77", 
                  "Taumako" = "#4e7705", 
                  "Watom" = "#86c73a")


# Set microshades colors
Lab_colors = c("Jena" = microshades_palette("micro_cvd_purple", 1, lightest = FALSE),
                "Oklahoma" =  microshades_palette("micro_cvd_blue", 1, lightest = FALSE),
                "Otago" = microshades_palette("micro_cvd_green", 1, lightest = FALSE))

Type_colors <- c("Arch. Bone" = "#148F77", "blank" = "#8B8B8B", "Arch. calculus" = "black")


```

```{r}
Island_shapes = c("Tongatapu" = 21, 
                  "Rurutu" = 21, 
                  "Flores" = 21, 
                  "Rapa Nui" = 21, 
                  "Futuna" = 21, 
                  "Efate" = 21, 
                  "Vao" = 21,
                  "Efate 3000 BP" = 21, 
                  "Uripiv" = 21, 
                  "Viti Levu" = 21, 
                  "Raiatea" = 21, 
                  "Taumako" = 21, 
                  "Watom" = 21)

Lab_shapes = c(21,21,21)
Lab_size = c(4,4,4)

Type_shapes = c(21,21,21,21,21,21,21,21)
Type_size = c(4,4,4,4,4,4,4,4)

```


```{r plot_pca_fxn}

# add functions for:
# making the input table for PCA?

# plotting PCA with colored dots
plot_pca <- function(df, pc1, pc2, color_group, shape_group) {
    metadata_group_colors <- get(paste(color_group, "_colors", sep = ""))
    metadata_group_shapes <- get(paste(shape_group, "_shapes", sep = ""))

    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%")
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(metadata, by = "Library_ID")
    
    color_group = df_X[[color_group]]
    shape_group = df_X[[shape_group]]

                       if (pc1 == 'PC1') {
                           pc1 <- df_X$PC1
                           exp_var_pc1 <- exp_var[1]
                           xaxis <- c("PC1")
                       }  else if (pc1 == 'PC2') {
                           pc1 <- df_X$PC2
                           exp_var_pc1 <- exp_var[2]
                           xaxis <- c("PC2")
                       } else if (pc1 == 'PC3') {
                           pc1 <- df_X$PC3
                           exp_var_pc1 <- exp_var[3]
                           xaxis <- c("PC3")
                       }

                       if (pc2 == 'PC1') {
                           pc2 <- df_X$PC1
                           exp_var_pc2 <- exp_var[1]
                           yaxis <- c("PC1")
                       }  else if (pc2 == 'PC2') {
                           pc2 <- df_X$PC2
                           exp_var_pc2 <- exp_var[2]
                           yaxis <- c("PC2")
                       } else if (pc2 == 'PC3') {
                           pc2 <- df_X$PC3
                           exp_var_pc2 <- exp_var[3]
                           yaxis <- c("PC3")
                       }

    pca_plot <- ggplot(df_X, aes(pc1, pc2)) +
                        geom_point(aes(fill = color_group, shape = shape_group), size = 4, stroke = 0.3, shape = 21) +
                        scale_fill_manual(values = metadata_group_colors) +
                        xlab(paste(xaxis, " - ", exp_var_pc1)) +
                        ylab(paste(yaxis, " - ", exp_var_pc2)) +
                        theme_minimal(base_size = 14) +
                        theme(legend.position = "right") +
                        theme(legend.title = element_blank())

    return(pca_plot)
}

```


```{r pca_plot_cont_fxn}
# for continuous data
plot_pca_cont <- function(df_pca, pc1, pc2, color_group, shape_group, ncomps, title_text) {

    exp_var <- paste0(round(df_pca$prop_expl_var$X * 100, 2), "%")
    df_X <- df_pca$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(metadata, by = "Library_ID")

    color_group = df_X[[color_group]]
    shape_group = df_X[[shape_group]]

    if (pc1 == 'PC1') {
        pc1 <- df_X$PC1
        exp_var_pc1 <- exp_var[1]
        xaxis <- c("PC1")
    }  else if (pc1 == 'PC2') {
        pc1 <- df_X$PC2
        exp_var_pc1 <- exp_var[2]
        xaxis <- c("PC2")
    } else if (pc1 == 'PC3') {
        pc1 <- df_X$PC3
        exp_var_pc1 <- exp_var[3]
        xaxis <- c("PC3")
    }

    if (pc2 == 'PC1') {
        pc2 <- df_X$PC1
        exp_var_pc2 <- exp_var[1]
        yaxis <- c("PC1")
    }  else if (pc2 == 'PC2') {
        pc2 <- df_X$PC2
        exp_var_pc2 <- exp_var[2]
        yaxis <- c("PC2")
    } else if (pc2 == 'PC3') {
        pc2 <- df_X$PC3
        exp_var_pc2 <- exp_var[3]
        yaxis <- c("PC3")
    }

    pca_plot <- ggplot(df_X, aes(x = pc1, y = pc2)) +
     geom_star(aes(fill = color_group, starshape = shape_group), size = 3.5, stroke = 0.3) + # , shape = 21
     scale_fill_viridis_c(option = "C") +
     scale_starshape_manual(values = c(15,9,11)) +
     # scale_shape_manual(values = c(16, 17, 15, 18, 14,9,13,24)) +
    # stat_ellipse() +
     xlab(paste(xaxis, " - ", exp_var_pc1)) +
     ylab(paste(yaxis, " - ", exp_var_pc2)) +
     theme_minimal(base_size = 16) +
     theme(text = element_text(size=16)) +
     theme(legend.title = element_blank(),
           legend.key.size = unit(2,"mm"),
           legend.text = element_text(size = 10)) + 
     theme(legend.position = "right") +
     ggtitle(title_text) + theme(plot.title = element_text(size = 10))

    return(pca_plot)
}

```


```{r plot_bi}
# this function plots the PCA with biplots to look at the categories with strongest loadings
plot_pca_bi <- function(df, pc1, pc2, metadata_group, columntitle, GeneFamilys = c("TRUE","FALSE")) {
    metadata_group_colors <- get(paste(metadata_group, "_colors", sep = ""))
    metadata_group_shapes <- get(paste(metadata_group, "_shapes", sep = ""))
   
    arrow_pc <- enquo(columntitle)
    
    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%") # explained variance for x- and y-labels
    
    # select only the PCs from the PCA and add metadata
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(metadata, by = "Library_ID")
    
    metadata_group = df_X[[metadata_group]]
    
    corr_lam <- df$sdev[c("PC1", "PC2", "PC3")] * sqrt(nrow(df_X))

    df_X <- df_X %>%
      mutate(PC1 = PC1 / corr_lam[1],
             PC2 = PC2 / corr_lam[2], 
             PC3 = PC3 / corr_lam[3])

    # select the correct PC column and explained variance for PC1
    if (pc1 == 'PC1') {
        Pc1 <- df_X$PC1
        exp_var_pc1 <- exp_var[1]
        xaxis <- c("PC1")
    } else if (pc1 == 'PC2') {
        Pc1 <- df_X$PC2
        exp_var_pc1 <- exp_var[2]
        xaxis <- c("PC2")
    } else if (pc1 == 'PC3') {
        Pc1 <- df_X$PC3
        exp_var_pc1 <- exp_var[3]
        xaxis <- c("PC3")
   }
    
    # select the correct PC column and explained variance for PC2
    if (pc2 == 'PC1') {
        Pc2 <- df_X$PC1
        exp_var_pc2 <- exp_var[1]
        yaxis <- c("PC1")
 } else if (pc2 == 'PC2') {
       Pc2 <- df_X$PC2
       exp_var_pc2 <- exp_var[2]
       yaxis <- c("PC2")
    } else if (pc2 == 'PC3') {
       Pc2 <- df_X$PC3
       exp_var_pc2 <- exp_var[3]
       yaxis <- c("PC3")
   }

    # Identify the 10 GeneFamilys that have highest positive and negative loadings in the selected PC
    pws_10 <- df$loadings$X %>%
      as.data.frame(.) %>%
      rownames_to_column(var = "Function") %>%
      top_n(10, !!arrow_pc)
   if (GeneFamilys == T) {
      pws_10 <- pws_10 %>%
        rename(GeneFamily = Function) %>%
        mutate(Function = sapply(GeneFamily, function(f) {
                                  unlist(str_split(f, ":"))[1]
                           })) 
    }

    neg_10 <- df$loadings$X %>%
      as.data.frame(.) %>%
      rownames_to_column(var = "Function") %>%
      top_n(-10, !!arrow_pc)
    if (GeneFamilys == T) {
      neg_10 <- neg_10 %>%
      rename(GeneFamily = Function) %>%
       mutate(Function = sapply(GeneFamily, function(f) {
                                  unlist(str_split(f, ":"))[1]
                           })) 
    }


    pca_plot_bi <- ggplot(df_X, aes(x = Pc1, y = Pc2)) +
      geom_point(aes(shape = metadata_group, fill = metadata_group), size = 3.5, stroke = 0.3) +
      geom_segment(data = pws_10,
                   aes(xend = get(paste(pc1)), yend = get(paste(pc2))),
                   x = 0, y = 0, colour = "black",
                   size = 0.5,
                   arrow = arrow(length = unit(0.03, "npc"))) +
      geom_label_repel(data = pws_10,
                   aes(x = get(paste(pc1)), y = get(paste(pc2)), label = Function),
                   size = 2.5, colour = "grey20", label.padding = 0.2, max.overlaps = 12) +
      geom_segment(data = neg_10,
                   aes(xend = get(paste(pc1)), yend = get(paste(pc2))),
                   x = 0, y = 0, colour = "grey50",
                   size = 0.5,
                   arrow = arrow(length = unit(0.03, "npc"))) +
      geom_label_repel(data = neg_10,
                   aes(x = get(paste(pc1)), y = get(paste(pc2)), label = Function),
                   size = 2.5, colour = "grey20", label.padding = 0.2, max.overlaps = 12) +
      labs(x = paste(xaxis, " - ", exp_var_pc1),
           y = paste(yaxis, " - ", exp_var_pc2)) +
      scale_fill_manual(values = metadata_group_colors) +
      scale_shape_manual(values = metadata_group_shapes) +
      theme_minimal() + theme(text = element_text(size = 14)) +
      theme(legend.position = "right")

    return(pca_plot_bi)
}


```


## Genus
```{r load_genus_tables}
# read in the species table
rs_raw_gn <- fread("./05-Documentation.backup/oceania_malt_cRefSeq_summarized_genus_20201105.txt") %>%
  rename(Genus = 1) %>%
  full_join(., fread("./05-Documentation.backup/sources_MALT_cRefSeq_JFY_genus_summarized.txt") %>%
              rename(Genus = 1), by = "Genus") %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_blanks_cRefSeq_genus_summarized_20201130.txt") %>%
              rename(Genus = 1), by = "Genus") %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_bones_cRefSeq_genus_summarized_20201202.txt") %>%
              rename(Genus = 1), by = "Genus") %>%
  filter(Genus != "Homo")

# clean the file names
colnames(rs_raw_gn) <- gsub("MeganServer::", "", colnames(rs_raw_gn)) 
colnames(rs_raw_gn) <- gsub(".unmapped", "", colnames(rs_raw_gn)) 
colnames(rs_raw_gn) <- gsub(".SG1.1", "", colnames(rs_raw_gn)) 
colnames(rs_raw_gn) <- gsub(".SG1.2", "", colnames(rs_raw_gn)) 
colnames(rs_raw_gn) <- gsub("_L001_R1_001.fastq.combined.fq.prefixed.extractunmapped.bam", "", colnames(rs_raw_gn)) 
colnames(rs_raw_gn) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg19unmapped", "", colnames(rs_raw_gn)) 
colnames(rs_raw_gn) <- gsub("_S0_L001_R1_001.fastq.truncated.prefixed.extractunmapped.bam", "", colnames(rs_raw_gn)) 
colnames(rs_raw_gn) <- gsub("_S0_L001_R1_000.fastq.merged.prefixed.hg1", "", colnames(rs_raw_gn)) 
colnames(rs_raw_gn) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg1", "", colnames(rs_raw_gn)) 
colnames(rs_raw_gn) <- gsub(".SG1.2_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(rs_raw_gn))  
colnames(rs_raw_gn) <- gsub("_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(rs_raw_gn)) 

# now do a pretend transpose to replace the NAs with 0s
rs_raw_gn <- rs_raw_gn %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = replace_na(Counts, 0)) %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts")

```


```{r outliers}
# HCLVMBCX2-3505-13-00-01_S13 appears to be a sample, although it's supposed to be a blank
# remove it from all further analyses

metadata_filt <- metadata %>%
  anti_join(., cfdp_fail) %>%
  filter(Library_ID != "HCLVMBCX2-3505-13-00-01_S13")

genus_raw_pass <- rs_raw_gn %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  pivot_wider(names_from = "Genus", values_from = "Counts") %>%
  anti_join(., cfdp_fail) %>%
  filter(Library_ID != "HCLVMBCX2-3505-13-00-01_S13") %>%
  # remove all taxa that have to entries anymore
  pivot_longer(!Library_ID, names_to = "Genus", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  adorn_totals(where = "col") %>%
  select(Total, everything()) %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  pivot_wider(names_from = "Genus", values_from = "Counts")

```

### Decontam all together

```{r all_table}
# note that all poorly-preserved samples are removed from genus_raw_pass although they're still in the lab lists

# select all samples and remove all contaminants
# save row names
rownames_gn_all_filt <- genus_raw_pass %>%
  filter(Library_ID %in% sample_list) %>%
  pivot_longer(!Library_ID, names_to = "Genus", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  anti_join(., cont_all_gn %>%
              select(Genus), by = "Genus") %>%
  adorn_totals(where = "col") %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  pull(Genus)

gn_all_filt <- genus_raw_pass %>%
  filter(Library_ID %in% sample_list) %>%
  pivot_longer(!Library_ID, names_to = "Genus", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  anti_join(., cont_all_gn %>%
              select(Genus), by = "Genus") %>%
  adorn_totals(where = "col") %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  # remove the totals attribute so can make column sums later if necessary by converting to matrix and back
  as.matrix(., rownames = "Genus") %>%
  as_tibble(.) %>%
  mutate(Genus = rownames_gn_all_filt) %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = as.numeric(Counts)) %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts")

# remove the totals attribute so can make column sums later if necessary
# attr(gn_all_filt, "totals") <- NULL

```


```{r all_plot_no_filt}

gn_all_filt_1 <- gn_all_filt %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Genus,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.gn_all_filt_1_pca <- tune.pca(gn_all_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
gn_all_filt_1.pca <- pca(gn_all_filt_1, ncomp = 3, logratio = 'CLR')

Type_colors <- c("black","blue","turquoise", "red")
Type_shapes <- c(19,17,15,9)


# plot the PCA
gn_all_plot <- plot_pca(gn_all_filt_1.pca, "PC1", "PC2", "Lab", "Lab")
gn_all_plot

# top 20 genera separating samples along PC1? (10+, 10-)
gn_all_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Genus = rownames(gn_all_filt_1.pca$loadings$X)) %>%
  select(Genus, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., gn_all_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Genus = rownames(gn_all_filt_1.pca$loadings$X)) %>%
            select(Genus, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r pca_cont_all_gn}
# average sequence length
plot_pca_cont(gn_all_filt_1.pca, "PC1", "PC2", "seq_len_avg", "Lab", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(gn_all_filt_1.pca, "PC1", "PC2", "gc_avg", "Lab", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(gn_all_filt_1.pca, "PC1", "PC2", "No_reads_no_human", "Lab", 3, "Total Reads")

# total number of reads
plot_pca_cont(gn_all_filt_1.pca, "PC1", "PC2", "Age_mean", "Lab", 3, "Mean Age (BP)")

```

```{r adonis_all_species}

metadata_all <- metadata %>%
    filter(Library_ID %in% sample_list)
  

gn_all_filt_1 <- gn_all_filt %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Genus,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.gn_all_filt_1_pca <- tune.pca(gn_all_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
gn_all_filt_1.pca <- pca(gn_all_filt_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
gn_all_filt_1.euc <- vegdist(gn_all_filt_1.pca$X, method = "euclidean", na.rm = F)
gn_all_filt_1.euc.pcoa <- ape::pcoa(gn_all_filt_1.euc)

gn_all_filt_1.euc.pcoavectors <- as.data.frame(gn_all_filt_1.euc.pcoa$vectors)
gn_all_filt_1.euc.pcoavectors <- gn_all_filt_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

gn_all_filt_1.euc.vecmet <- left_join(gn_all_filt_1.euc.pcoavectors, metadata_filt, by = "Library_ID")

gn_all_filt_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Island, shape = Island)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Island_colors) +
  scale_shape_manual(values = Island_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Site_code")

# gn_all_filt_1 %>%
#   rownames_to_column("Library_ID") %>%
#   select(Library_ID) %>%
#   anti_join(., metadata %>% select(Library_ID, Anginosus), by  = "Library_ID")


# test for difference between different metadata categories 
adonis2(gn_all_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")

all_permanova <- adonis2(gn_all_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(., adonis2(gn_all_filt_1.euc ~ Island, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(gn_all_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(gn_all_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(gn_all_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(gn_all_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","***","***","***","ns"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)

# all together
adonis2(gn_all_filt_1.euc ~ Lab + Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(gn_all_filt_1.euc ~ Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)

```

### Filter out taxa in only 1 lab
```{r all_plot_1lab_filt}

# Taxa in only 1 group
gn_1_lab <- gn_all_filt %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts == 0, 0, 1)) %>%
  full_join(., metadata %>%
              select(Library_ID, Lab)) %>%
  # get counts per lab for each genus
  group_by(Genus, Lab) %>%
  summarize(In_lab = sum(Counts)) %>%
  mutate(In_lab = ifelse(In_lab == 0,0,1)) %>%
  ungroup() %>%
  group_by(Genus) %>%
  summarize(num_labs = sum(In_lab)) %>%
  arrange(num_labs) %>%
  filter(num_labs == 1)


gn_all_1lab_filt_1 <- gn_all_filt %>%
  anti_join(., gn_1_lab) %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Genus,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.gn_all_1lab_filt_1_pca <- tune.pca(gn_all_1lab_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
gn_all_1lab_filt_1.pca <- pca(gn_all_1lab_filt_1, ncomp = 3, logratio = 'CLR')

Type_colors <- c("black","blue","turquoise", "red")
Type_shapes <- c(19,17,15,9)


# plot the PCA
gn_all_1lab_plot <- plot_pca(gn_all_1lab_filt_1.pca, "PC1", "PC2", "Lab", "Lab")
gn_all_1lab_plot

# top 20 genera separating samples along PC1? (10+, 10-)
gn_all_1lab_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Genus = rownames(gn_all_1lab_filt_1.pca$loadings$X)) %>%
  select(Genus, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., gn_all_1lab_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Genus = rownames(gn_all_1lab_filt_1.pca$loadings$X)) %>%
            select(Genus, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r pca_cont_1lab_gn}
# average sequence length
plot_pca_cont(gn_all_1lab_filt_1.pca, "PC1", "PC2", "seq_len_avg",  "Lab", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(gn_all_1lab_filt_1.pca, "PC1", "PC2", "gc_avg",  "Lab", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(gn_all_1lab_filt_1.pca, "PC1", "PC2", "No_reads_no_human",  "Lab", 3, "Total Reads")

# total number of reads
plot_pca_cont(gn_all_1lab_filt_1.pca, "PC1", "PC2", "Age_mean",  "Lab", 3, "Mean Age (BP)")

```

### Filter out taxa in <= 1/3 samples in each Lab
```{r all_plot_1lab_filt}
# Taxa in at least 1/3 of samples in at least 2 of the 3 labs
gn_all_filt_presence_more_30 <- gn_all_filt %>%
  # mutate_if(is.numeric, ~1 * (. > 0)) %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts == 0, 0, 1)) %>%
  left_join(., metadata %>%
              select(Library_ID, Lab), by = "Library_ID") %>%
  arrange(Genus, Lab) %>%
  group_by(Genus,Lab) %>%
  summarize(Percent_presence = sum(Counts)/length(Counts)*100) %>%
  # select the pathways that are present in at least 1/3 of at least 2 groups
  ungroup() %>%
  # pathways present in > 1/3 of samples per lab
  filter(Percent_presence >= 33.33) %>%
  group_by(Genus) %>%
  count()%>%
  filter(n > 1)%>%
  # pull out these genera to keep
  select(Genus)

```

```{r}


gn_all_pres33_filt_1 <- gn_all_filt %>%
  inner_join(., gn_all_filt_presence_more_30) %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Genus,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.gn_all_pres33_filt_1_pca <- tune.pca(gn_all_pres33_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
gn_all_pres33_filt_1.pca <- pca(gn_all_pres33_filt_1, ncomp = 3, logratio = 'CLR')

Type_colors <- c("black","blue","turquoise", "red")
Type_shapes <- c(19,17,15,9)


# plot the PCA
gn_all_pres33_plot <- plot_pca(gn_all_pres33_filt_1.pca, "PC1", "PC2", "Lab", "Lab")
gn_all_pres33_plot
plot_pca(gn_all_pres33_filt_1.pca, "PC1", "PC2", "Island", "Lab")

# top 20 genera separating samples along PC1? (10+, 10-)
gn_all_pres33_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Genus = rownames(gn_all_pres33_filt_1.pca$loadings$X)) %>%
  select(Genus, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., gn_all_pres33_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Genus = rownames(gn_all_pres33_filt_1.pca$loadings$X)) %>%
            select(Genus, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r, eval = F}
# use this table for downstream analyses
fwrite(gn_all_pres33_filt_1, "./05-Documentation.backup/pacific_genus_decontam_pres33_filt.tsv", quote = F, sep = "\t")
# fwrite(gn_all_pres33_filt_1, "../strep_clades/00-documentation/pacific_genus_decontam_pres33_filt.tsv", quote = F, sep = "\t")

```


```{r pca_cont_pres33_gn}
# average sequence length
plot_pca_cont(gn_all_pres33_filt_1.pca, "PC1", "PC2", "seq_len_avg",  "Lab", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(gn_all_pres33_filt_1.pca, "PC1", "PC2", "gc_avg",  "Lab", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(gn_all_pres33_filt_1.pca, "PC1", "PC2", "No_reads_no_human",  "Lab", 3, "Total Reads")

# total number of reads
plot_pca_cont(gn_all_pres33_filt_1.pca, "PC1", "PC2", "Age_mean",  "Lab", 3, "Mean Age (BP)")

```

Run a PERMANOVA with adonis2 on the abundance-filtered clr-transformed species matrix.
```{r adonis_genus}

metadata_all <- metadata %>%
    filter(Library_ID %in% sample_list)
  

gn_all_pres33_filt_1 <- gn_all_filt %>%
  inner_join(., gn_all_filt_presence_more_30) %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Genus,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.gn_all_pres33_filt_1_pca <- tune.pca(gn_all_pres33_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
gn_all_pres33_filt_1.pca <- pca(gn_all_pres33_filt_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
gn_all_pres33_filt_1.euc <- vegdist(gn_all_pres33_filt_1.pca$X, method = "euclidean", na.rm = F)
gn_all_pres33_filt_1.euc.pcoa <- ape::pcoa(gn_all_pres33_filt_1.euc)

gn_all_pres33_filt_1.euc.pcoavectors <- as.data.frame(gn_all_pres33_filt_1.euc.pcoa$vectors)
gn_all_pres33_filt_1.euc.pcoavectors <- gn_all_pres33_filt_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

gn_all_pres33_filt_1.euc.vecmet <- left_join(gn_all_pres33_filt_1.euc.pcoavectors, metadata_filt, by = "Library_ID")

gn_all_pres33_filt_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Island, shape = Island)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Island_colors) +
  scale_shape_manual(values = Island_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Site_code")

# gn_all_pres33_filt_1 %>%
#   rownames_to_column("Library_ID") %>%
#   select(Library_ID) %>%
#   anti_join(., metadata %>% select(Library_ID, Anginosus), by  = "Library_ID")


# test for difference between different metadata categories 
adonis2(gn_all_pres33_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")

all_permanova <- adonis2(gn_all_pres33_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(., adonis2(gn_all_pres33_filt_1.euc ~ Island, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(gn_all_pres33_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(gn_all_pres33_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(gn_all_pres33_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(gn_all_pres33_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","***","**","**","ns"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)

# all together
adonis2(gn_all_pres33_filt_1.euc ~ Lab + Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(gn_all_pres33_filt_1.euc ~ Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)

```


### Filter out taxa that are less than 0.005% relative abundane across all samples?
```{r}
# let's look at the abundance of these taxa
gn_all_filt %>%
  as_tibble(.) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  select(Genus, Percent) %>%
  arrange(desc(Percent)) %>%
  mutate(Percent = round(Percent, 4))

```


```{r all_plot_1lab_filt}

# Taxa present at < 0.005% relative abundance
gn_abund005_lab <- gn_all_filt %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent <= 0.005) %>%
  # select(Genus, Total, Percent)
  select(Genus)

gn_all_abund005_filt_1 <- gn_all_filt %>%
  anti_join(., gn_abund005_lab) %>%
  pivot_longer(!Genus, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Genus,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.gn_all_abund005_filt_1_pca <- tune.pca(gn_all_abund005_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
gn_all_abund005_filt_1.pca <- pca(gn_all_abund005_filt_1, ncomp = 3, logratio = 'CLR')

Type_colors <- c("black","blue","turquoise", "red")
Type_shapes <- c(19,17,15,9)


# plot the PCA
gn_all_abund005_plot <- plot_pca(gn_all_abund005_filt_1.pca, "PC1", "PC2", "Lab", "Lab")
gn_all_abund005_plot

# top 20 genera separating samples along PC1? (10+, 10-)
gn_all_abund005_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Genus = rownames(gn_all_abund005_filt_1.pca$loadings$X)) %>%
  select(Genus, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., gn_all_abund005_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Genus = rownames(gn_all_abund005_filt_1.pca$loadings$X)) %>%
            select(Genus, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

patchwork example
```{r}

gn_plot <- gn_all_plot + gn_all_1lab_plot + gn_all_pres33_plot + gn_all_abund005_plot +
  plot_layout(nrow = 1)  + 
  plot_layout(guides = "collect")

gn_plot

```



Biplots
```{r genus_pres33_biplot}
# which paths distinguish samples?
gn_unfilt_1_pc1 <- plot_pca_bi(gn_all_pres33_filt_1.pca, "PC1", "PC2", "Lab", PC1, GeneFamily = F)
gn_unfilt_1_pc1

# which paths distinguish samples?
gn_unfilt_1_pc2 <- plot_pca_bi(gn_all_pres33_filt_1.pca, "PC1", "PC2", "Lab", PC2, GeneFamily = F)
gn_unfilt_1_pc2

```



## Species
```{r load_species_tables}
# read in the species table
rs_raw_sp <- fread("./05-Documentation.backup/oceania_malt_cRefSeq_summarized_species_20201105.txt") %>%
  rename(Species = 1) %>%
  full_join(., fread("./05-Documentation.backup/sources_MALT_cRefSeq_JFY_species_summarized.txt") %>%
              rename(Species = 1), by = "Species") %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_blanks_cRefSeq_species_summarized_20201130.txt") %>%
              rename(Species = 1), by = "Species") %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_bones_cRefSeq_species_summarized_20201202.txt") %>%
              rename(Species = 1), by = "Species") %>%
  filter(Species != "Homo")

# clean the file names
colnames(rs_raw_sp) <- gsub("MeganServer::", "", colnames(rs_raw_sp)) 
colnames(rs_raw_sp) <- gsub(".unmapped", "", colnames(rs_raw_sp)) 
colnames(rs_raw_sp) <- gsub(".SG1.1", "", colnames(rs_raw_sp)) 
colnames(rs_raw_sp) <- gsub(".SG1.2", "", colnames(rs_raw_sp)) 
colnames(rs_raw_sp) <- gsub("_L001_R1_001.fastq.combined.fq.prefixed.extractunmapped.bam", "", colnames(rs_raw_sp)) 
colnames(rs_raw_sp) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg19unmapped", "", colnames(rs_raw_sp)) 
colnames(rs_raw_sp) <- gsub("_S0_L001_R1_001.fastq.truncated.prefixed.extractunmapped.bam", "", colnames(rs_raw_sp)) 
colnames(rs_raw_sp) <- gsub("_S0_L001_R1_000.fastq.merged.prefixed.hg1", "", colnames(rs_raw_sp)) 
colnames(rs_raw_sp) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg1", "", colnames(rs_raw_sp)) 
colnames(rs_raw_sp) <- gsub(".SG1.2_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(rs_raw_sp))  
colnames(rs_raw_sp) <- gsub("_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(rs_raw_sp)) 

# now do a pretend transpose to replace the NAs with 0s
rs_raw_sp <- rs_raw_sp %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = replace_na(Counts, 0)) %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts")

```


```{r outliers}
# HCLVMBCX2-3505-13-00-01_S13 appears to be a sample, although it's supposed to be a blank
# remove it from all further analyses

metadata_filt <- metadata %>%
  anti_join(., cfdp_fail) %>%
  filter(Library_ID != "HCLVMBCX2-3505-13-00-01_S13")

species_raw_pass <- rs_raw_sp %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  pivot_wider(names_from = "Species", values_from = "Counts") %>%
  anti_join(., cfdp_fail) %>%
  filter(Library_ID != "HCLVMBCX2-3505-13-00-01_S13") %>%
  # remove all taxa that have to entries anymore
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  adorn_totals(where = "col") %>%
  select(Total, everything()) %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  pivot_wider(names_from = "Species", values_from = "Counts")

```


### Decontam all together

```{r all_table}
# note that all poorly-preserved samples are removed from species_raw_pass although they're still in the lab lists

# select all samples and remove all contaminants
# save row names
rownames_sp_all_filt <- species_raw_pass %>%
  filter(Library_ID %in% sample_list) %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  anti_join(., cont_all_sp %>%
              select(Species), by = "Species") %>%
  adorn_totals(where = "col") %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  pull(Species)

sp_all_filt <- species_raw_pass %>%
  filter(Library_ID %in% sample_list) %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  anti_join(., cont_all_sp %>%
              select(Species), by = "Species") %>%
  adorn_totals(where = "col") %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  # remove the totals attribute so can make column sums later if necessary by converting to matrix and back
  as.matrix(., rownames = "Species") %>%
  as_tibble(.) %>%
  mutate(Species = rownames_sp_all_filt) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = as.numeric(Counts)) %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts")

# remove the totals attribute so can make column sums later if necessary
# attr(sp_all_filt, "totals") <- NULL

```


```{r all_plot_no_filt}

sp_all_filt_1 <- sp_all_filt %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.sp_all_filt_1_pca <- tune.pca(sp_all_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_all_filt_1.pca <- pca(sp_all_filt_1, ncomp = 3, logratio = 'CLR')

Type_colors <- c("black","blue","turquoise", "red")
Type_shapes <- c(19,17,15,9)


# plot the PCA
sp_all_plot <- plot_pca(sp_all_filt_1.pca, "PC1", "PC2", "Lab", "Lab")
sp_all_plot

# top 20 genera separating samples along PC1? (10+, 10-)
sp_all_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(sp_all_filt_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., sp_all_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(sp_all_filt_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r pca_cont_species}
# average sequence length
plot_pca_cont(sp_all_filt_1.pca, "PC1", "PC2", "seq_len_avg",  "Lab", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(sp_all_filt_1.pca, "PC1", "PC2", "gc_avg",  "Lab", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(sp_all_filt_1.pca, "PC1", "PC2", "No_reads_no_human",  "Lab", 3, "Total Reads")

# total number of reads
plot_pca_cont(sp_all_filt_1.pca, "PC1", "PC2", "Age_mean",  "Lab", 3, "Mean Age (BP)")

```

```{r adonis_all_species}

metadata_all <- metadata %>%
    filter(Library_ID %in% sample_list)
  

sp_all_filt_1 <- sp_all_filt %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.sp_all_filt_1_pca <- tune.pca(sp_all_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_all_filt_1.pca <- pca(sp_all_filt_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
sp_all_filt_1.euc <- vegdist(sp_all_filt_1.pca$X, method = "euclidean", na.rm = F)
sp_all_filt_1.euc.pcoa <- ape::pcoa(sp_all_filt_1.euc)

sp_all_filt_1.euc.pcoavectors <- as.data.frame(sp_all_filt_1.euc.pcoa$vectors)
sp_all_filt_1.euc.pcoavectors <- sp_all_filt_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

sp_all_filt_1.euc.vecmet <- left_join(sp_all_filt_1.euc.pcoavectors, metadata_filt, by = "Library_ID")

sp_all_filt_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Island, shape = Island)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Island_colors) +
  scale_shape_manual(values = Island_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Site_code")

# sp_all_filt_1 %>%
#   rownames_to_column("Library_ID") %>%
#   select(Library_ID) %>%
#   anti_join(., metadata %>% select(Library_ID, Anginosus), by  = "Library_ID")


# test for difference between different metadata categories 
adonis2(sp_all_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")

all_permanova <- adonis2(sp_all_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(., adonis2(sp_all_filt_1.euc ~ Island, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","***","***","***","ns"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)

# all together
adonis2(sp_all_filt_1.euc ~ Lab + Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(sp_all_filt_1.euc ~ Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)

```


### Filter out taxa in only 1 lab
```{r all_plot_1lab_filt}

# Taxa in only 1 group
sp_1_lab <- sp_all_filt %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts == 0, 0, 1)) %>%
  full_join(., metadata %>%
              select(Library_ID, Lab)) %>%
  # get counts per lab for each species
  group_by(Species, Lab) %>%
  summarize(In_lab = sum(Counts)) %>%
  mutate(In_lab = ifelse(In_lab == 0,0,1)) %>%
  ungroup() %>%
  group_by(Species) %>%
  summarize(num_labs = sum(In_lab)) %>%
  arrange(num_labs) %>%
  filter(num_labs == 1)


sp_all_1lab_filt_1 <- sp_all_filt %>%
  anti_join(., sp_1_lab) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.sp_all_1lab_filt_1_pca <- tune.pca(sp_all_1lab_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_all_1lab_filt_1.pca <- pca(sp_all_1lab_filt_1, ncomp = 3, logratio = 'CLR')

# plot the PCA
sp_all_1lab_plot <- plot_pca(sp_all_1lab_filt_1.pca, "PC1", "PC2", "Lab", "Lab")
sp_all_1lab_plot

# top 20 genera separating samples along PC1? (10+, 10-)
sp_all_1lab_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(sp_all_1lab_filt_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., sp_all_1lab_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(sp_all_1lab_filt_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r pca_cont_1lab_species}
# average sequence length
plot_pca_cont(sp_all_1lab_filt_1.pca, "PC1", "PC2", "seq_len_avg",  "Lab", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(sp_all_1lab_filt_1.pca, "PC1", "PC2", "gc_avg",  "Lab", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(sp_all_1lab_filt_1.pca, "PC1", "PC2", "No_reads_no_human",  "Lab", 3, "Total Reads")

# total number of reads
plot_pca_cont(sp_all_1lab_filt_1.pca, "PC1", "PC2", "Age_mean",  "Lab", 3, "Mean Age (BP)")

```

```{r adonis_all_species}

metadata_all <- metadata %>%
    filter(Library_ID %in% sample_list)
  

sp_1_lab <- sp_all_filt %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts == 0, 0, 1)) %>%
  full_join(., metadata %>%
              select(Library_ID, Lab)) %>%
  # get counts per lab for each species
  group_by(Species, Lab) %>%
  summarize(In_lab = sum(Counts)) %>%
  mutate(In_lab = ifelse(In_lab == 0,0,1)) %>%
  ungroup() %>%
  group_by(Species) %>%
  summarize(num_labs = sum(In_lab)) %>%
  arrange(num_labs) %>%
  filter(num_labs == 1)


sp_all_1lab_filt_1 <- sp_all_filt %>%
  anti_join(., sp_1_lab) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.sp_all_1lab_filt_1_pca <- tune.pca(sp_all_1lab_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_all_1lab_filt_1.pca <- pca(sp_all_1lab_filt_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
sp_all_1lab_filt_1.euc <- vegdist(sp_all_1lab_filt_1.pca$X, method = "euclidean", na.rm = F)
sp_all_1lab_filt_1.euc.pcoa <- ape::pcoa(sp_all_1lab_filt_1.euc)

sp_all_1lab_filt_1.euc.pcoavectors <- as.data.frame(sp_all_1lab_filt_1.euc.pcoa$vectors)
sp_all_1lab_filt_1.euc.pcoavectors <- sp_all_1lab_filt_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

sp_all_1lab_filt_1.euc.vecmet <- left_join(sp_all_1lab_filt_1.euc.pcoavectors, metadata_filt, by = "Library_ID")

sp_all_1lab_filt_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Island, shape = Island)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Island_colors) +
  scale_shape_manual(values = Island_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Site_code")

# sp_all_1lab_filt_1 %>%
#   rownames_to_column("Library_ID") %>%
#   select(Library_ID) %>%
#   anti_join(., metadata %>% select(Library_ID, Anginosus), by  = "Library_ID")


# test for difference between different metadata categories 
adonis2(sp_all_1lab_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_1lab_filt_1.euc ~ Island, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_1lab_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_1lab_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_1lab_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_1lab_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin")

all_1lab_permanova <- adonis2(sp_all_1lab_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_1lab_permanova  <- all_1lab_permanova %>% bind_rows(., adonis2(sp_all_1lab_filt_1.euc ~ Island, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_1lab_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_1lab_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_1lab_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_1lab_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_1lab_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","***","***","***","ns"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_1lab_permanova.tsv", sep = "\t", quote =  F)

# all together
adonis2(sp_all_1lab_filt_1.euc ~ Lab + Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(sp_all_1lab_filt_1.euc ~ Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)

```


### Filter out taxa in <= 1/3 of samples in each Lab
```{r all_plot_1lab_filt}
# Taxa in at least 1/3 of samples in at least 2 of the 3 labs
sp_all_filt_presence_more_30 <- sp_all_filt %>%
  # mutate_if(is.numeric, ~1 * (. > 0)) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts == 0, 0, 1)) %>%
  left_join(., metadata %>%
              select(Library_ID, Lab), by = "Library_ID") %>%
  arrange(Species, Lab) %>%
  group_by(Species,Lab) %>%
  summarize(Percent_presence = sum(Counts)/length(Counts)*100) %>%
  # select the pathways that are present in at least 1/3 of at least 2 groups
  ungroup() %>%
  # pathways present in > 1/3 of samples per lab
  filter(Percent_presence >= 33.33) %>%
  group_by(Species) %>%
  count() %>%
  filter(n > 1) %>%
  # pull out these pathways to keep
  select(Species)

```

```{r}


sp_all_pres33_filt_1 <- sp_all_filt %>%
  inner_join(., sp_all_filt_presence_more_30) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.sp_all_pres33_filt_1_pca <- tune.pca(sp_all_pres33_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_all_pres33_filt_1.pca <- pca(sp_all_pres33_filt_1, ncomp = 3, logratio = 'CLR')

Type_colors <- c("black","blue","turquoise", "red")
Type_shapes <- c(19,17,15,9)


# plot the PCA
sp_all_pres33_plot <- plot_pca(sp_all_pres33_filt_1.pca, "PC1", "PC2", "Lab", "Lab")
sp_all_pres33_plot
plot_pca(sp_all_pres33_filt_1.pca, "PC1", "PC2", "Island", "Lab")

# top 20 genera separating samples along PC1? (10+, 10-)
sp_all_pres33_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(sp_all_pres33_filt_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., sp_all_pres33_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(sp_all_pres33_filt_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r pca_cont_pres33_species}
# average sequence length
plot_pca_cont(sp_all_pres33_filt_1.pca, "PC1", "PC2", "seq_len_avg",  "Lab", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(sp_all_pres33_filt_1.pca, "PC1", "PC2", "gc_avg",  "Lab", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(sp_all_pres33_filt_1.pca, "PC1", "PC2", "No_reads_no_human",  "Lab", 3, "Total Reads")

# total number of reads
plot_pca_cont(sp_all_pres33_filt_1.pca, "PC1", "PC2", "Age_mean",  "Lab", 3, "Mean Age (BP)")

```


Run a PERMANOVA with adonis2 on the abundance-filtered clr-transformed species matrix.
```{r adonis_pres33_species}

metadata_all <- metadata %>%
    filter(Library_ID %in% sample_list)
  

sp_all_pres33_filt_1 <- sp_all_filt %>%
  inner_join(., sp_all_filt_presence_more_30) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.sp_all_pres33_filt_1_pca <- tune.pca(sp_all_pres33_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_all_pres33_filt_1.pca <- pca(sp_all_pres33_filt_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
sp_all_pres33_filt_1.euc <- vegdist(sp_all_pres33_filt_1.pca$X, method = "euclidean", na.rm = F)
sp_all_pres33_filt_1.euc.pcoa <- ape::pcoa(sp_all_pres33_filt_1.euc)

sp_all_pres33_filt_1.euc.pcoavectors <- as.data.frame(sp_all_pres33_filt_1.euc.pcoa$vectors)
sp_all_pres33_filt_1.euc.pcoavectors <- sp_all_pres33_filt_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

sp_all_pres33_filt_1.euc.vecmet <- left_join(sp_all_pres33_filt_1.euc.pcoavectors, metadata_filt, by = "Library_ID")

sp_all_pres33_filt_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Island, shape = Island)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Island_colors) +
  scale_shape_manual(values = Island_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Site_code")

# sp_all_pres33_filt_1 %>%
#   rownames_to_column("Library_ID") %>%
#   select(Library_ID) %>%
#   anti_join(., metadata %>% select(Library_ID, Anginosus), by  = "Library_ID")


# test for difference between different metadata categories 
adonis2(sp_all_pres33_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")

all_permanova <- adonis2(sp_all_pres33_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(., adonis2(sp_all_pres33_filt_1.euc ~ Island, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_pres33_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_pres33_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_pres33_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_pres33_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","***","***","***","ns"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)


# all together
adonis2(sp_all_pres33_filt_1.euc ~ Lab + Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(sp_all_pres33_filt_1.euc ~ Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)


```

```{r betadisper_noblanks_noyanomami_species}
# Test the 3 metadata categories that are significant (p < 0.05) based on adonis for significant differences in dispersion
# Industry, tooth site, village
# Ok, sometimes Village is not significantly different, it fluxuates around 0.05. Maybe better to leave it out as not significant

# Lab
groups_tooth_site <- metadata_all %>%
  select(Lab) %>% 
  pull()

## Calculate multivariate dispersions
lab_betadisper_noblanks_mr <- betadisper(sp_all_pres33_filt_1.euc, groups_tooth_site, bias.adjust =TRUE)
lab_betadisper_noblanks_mr

## Perform test
anova(lab_betadisper_noblanks_mr)

## Permutation test for F
permutest(lab_betadisper_noblanks_mr, pairwise = TRUE, permutations = 99)

## Tukey's Honest Significant Differences
lab_betadisper_noblanks_mr.HSD <- TukeyHSD(lab_betadisper_noblanks_mr)
plot(lab_betadisper_noblanks_mr.HSD)

## Plot the groups and distances to centroids on the
## first two PCoA axes
## with data ellipses instead of hulls
plot(lab_betadisper_noblanks_mr, ellipse = TRUE, hull = FALSE) # 1 sd data ellipse

## can also specify which axes to plot, ordering respected
# plot(lab_betadisper_noblanks_mr, axes = c(3,1), seg.col = "forestgreen", seg.lty = "dashed")

## Draw a boxplot of the distances to centroid for each group
boxplot(lab_betadisper_noblanks_mr)


```

```{r, eval = F}
# use this table for downstream analyses
fwrite(sp_all_pres33_filt_1, "./05-Documentation.backup/pacific_species_decontam_pres33_filt.tsv", quote = F, sep = "\t")
# fwrite(sp_all_pres33_filt_1, "../strep_clades/00-documentation/pacific_species_decontam_pres33_filt.tsv", quote = F, sep = "\t")

```


### Filter out taxa that are less than 0.005% relative abundane across all samples?
Try instead filtering out the taxa that have fewer than X number of reads
and see what percent that is as you ajust X
```{r, eval = F}
# let's look at the abundance of these taxa
sp_all_filt %>%
  as_tibble(.) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  select(Species, Percent) %>%
  arrange(desc(Percent)) %>%
  mutate(Percent = round(Percent, 4))

```


```{r all_plot_abund005_filt}

# Taxa present at < 0.005% relative abundance
sp_abund005_lab <- sp_all_filt %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent <= 0.005) %>%
  # select(Species, Total, Percent)
  select(Species)

sp_all_abund005_filt_1 <- sp_all_filt %>%
  anti_join(., sp_abund005_lab) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.sp_all_abund005_filt_1_pca <- tune.pca(sp_all_abund005_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_all_abund005_filt_1.pca <- pca(sp_all_abund005_filt_1, ncomp = 3, logratio = 'CLR')

Type_colors <- c("black","blue","turquoise", "red")
Type_shapes <- c(19,17,15,9)


# plot the PCA
sp_all_abund005_plot <- plot_pca(sp_all_abund005_filt_1.pca, "PC1", "PC2", "Lab", "Lab")
sp_all_abund005_plot

# top 20 genera separating samples along PC1? (10+, 10-)
sp_all_abund005_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(sp_all_abund005_filt_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., sp_all_abund005_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(sp_all_abund005_filt_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r pca_cont_abund005_species}
# average sequence length
plot_pca_cont(sp_all_abund005_filt_1.pca, "PC1", "PC2", "seq_len_avg",  "Lab", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(sp_all_abund005_filt_1.pca, "PC1", "PC2", "gc_avg",  "Lab", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(sp_all_abund005_filt_1.pca, "PC1", "PC2", "No_reads_no_human",  "Lab", 3, "Total Reads")

# total number of reads
plot_pca_cont(sp_all_abund005_filt_1.pca, "PC1", "PC2", "Age_mean",  "Lab", 3, "Mean Age (BP)")

```

Run a PERMANOVA with adonis2 on the abundance-filtered clr-transformed species matrix.
```{r adonis_abund005_species}

metadata_all <- metadata %>%
    filter(Library_ID %in% sample_list)
  

# Taxa present at < 0.005% relative abundance
sp_abund005_lab <- sp_all_filt %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent <= 0.005) %>%
  # select(Species, Total, Percent)
  select(Species)

sp_all_abund005_filt_1 <- sp_all_filt %>%
  anti_join(., sp_abund005_lab) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.sp_all_abund005_filt_1_pca <- tune.pca(sp_all_abund005_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_all_abund005_filt_1.pca <- pca(sp_all_abund005_filt_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
sp_all_abund005_filt_1.euc <- vegdist(sp_all_abund005_filt_1.pca$X, method = "euclidean", na.rm = F)
sp_all_abund005_filt_1.euc.pcoa <- ape::pcoa(sp_all_abund005_filt_1.euc)

sp_all_abund005_filt_1.euc.pcoavectors <- as.data.frame(sp_all_abund005_filt_1.euc.pcoa$vectors)
sp_all_abund005_filt_1.euc.pcoavectors <- sp_all_abund005_filt_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

sp_all_abund005_filt_1.euc.vecmet <- left_join(sp_all_abund005_filt_1.euc.pcoavectors, metadata_filt, by = "Library_ID")

sp_all_abund005_filt_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Island, shape = Island)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Island_colors) +
  scale_shape_manual(values = Island_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Site_code")

# sp_all_abund005_filt_1 %>%
#   rownames_to_column("Library_ID") %>%
#   select(Library_ID) %>%
#   anti_join(., metadata %>% select(Library_ID, Anginosus), by  = "Library_ID")


# test for difference between different metadata categories 
adonis2(sp_all_abund005_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_abund005_filt_1.euc ~ Island, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_abund005_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_abund005_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_abund005_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin")
adonis2(sp_all_abund005_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin")

all_permanova <- adonis2(sp_all_abund005_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(., adonis2(sp_all_abund005_filt_1.euc ~ Island, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_abund005_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_abund005_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_abund005_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_all_abund005_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","***","***","**","ns"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)

# all together
adonis2(sp_all_abund005_filt_1.euc ~ Lab + Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(sp_all_abund005_filt_1.euc ~ Island + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)

```

Remove species that are present at <0.005% relative abundance in each lab
```{r each_plot_abund005_filt}

# species present at <0.005% per lab
sp_abund005_each <- sp_all_filt %>%
  # Jena
  select(matches(c("Species",jena))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent <= 0.005) %>%
  select(Species) %>%
  full_join(., sp_all_filt %>%
            # Oklahoma
            select(matches(c("Species",oklahoma))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent <= 0.005) %>%
            select(Species)) %>%
  full_join(., sp_all_filt %>%
            # Otago
            select(matches(c("Species",otago))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent <= 0.005) %>%
            select(Species))


sp_each_abund005_filt_1 <- sp_all_filt %>%
  anti_join(., sp_abund005_each) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.sp_each_abund005_filt_1_pca <- tune.pca(sp_each_abund005_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_each_abund005_filt_1.pca <- pca(sp_each_abund005_filt_1, ncomp = 3, logratio = 'CLR')

# plot the PCA
sp_each_abund005_plot <- plot_pca(sp_each_abund005_filt_1.pca, "PC1", "PC2", "Lab", "Lab")
sp_each_abund005_plot
sp_each_abund005_plot_island <- plot_pca(sp_each_abund005_filt_1.pca, "PC1", "PC2", "Island", "Lab")
sp_each_abund005_plot_island

ggsave("./06-publication/prelim_figs/pacific_species_island_pca12.pdf", plot = sp_each_abund005_plot_island, device = "pdf",
       scale = 1, width = 8, height = 4, units = c("in"), dpi = 300)


# top 20 genera separating samples along PC1? (10+, 10-)
sp_each_abund005_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(sp_each_abund005_filt_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., sp_each_abund005_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(sp_each_abund005_filt_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

# top 20 genera separating samples along PC2? (10+, 10-)
sp_each_abund005_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(sp_each_abund005_filt_1.pca$loadings$X)) %>%
  select(Species, PC2) %>%
  arrange(PC2) %>%
  slice_max(PC2, n = 10) %>%
  bind_rows(., sp_each_abund005_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(sp_each_abund005_filt_1.pca$loadings$X)) %>%
            select(Species, PC2) %>%
            arrange(PC2) %>%
            slice_min(PC2, n = 10))

```

```{r pca_cont_abund005_each_species}
# average sequence length
plot_pca_cont(sp_each_abund005_filt_1.pca, "PC1", "PC2", "seq_len_avg",  "Lab", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(sp_each_abund005_filt_1.pca, "PC1", "PC2", "gc_avg",  "Lab", 3, "Average GC (%)")
plot_pca_cont(sp_each_abund005_filt_1.pca, "PC1", "PC3", "gc_avg",  "Lab", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(sp_each_abund005_filt_1.pca, "PC1", "PC2", "No_reads_no_human",  "Lab", 3, "Total Reads")

# total number of reads
plot_pca_cont(sp_each_abund005_filt_1.pca, "PC1", "PC2", "Age_mean",  "Lab", 3, "Mean Age (BP)")

```

Run a PERMANOVA with adonis2 on the abundance-filtered clr-transformed species matrix.
```{r adonis_abund005_species}

metadata_all <- metadata %>%
    filter(Library_ID %in% sample_list)
  

# species present at <0.005% per lab
sp_abund005_each <- sp_all_filt %>%
  # Jena
  select(matches(c("Species",jena))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent <= 0.005) %>%
  select(Species) %>%
  full_join(., sp_all_filt %>%
            # Oklahoma
            select(matches(c("Species",oklahoma))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent <= 0.005) %>%
            select(Species)) %>%
  full_join(., sp_all_filt %>%
            # Otago
            select(matches(c("Species",otago))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent <= 0.005) %>%
            select(Species))


sp_each_abund005_filt_1 <- sp_all_filt %>%
  anti_join(., sp_abund005_each) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")


# check the number of components to retain by tuning the PCA
tune.sp_each_abund005_filt_1_pca <- tune.pca(sp_each_abund005_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
sp_each_abund005_filt_1.pca <- pca(sp_each_abund005_filt_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
sp_each_abund005_filt_1.euc <- vegdist(sp_each_abund005_filt_1.pca$X, method = "euclidean", na.rm = F)
sp_each_abund005_filt_1.euc.pcoa <- ape::pcoa(sp_each_abund005_filt_1.euc)

sp_each_abund005_filt_1.euc.pcoavectors <- as.data.frame(sp_each_abund005_filt_1.euc.pcoa$vectors)
sp_each_abund005_filt_1.euc.pcoavectors <- sp_each_abund005_filt_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

sp_each_abund005_filt_1.euc.vecmet <- left_join(sp_each_abund005_filt_1.euc.pcoavectors, metadata_filt, by = "Library_ID")

sp_each_abund005_filt_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Island, shape = Island)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Island_colors) +
  scale_shape_manual(values = Island_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Site_code")

# sp_each_abund005_filt_1 %>%
#   rownames_to_column("Library_ID") %>%
#   select(Library_ID) %>%
#   anti_join(., metadata %>% select(Library_ID, Anginosus), by  = "Library_ID")


# test for difference between different metadata categories 
adonis2(sp_each_abund005_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")
each_permanova <- adonis2(sp_each_abund005_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
each_permanova  <- each_permanova %>% bind_rows(., adonis2(sp_each_abund005_filt_1.euc ~ Island, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_each_abund005_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_each_abund005_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_each_abund005_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(sp_each_abund005_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


each_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","**","**","***","ns"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/each_permanova.tsv", sep = "\t", quote =  F)

# all together
adonis2(sp_each_abund005_filt_1.euc ~ Lab + Island + Age_mean + gc_avg + seq_len_avg + No_reads_no_human, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(sp_each_abund005_filt_1.euc ~ Island + Age_mean + gc_avg + seq_len_avg + No_reads_no_human, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)

```


### Plot together
patchwork example
```{r}

sp_plot <- sp_all_plot + sp_all_1lab_plot + sp_all_pres33_plot + sp_all_abund005_plot +  sp_each_abund005_plot +
  plot_layout(nrow = 1)  + 
  plot_layout(guides = "collect")

sp_plot

```



Biplots
```{r pres33_filt_biplot}
# which paths distinguish sample types 
sp_unfilt_1_pc1 <- plot_pca_bi(sp_each_abund005_filt_1.pca, "PC1", "PC2", "Lab", PC1, GeneFamily = F)
sp_unfilt_1_pc1

# which paths distinguish sample types 
sp_unfilt_1_pc2 <- plot_pca_bi(sp_each_abund005_filt_1.pca, "PC1", "PC2", "Lab", PC2, GeneFamily = F)
sp_unfilt_1_pc2

```

## Canonical correlation analysis
```{r pca_species_all}

exp_var_all <- paste0(round(sp_each_abund005_filt_1.pca$prop_expl_var$X * 100, 2), "%")
all_pca_values <- sp_each_abund005_filt_1.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  full_join(metadata, by = "Library_ID") %>%
  full_join(., fread("./05-Documentation.backup/oceana_climate.tsv")) %>%
  full_join(., cfdp_fail %>% mutate(Preservation = "Poor"), by = "Library_ID") %>%
  mutate(Preservation = ifelse(is.na(Preservation), "Good", Preservation)) %>%
  mutate(Preservation = ifelse(Type == "blank","Blank",
                               ifelse(Type == "Arch. Bone", "Arch. Bone", Preservation))) %>%
  drop_na(PC1) %>%
  # now convert all non-numeric columns to numbers
  left_join(., metadata %>%
    select(Site) %>%
    unique() %>%
    mutate(Site_n = 1:n()), by = "Site") %>%
  left_join(., metadata %>%
    select(Island) %>%
    unique() %>%
    mutate(Island_n = 1:n()), by = "Island") %>%
  mutate(Lab_n = ifelse(Lab == "Jena",0,
                        ifelse(Lab == "Oklahoma",1,2))) 

```

```{r all_corrs}
# rename variables to look nice in figures
all_pca_values_select <- all_pca_values %>%
  select(Library_ID,PC1, PC2, PC3, seq_len_avg, gc_avg, No_reads_no_human,  Age_mean, Lab_n, Site_n, Island_n, Lat, Long, Rainfall, Temperature, ann_evapotransporation) %>%
  rename(`Seq. length` = seq_len_avg,
         GC = gc_avg,
         `Total reads` = No_reads_no_human,
         Lab = Lab_n,
         Site = Site_n,
         Island = Island_n,
         `Mean age (BP)` = Age_mean,
         Latitude = Lat,
         Longitude = Long,
         `Ann. evapotransporation` = ann_evapotransporation)

form <- ~ PC1 + PC2 + PC3 + `Seq. length` + GC + `Total reads` + `Mean age (BP)` + Lab + Site + Island + Latitude + Longitude  + Rainfall + Temperature + `Ann. evapotransporation`
C_all = canCorPairs(form, all_pca_values_select)
plotCorrMatrix( C_all )

```


```{r pearson_corr_mid}
# test for significant differences in diversity between the metadata categories

pearson_all <- all_pca_values %>%
  # remove columns to avoid duplicates in renaming
  select(-Island, -Site, -Lab)  %>%
  # rename to match the chunk above
  rename(`Seq. length` = seq_len_avg,
         GC = gc_avg,
         `Total reads` = No_reads_no_human,
         Lab = Lab_n,
         Site = Site_n,
         Island = Island_n,
         `Mean age (BP)` = Age_mean,
         Latitude = Lat,
         Longitude = Long,
         `Ann. evapotransporation` = ann_evapotransporation) %>%
  rstatix::cor_test(., PC1, PC2, PC3, "Seq. length", GC, "Total reads", "Mean age (BP)", Lab, Site, Island, Latitude, Longitude, Rainfall, Temperature, "Ann. evapotransporation", method="pearson")

pearson_all

pearson_all %>%
  mutate(same = var1 == var2) %>%
  filter(same == FALSE) %>%
  select(-same) %>%
  filter(cor > 0.40 | cor < -0.40) %>%
  filter(p <= 0.001) %>%
  distinct(statistic, .keep_all = TRUE) 
# %>%
#   arrange(var1,var2)
  
```

```{r}

C_all %>%
  as_data_frame(.) %>%
  mutate(names = rownames(C_all)) %>%
  select(names, everything()) %>%
  pivot_longer(!names, names_to = "Corrs", values_to = "values") %>%
  filter(values != 1) %>%
  filter(values >= 0.4) %>%
  distinct(values, .keep_all = TRUE)

```


```{r}
# colorset <- rev(colorspace::sequential_hcl(10, palette = "BuPu"))
# maybe get colors from here: https://colordesigner.io/

colorset <- c("#ffffff","#e4f4fe","#c8e9fd","#addefc","#92d3fa","#77c8f9","#5bbdf8","#40b2f7","#25a8f6","#0a9cf4","#098BD9") # ,"#087dc3","#076fae","#066198","#055382"

 pvals_all <- all_pca_values_select %>%
   drop_na(Library_ID) %>%
  column_to_rownames("Library_ID") %>%
  as.matrix(.) %>%
  corrplot::cor.mtest(., method="pearson")

C_all %>%
  corrplot::corrplot(., p.mat = pvals_all$p, order = "original", type = "upper", col = colorset, is.corr = FALSE, cl.cex = 1.0, diag = FALSE, tl.cex = 1.0,
                     tl.col = 'black', sig.level = c(0.001), pch.cex = 2.2, insig = 'label_sig', pch.col = 'grey20')

```

```{r, eval = F}
svg(file = "./06-publication/prelim_figs/taxonomy_corr_plot.svg") 

C_all %>%
  corrplot::corrplot(., p.mat = pvals_all$p, order = "original", type = "upper", col = colorset, is.corr = FALSE, cl.cex = 1.5, diag = FALSE, tl.cex = 1.7,
                     tl.col = 'black', sig.level = c(0.001), pch.cex = 2.2, insig = 'label_sig', pch.col = 'grey20')

dev.off()

```








