---
title: "Pacific calculus taxonomy compared to European ancient dental calculus"
author: "Irina Velsko"
date: "`r format(Sys.time(), '%b %d, %Y')`"
output:
    github_document:
        fig_width: 8
        fig_height: 6
        toc: true
---


```{r load_libraries, echo = F, message = F}
library(knitr)
library(data.table)
library(mixOmics)
library(janitor)
library(pander)
library(vegan)
library(tidyverse)
library(ggrepel)
library(viridis)
library(patchwork)
library(RColorBrewer)
library(microshades)
opts_chunk$set(echo=F, warning=F, message=F, dpi=150)
```

Set the working directory of the notebook to the parent directory of the 
notebook itself.  

```{r setup}
knitr::opts_knit$set(root.dir = normalizePath(".."))
```


```{r colors}
Island_colors = c("Tongatapu" = "#616161", 
                  "Flores" = "#f6865e", 
                  "Rapa Nui" = "#cc79a7", 
                  "Futuna" = "#7D3560", 
                  "Efate" = "#7472af", 
                  "Vao" = "#098BD9",
                  "Efate 3000 BP" = "#7DCCFF", 
                  "Uripiv" = "#e1f2fe", 
                  "Viti Levu" = "#1ec1a7", 
                  "Raiatea" = "#148f77", 
                  "Taumako" = "#4e7705", 
                  "Watom" = "#86c73a")


type_cols <- c("Saliva - Industrial" = "#AFDFEF", "Buccal mucosa - Industrial" = "#4e7705", "Plaque - Industrial" = "#FC6882", "Plaque - Non-industrial" = "#C70E7B", "Saliva - Non-industrial" = "#54BCD1", "Buccal mucosa - Non-industrial" = "#8FDA04", "Calculus - Industrial" = "#F4E3C7")


# Set microshades colors
Lab_colors = c("Jena" = microshades_palette("micro_cvd_purple", 3, lightest = FALSE)[3],
                "Oklahoma" =  microshades_palette("micro_cvd_blue", 3, lightest = FALSE)[3],
                "Otago" = microshades_palette("micro_cvd_green", 3, lightest = FALSE)[3],
                "Oxford" = microshades_palette("micro_cvd_turquoise", 3, lightest = FALSE)[1],
                "Tuebingen" = microshades_palette("micro_cvd_purple", 3, lightest = FALSE)[1],
                "ACAD" = microshades_palette("micro_cvd_blue", 4, lightest = FALSE)[1],
                "Vienna" = microshades_palette("micro_cvd_green", 3, lightest = FALSE)[1])

Study_colors <- c("Pacific" = "#098BD9", 
                  "Middenbeemster" = "#7472af", 
                  "Radcliffe" = "#cc79a7", 
                  "Kilteasheen" = "#f6865e", 
                  "Iberia" = "#1ec1a7",
                  "Eisenhoffer2020" = "#86c73a",
                  "Ottoni2021" = "#616161",
                  "FellowsYates2021" = "#e1f2fe",
                  "Mann2018" = "#f5c710ff") # 7D3560

Continent_colors <- c("Pacific" = "#098BD9", 
                  "Europe" = "#7472af", 
                  "Africa" = "#cc79a7", 
                  "Caribbean" = "#f6865e", 
                  "Asia" = "#86c73a",
                  "North America" = "#1ec1a7")

```

```{r}
Island_shapes = c("Tongatapu" = 21, 
                  "Flores" = 21, 
                  "Rapa Nui" = 21, 
                  "Futuna" = 21, 
                  "Efate" = 21, 
                  "Vao" = 21,
                  "Efate 3000 BP" = 21, 
                  "Uripiv" = 21, 
                  "Viti Levu" = 21, 
                  "Raiatea" = 21, 
                  "Taumako" = 21, 
                  "Watom" = 21)

Lab_shapes = c(21,21,21,21,21,21,21)
Lab_size = c(4,4,4,4,4,4,4)

Type_shapes = c(21,21,21,21,21,21,21,21)
Type_size = c(4,4,4,4,4,4,4,4)

```


```{r plot_pca_fxn}

# add functions for:
# making the input table for PCA?

# plotting PCA with colored dots
plot_pca <- function(df, pc1, pc2, color_group, shape_group) {
    metadata_group_colors <- get(paste(color_group, "_colors", sep = ""))
    metadata_group_shapes <- get(paste(shape_group, "_shapes", sep = ""))

    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%")
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(metadata, by = "Library_ID")
    
    color_group = df_X[[color_group]]
    shape_group = df_X[[shape_group]]

                       if (pc1 == 'PC1') {
                           pc1 <- df_X$PC1
                           exp_var_pc1 <- exp_var[1]
                           xaxis <- c("PC1")
                       }  else if (pc1 == 'PC2') {
                           pc1 <- df_X$PC2
                           exp_var_pc1 <- exp_var[2]
                           xaxis <- c("PC2")
                       } else if (pc1 == 'PC3') {
                           pc1 <- df_X$PC3
                           exp_var_pc1 <- exp_var[3]
                           xaxis <- c("PC3")
                       }

                       if (pc2 == 'PC1') {
                           pc2 <- df_X$PC1
                           exp_var_pc2 <- exp_var[1]
                           yaxis <- c("PC1")
                       }  else if (pc2 == 'PC2') {
                           pc2 <- df_X$PC2
                           exp_var_pc2 <- exp_var[2]
                           yaxis <- c("PC2")
                       } else if (pc2 == 'PC3') {
                           pc2 <- df_X$PC3
                           exp_var_pc2 <- exp_var[3]
                           yaxis <- c("PC3")
                       }

    pca_plot <- ggplot(df_X, aes(pc1, pc2)) +
                        geom_point(aes(fill = color_group, shape = shape_group), size = 2, stroke = 0.3, shape = 21) +
                        scale_fill_manual(values = metadata_group_colors) +
                        xlab(paste(xaxis, " - ", exp_var_pc1)) +
                        ylab(paste(yaxis, " - ", exp_var_pc2)) +
                        theme_minimal(base_size = 14) +
                        theme(legend.position = "right") +
                        theme(legend.title = element_blank())

    return(pca_plot)
}

```


```{r plot_bi}
# this function plots the PCA with biplots to look at the categories with strongest loadings
plot_pca_bi <- function(df, pc1, pc2, metadata_group, columntitle) {
    metadata_group_colors <- get(paste(metadata_group, "_colors", sep = ""))
    metadata_group_shapes <- get(paste(metadata_group, "_shapes", sep = ""))
   
    arrow_pc <- enquo(columntitle)
    
    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%") # explained variance for x- and y-labels
    
    # select only the PCs from the PCA and add metadata
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(metadata, by = "Library_ID")
    
    metadata_group = df_X[[metadata_group]]
    
    corr_lam <- df$sdev[c("PC1", "PC2", "PC3")] * sqrt(nrow(df_X))

    df_X <- df_X %>%
      mutate(PC1 = PC1 / corr_lam[1],
             PC2 = PC2 / corr_lam[2], 
             PC3 = PC3 / corr_lam[3])

    # select the correct PC column and explained variance for PC1
    if (pc1 == 'PC1') {
        Pc1 <- df_X$PC1
        exp_var_pc1 <- exp_var[1]
        xaxis <- c("PC1")
    } else if (pc1 == 'PC2') {
        Pc1 <- df_X$PC2
        exp_var_pc1 <- exp_var[2]
        xaxis <- c("PC2")
    } else if (pc1 == 'PC3') {
        Pc1 <- df_X$PC3
        exp_var_pc1 <- exp_var[3]
        xaxis <- c("PC3")
   }
    
    # select the correct PC column and explained variance for PC2
    if (pc2 == 'PC1') {
        Pc2 <- df_X$PC1
        exp_var_pc2 <- exp_var[1]
        yaxis <- c("PC1")
 } else if (pc2 == 'PC2') {
       Pc2 <- df_X$PC2
       exp_var_pc2 <- exp_var[2]
       yaxis <- c("PC2")
    } else if (pc2 == 'PC3') {
       Pc2 <- df_X$PC3
       exp_var_pc2 <- exp_var[3]
       yaxis <- c("PC3")
   }

    # Identify the 10 taxa that have highest positive and negative loadings in the selected PC
    pws_10 <- df$loadings$X %>%
      as.data.frame(.) %>%
      rownames_to_column(var = "Function") %>%
      top_n(10, !!arrow_pc)

    neg_10 <- df$loadings$X %>%
      as.data.frame(.) %>%
      rownames_to_column(var = "Function") %>%
      top_n(-10, !!arrow_pc)


    pca_plot_bi <- ggplot(df_X, aes(x = Pc1, y = Pc2)) +
      geom_point(aes(shape = metadata_group, fill = metadata_group), size = 3.5, stroke = 0.3) +
      geom_segment(data = pws_10,
                   aes(xend = get(paste(pc1)), yend = get(paste(pc2))),
                   x = 0, y = 0, colour = "black",
                   linewidth = 0.5,
                   arrow = arrow(length = unit(0.03, "npc"))) +
      geom_label_repel(data = pws_10,
                   aes(x = get(paste(pc1)), y = get(paste(pc2)), label = Function),
                   size = 2.5, colour = "grey20", label.padding = 0.2, max.overlaps = 18) +
      geom_segment(data = neg_10,
                   aes(xend = get(paste(pc1)), yend = get(paste(pc2))),
                   x = 0, y = 0, colour = "grey50",
                   linewidth = 0.5,
                   arrow = arrow(length = unit(0.03, "npc"))) +
      geom_label_repel(data = neg_10,
                   aes(x = get(paste(pc1)), y = get(paste(pc2)), label = Function),
                   size = 2.5, colour = "grey20", label.padding = 0.2, max.overlaps = 18) +
      labs(x = paste(xaxis, " - ", exp_var_pc1),
           y = paste(yaxis, " - ", exp_var_pc2)) +
      scale_fill_manual(values = metadata_group_colors) +
      scale_shape_manual(values = metadata_group_shapes) +
      theme_minimal() + theme(text = element_text(size = 14)) +
      theme(legend.position = "right")

    return(pca_plot_bi)
}


```

```{r pca_plot_cont_fxn}
# for continuous data
plot_pca_cont <- function(df_pca, pc1, pc2, color_group, ncomps, title_text) {

    exp_var <- paste0(round(df_pca$prop_expl_var$X * 100, 2), "%")
    df_X <- df_pca$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(metadata, by = "Library_ID")

    color_group = df_X[[color_group]]

    if (pc1 == 'PC1') {
        pc1 <- df_X$PC1
        exp_var_pc1 <- exp_var[1]
        xaxis <- c("PC1")
    }  else if (pc1 == 'PC2') {
        pc1 <- df_X$PC2
        exp_var_pc1 <- exp_var[2]
        xaxis <- c("PC2")
    } else if (pc1 == 'PC3') {
        pc1 <- df_X$PC3
        exp_var_pc1 <- exp_var[3]
        xaxis <- c("PC3")
    }

    if (pc2 == 'PC1') {
        pc2 <- df_X$PC1
        exp_var_pc2 <- exp_var[1]
        yaxis <- c("PC1")
    }  else if (pc2 == 'PC2') {
        pc2 <- df_X$PC2
        exp_var_pc2 <- exp_var[2]
        yaxis <- c("PC2")
    } else if (pc2 == 'PC3') {
        pc2 <- df_X$PC3
        exp_var_pc2 <- exp_var[3]
        yaxis <- c("PC3")
    }

    pca_plot <- ggplot(df_X, aes(x = pc1, y = pc2)) +
      geom_point(aes(fill = color_group), size = 3.5, stroke = 0.3, shape = 21) +
     scale_fill_viridis_c(option = "C") +
     # scale_shape_manual(values = c(16, 16, 16, 16, 16,16,16,16)) +
     # scale_shape_manual(values = c(16, 17, 15, 18, 14,9,13,24)) +
    # stat_ellipse() +
     xlab(paste(xaxis, " - ", exp_var_pc1)) +
     ylab(paste(yaxis, " - ", exp_var_pc2)) +
     theme_minimal(base_size = 16) +
     theme(text = element_text(size=16)) +
     theme(legend.title = element_blank(),
           legend.key.size = unit(2,"mm"),
           legend.text = element_text(size = 10)) + 
     theme(legend.position = "right") +
     ggtitle(title_text) + theme(plot.title = element_text(size = 10))

    return(pca_plot)
}

```

## Prep Pacific data
Load Pacific contaminant tables 
```{r load_contam_lists}
# genus lists
cont_otago_gn <- fread("./05-Documentation.backup/contaminant_otago_gn.tsv")
cont_oklahoma_gn <- fread("./05-Documentation.backup/contaminant_oklahoma_gn.tsv")
cont_jena_gn <- fread("./05-Documentation.backup/contaminant_jena_gn.tsv")

cont_all_gn <- fread("./05-Documentation.backup/contaminant_pacific_ooj_gn.tsv")

# species lists
cont_otago_sp <- fread("./05-Documentation.backup/contaminant_otago_sp.tsv", sep = "\t")
cont_oklahoma_sp <- fread("./05-Documentation.backup/contaminant_oklahoma_sp.tsv", sep = "\t")
cont_jena_sp <- fread("./05-Documentation.backup/contaminant_jena_sp.tsv", sep = "\t")

cont_all_sp <- fread("./05-Documentation.backup/contaminant_pacific_ooj_sp.tsv", sep = "\t")

```

Load Pacific metadata
```{r}
pacific_metadata <- fread("05-Documentation.backup/oceania_metadata_full.tsv") %>%
  select(Library_ID, Lab, Type, Age_mean, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = "Pacific",
         Continent = "Pacific",
         Age_mean_log = log10(Age_mean))

```

Pacific poorly preserved samples
```{r poorly_preserved}
cfdp_fail <- fread("05-Documentation.backup/oceania_cuperdec_species_poor_samples.tsv") %>%
  rename(Library_ID = 1)

```


```{r pacific_species_tables}
# read in the species table
pacific_rs_raw_sp <- fread("./05-Documentation.backup/oceania_malt_cRefSeq_summarized_species_20201105.txt") %>%
  rename(Species = 1) %>%
  # remove HPD samples
  select(-matches("HPD"))  %>%
  # full_join(., fread("./05-Documentation.backup/sources_MALT_cRefSeq_JFY_species_summarized.txt") %>%
  #             rename(Species = 1), by = "Species") %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_blanks_cRefSeq_species_summarized_20201130.txt") %>%
              rename(Species = 1), by = "Species") %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_bones_cRefSeq_species_summarized_20201202.txt") %>%
              rename(Species = 1), by = "Species") %>%
  filter(!str_detect(Species, "Homo"))

# clean the file names
colnames(pacific_rs_raw_sp) <- gsub("MeganServer::", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub(".unmapped", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub(".SG1.1", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub(".SG1.2", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_L001_R1_001.fastq.combined.fq.prefixed.extractunmapped.bam", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg19unmapped", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_S0_L001_R1_001.fastq.truncated.prefixed.extractunmapped.bam", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_S0_L001_R1_000.fastq.merged.prefixed.hg1", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg1", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub(".SG1.2_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(pacific_rs_raw_sp))  
colnames(pacific_rs_raw_sp) <- gsub("_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(pacific_rs_raw_sp)) 

# now do a pretend transpose to replace the NAs with 0s
pacific_rs_raw_sp <- pacific_rs_raw_sp %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = replace_na(Counts, 0)) %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts")

```


```{r pacific_outliers}
# HCLVMBCX2-3505-13-00-01_S13 appears to be a sample, although it's supposed to be a blank
# remove it from all further analyses

metadata_filt <- pacific_metadata %>%
  anti_join(., cfdp_fail) %>%
  filter(!str_detect(Library_ID, "01_S13"))

pacific_species_raw_pass <- pacific_rs_raw_sp %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  pivot_wider(names_from = "Species", values_from = "Counts") %>%
  anti_join(., cfdp_fail) %>%
  # remove blanks
  filter(!str_detect(Library_ID, "01_S13")) %>%
  # remove all taxa that have to entries anymore
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  adorn_totals(where = "col") %>%
  select(Total, everything()) %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  pivot_wider(names_from = "Species", values_from = "Counts")

```

```{r}
# list all bones that are part of this study
# the ARS bones are a little different in the PCA, so we won't use them
bones <- pacific_metadata %>%
  filter(Type == "bone") %>%
  select(Library_ID, Lab, Type) %>%
  filter(!str_detect(Library_ID, "ARS|HPD"))

# list all samples and blanks from Otago
otago <- pacific_metadata %>%
  filter(Lab == "Otago") %>%
  select(Library_ID, Lab, Type) %>%
  filter(Library_ID != "HCLVMBCX2-3505-13-00-01_S13") %>%
  arrange(Library_ID) %>%
  filter(Type != "blank") %>%
  pull(Library_ID)

# list all sampmles and blanks from Oklahoma
oklahoma <- pacific_metadata %>%
  filter(Lab == "Oklahoma") %>%
  select(Library_ID, Lab, Type) %>%
  filter(Type != "bone") %>%
  filter(Type != "blank") %>%
  pull(Library_ID)

# list all samples and blanks from Jena
jena <- pacific_metadata %>%
  filter(Lab == "Jena") %>%
  select(Library_ID, Lab, Type) %>%
  filter(Type != "bone") %>%
  filter(Type != "blank") %>%
  filter(!str_detect(Library_ID, "HPD")) %>%
  pull(Library_ID)

# list all samples to include
pacific_sample_list <- c(otago,oklahoma,jena)

```

```{r pacific_decontam_all_table}
# note that all poorly-preserved samples are removed from species_raw_pass

# select all samples and remove all contaminants
# save row names
rownames_sp_all_filt <- pacific_species_raw_pass %>%
  filter(Library_ID %in% pacific_sample_list) %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  anti_join(., cont_all_sp %>%
              select(Species), by = "Species") %>%
  adorn_totals(where = "col") %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  pull(Species)

pacific_sp_all_filt <- pacific_species_raw_pass %>%
  filter(Library_ID %in% pacific_sample_list) %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  anti_join(., cont_all_sp %>%
              select(Species), by = "Species") %>%
  adorn_totals(where = "col") %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  # remove the totals attribute so can make column sums later if necessary by converting to matrix and back
  as.matrix(., rownames = "Species") %>%
  as_tibble(.) %>%
  mutate(Species = rownames_sp_all_filt) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = as.numeric(Counts)) %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts")

# remove the totals attribute so can make column sums later if necessary
# attr(sp_all_filt, "totals") <- NULL

```


```{r individual_tables}
# note that all poorly-preserved samples are removed from species_raw_pass although they're still in the lab lists

# select only Otago samples
otago_sp <- pacific_species_raw_pass %>%
  filter(Library_ID %in% otago) %>%
  anti_join(., cfdp_fail) %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  anti_join(., cont_otago_sp, by = "Species") %>%
  adorn_totals(where = "col") %>%
  filter(Total != 0) %>%
  select(-Total)

# select only Oklahoma samples
oklahoma_sp <- pacific_species_raw_pass %>%
  filter(Library_ID %in% oklahoma) %>%
  anti_join(., cfdp_fail) %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  anti_join(., cont_oklahoma_sp, by = "Species") %>%
  adorn_totals(where = "col") %>%
  filter(Total != 0) %>%
  select(-Total)

# select only Jena samples
jena_sp <- pacific_species_raw_pass %>%
  filter(Library_ID %in% jena) %>%
  anti_join(., cfdp_fail) %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  anti_join(., cont_jena_sp, by = "Species") %>%
  adorn_totals(where = "col") %>%
  filter(Total != 0) %>%
  select(-Total)

```

### Prep World data
Load European contaminant tables - note these are only at the Species level (didn't do genus-level analysis)
```{r}
# contaminants
mid_contaminants <- fread("../smoking_calculus/05-results.backup/contaminant_species_mid.tsv", sep = "\t")
iber_contaminants <- fread("../smoking_calculus/05-results.backup/contaminant_species_iber.tsv", sep = "\t")
rad_contaminants <- fread("../smoking_calculus/05-results.backup/contaminant_species_rad.tsv", sep = "\t")
kil_contaminants_freq <- fread("../smoking_calculus/05-results.backup/contaminant_species_freq_kil.tsv", sep = "\t")
ei_contaminants <- fread("./05-Documentation.backup/contaminant_eisenhoffer_sp.tsv", sep = "\t")
ot_contaminants <- fread("./05-Documentation.backup/contaminant_ottoni_sp.tsv", sep = "\t")
de_contaminants <- fread("./05-Documentation.backup/contaminant_deepevo_sp.tsv", sep = "\t")
mw_contaminants <- fread("./05-Documentation.backup/contaminant_mw2018_sp.tsv")

```

```{r}
# poorly-preserved samples
world_fail <- fread("./05-Documentation.backup/cuperdec_malt_ei_ot_de_poor_samples.tsv")


```


```{r world_tables}
# MALT tables
mid_raw <- fread("../smoking_calculus/05-results.backup/MID_malt_bactarchhomo2018_comparison-species.txt")
radcliffe_raw <- fread("../smoking_calculus/05-results.backup/Radcliffe_pfuT_comparison-species.txt")
kilteasheen_raw <- fread("../smoking_calculus/05-results.backup/Kilteasheen_comparison-species.txt")
iberian_raw <- fread("../smoking_calculus/05-results.backup/iberian_medieval_malt_species_raw.tsv")
# mod_raw <- fread("../smoking_calculus/05-results.backup/mod_malt_species_raw.tsv")
ei2020_raw <- fread("./05-Documentation.backup/eisenhoffer2020_malt_refseq_species.tsv")
ot2021_raw <- fread("./05-Documentation.backup/ottoni2021_malt_refseq_species.tsv")
de_raw <- fread("./05-Documentation.backup/deep_evo_malt_refseq_comparison_species.tsv")
mw_raw <- fread("./05-Documentation.backup/mann2018_world_malt_refseq_comparison_species.tsv")

```

```{r}
metadataW <- fread("../smoking_calculus/00-documentation.backup/full_combined_metadata.tsv") %>%
  select(Library_ID, Site_code, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = str_replace_all(Site_code, "MID","Middenbeemster"),
         Study = str_replace_all(Study, "CMB","Middenbeemster"),
         Study = str_replace_all(Study, "KIL","Kilteasheen"),
         Study = str_replace_all(Study, "RAD","Radcliffe"),
         Study = str_replace_all(Study, "ELR","Iberia"),
         Study = str_replace_all(Study, "IVE","Iberia"),
         Lab = str_replace_all(Site_code, "MID","Jena"),
         Lab = str_replace_all(Lab, "CMB","Jena"),
         Lab = str_replace_all(Lab, "ELR","Jena"),
         Lab = str_replace_all(Lab, "IVE","Jena"),
         Lab = str_replace_all(Lab, "RAD","Oxford"),
         Lab = str_replace_all(Lab, "KIL","Tuebingen")) %>%
  mutate(Age_mean = ifelse(str_detect(Site_code, "MID|CMB|RAD"), 250,
                           ifelse(Site_code == "KIL", 700,
                                  ifelse(str_detect(Site_code, "ELR|IVE"), 620, 7)))) %>%
  mutate(Continent = "Europe") %>%
  full_join(., fread("./05-Documentation.backup/eisenhoffer2020_ottoni2021_deepevo_mann2018w_metadata.tsv") %>%
              mutate(Run_accession = ifelse(Study == "FellowsYates2021",Sample_alias,Run_accession)) %>%
              rename(Library_ID = Run_accession)) %>%
  mutate(Sample_type = ifelse(str_detect(Study, "EXB|LIB|DEN|SOL"),"blank",Sample_type)) %>%
  mutate(Sample_type = ifelse(is.na(Sample_type),"calculus",Sample_type)) %>%
  mutate(Sample_group = ifelse(Sample_type == "blank","blank",
                               ifelse(str_detect(Library_ID, "VLC|JAE"),"modern_calculus",Sample_group))) %>%
  mutate(Sample_group = ifelse(is.na(Sample_group),"ancient_calculus",Sample_group)) %>%
  mutate(Sample_alias = ifelse(is.na(Sample_alias),Library_ID,Sample_alias)) %>%
  mutate(Age_mean_log = log10(Age_mean))

world_metadata <- metadataW %>%
  # mutate(Sample_group = ifelse(is.na(Sample_group),"bone",Sample_group)) %>%
  filter(!str_detect(Sample_group, "primate|modern"))

# list the non-human primates and modern calculus to remove from the species table
nhp <- metadataW %>%
  filter(str_detect(Sample_group, "primate|modern")) %>%
  select(Sample_alias) %>%
  unique() %>%
  pull()

```

```{r world_studies}

fy <- world_metadata %>%
  anti_join(., world_fail %>% rename(Library_ID = Sample)) %>%
  filter(!str_detect(Sample_type, "blank")) %>%
  filter(Study == "FellowsYates2021") %>%
  select(Library_ID) %>%
  pull()

ot <- world_metadata %>%
  anti_join(., world_fail %>% rename(Library_ID = Sample)) %>%
  filter(!str_detect(Sample_type, "blank")) %>%
  filter(Study == "Ottoni2021") %>%
  select(Library_ID) %>%
  pull()

ei <- world_metadata %>%
  anti_join(., world_fail %>% rename(Library_ID = Sample)) %>%
  filter(!str_detect(Sample_type, "blank")) %>%
  filter(Study == "Eisenhoffer2020") %>%
  select(Library_ID) %>%
  pull()

ma <- world_metadata %>%
  filter(Study == "Kilteasheen") %>%
  select(Library_ID) %>%
  pull()

mw <- world_metadata %>%
  filter(Study == "Mann2018_world") %>%
  select(Library_ID) %>%
  pull()

ve19 <- world_metadata %>%
  filter(!str_detect(Sample_type, "blank")) %>%
  filter(Study == "Radcliffe") %>%
  select(Library_ID) %>%
  pull()

ve22 <- world_metadata %>%
  filter(!str_detect(Sample_type, "blank")) %>%
  filter(str_detect(Study, "Middenbeemster|Iberia")) %>%
  select(Library_ID) %>%
  pull()

```


```{r world_decontam_individual}
# remove contaminant species from MID/CMD
world_species_indiv_decontam <- mid_raw %>%
  rename(Species = `#Datasets`) %>%
  select(-matches("EXB|LIB")) %>%
  anti_join(., mid_contaminants) %>%
  full_join(., iberian_raw %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks
            select(-matches("EXB|LIB")) %>%
            anti_join(., iber_contaminants), by = "Species") %>%
  full_join(., radcliffe_raw %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks
            select(-`CSD.unmapped`, -`CSL.unmapped`, -`CSN.unmapped`, -`CSS.unmapped`) %>%
            anti_join(., rad_contaminants), by = "Species") %>%
  full_join(., kilteasheen_raw %>%
            rename(Species = `#Datasets`) %>%
            anti_join(., kil_contaminants_freq), by = "Species") %>%
  full_join(., mw_raw %>%
            rename(Species = `#Datasets`) %>%
            anti_join(., mw_contaminants), by = "Species") %>%
  full_join(., ei2020_raw %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks
            select(-`SRR11176637.unmapped`, -`SRR11176643.unmapped`) %>%
            anti_join(., ei_contaminants), by = "Species") %>%
  full_join(., ot2021_raw %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks
            select(-`ERR5729658.unmapped`, -`ERR5729659.unmapped`, -`ERR5729660.unmapped`, -`ERR5729661.unmapped`) %>%
            anti_join(., ot_contaminants), by = "Species") %>%
  full_join(., de_raw %>%
            rename(Species = `#Datasets`) %>%
            anti_join(., de_contaminants) %>%
            select(-matches(nhp %>% str_c(collapse = "|"))), by = "Species") %>%
  filter(Species != "Homo sapiens")

  
  

# remove contaminant species from Iberian medieval samples
# iberian_species_decontam <- iberian_raw %>%
#   rename(Species = `#Datasets`) %>%
#   anti_join(., iber_contaminants)

# remove contaminant species from Iberian medieval samples
# radcliffe_species_decontam <- radcliffe_raw %>%
#   rename(Species = `#Datasets`) %>%
#   anti_join(., rad_contaminants)

# remove contaminant species from JAE/VLC
# mod_species_decontam <- mod_raw %>%
#   rename(Species = `#Datasets`) %>%
#   anti_join(., rad_contaminants)

# remove contaminant species from Kilteasheen
# kilteasheen_species_decontam_freq <- kilteasheen_raw %>%
#   rename(Species = `#Datasets`) %>%
#   anti_join(., kil_contaminants_freq)

# world_species_indiv_decontam <- mid_species_decontam %>%
#   full_join(., radcliffe_species_decontam) %>%
#   full_join(., kilteasheen_species_decontam_freq) %>%
#   full_join(., iberian_species_decontam) 
# %>%
#   full_join(., mod_species_decontam)

# clean the file names
colnames(world_species_indiv_decontam) <- gsub(".unmapped","", colnames(world_species_indiv_decontam))
colnames(world_species_indiv_decontam) <- gsub("MeganServer::","", colnames(world_species_indiv_decontam))

# remove human and unassigned reads, convert NA to 0, and remove the 8 MID samples with no metadata
world_species_indiv_decontam <- world_species_indiv_decontam %>% 
  filter(!str_detect(Species, "Homo sapiens|Not assigned")) %>%
  replace(is.na(.), 0) %>%
  # remove poorly-preserved samples
  select(-matches(world_fail %>% pull(Sample) %>% str_c(collapse = "|"))) %>%
  # remove blanks, and soil and dentin
  select(-matches("CSS|CSD|EXB|LIB|CSN|CSL")) %>%
  # remove any taxa that are not present in the remaining samples
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total) %>%
  arrange(Species)


colnames(world_species_indiv_decontam) %>%
  as.data.frame(.) %>%
  rename(guess = 1) %>%
  filter(guess == "SRR6877282")

```

```{r world_all_decontam}
# remove all contaminants from all samples
world_species_all_decontam <- mid_raw %>%
  select(-matches("EXB|LIB")) %>%
  full_join(., radcliffe_raw %>%
            # remove blanks
            select(-`CSD.unmapped`, -`CSL.unmapped`, -`CSN.unmapped`, -`CSS.unmapped`)) %>%
  full_join(., kilteasheen_raw) %>%
  full_join(., mw_raw) %>%
  full_join(., iberian_raw %>%
              # remove blanks
              select(-matches("EXB|LIB"))) %>%
  full_join(., ei2020_raw %>%
              # remove blanks
              select(-`SRR11176637.unmapped`, -`SRR11176643.unmapped`)) %>%
  full_join(., ot2021_raw %>%
              # remove blanks
              select(-`ERR5729658.unmapped`, -`ERR5729659.unmapped`, -`ERR5729660.unmapped`, -`ERR5729661.unmapped`)) %>%
  full_join(., de_raw %>%
              select(-matches(nhp %>% str_c(collapse = "|")))) %>%
  rename(Species = `#Datasets`) %>%
  # remove blanks, dentin and soil
  select(-matches("CSS|CSD|EXB|LIB|CSN|CSL")) %>%
  # decontam all species
  anti_join(.,  mid_contaminants %>%
            bind_rows(., rad_contaminants) %>%
            bind_rows(., kil_contaminants_freq) %>%
            bind_rows(., iber_contaminants) %>%
            bind_rows(., ei_contaminants) %>%
            bind_rows(., ot_contaminants) %>%
            bind_rows(., mw_contaminants) %>%
            bind_rows(., de_contaminants) %>%
            distinct(), by = "Species") %>%
  filter(!str_detect(Species, "Homo sapiens|Not assigned")) %>%
  replace(is.na(.), 0) %>%
  # remove any species that have no counts after removing the blanks
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)
# %>%
#   full_join(., mod_raw)


# clean the file names
colnames(world_species_all_decontam) <- gsub(".unmapped","", colnames(world_species_all_decontam))
colnames(world_species_all_decontam) <- gsub("MeganServer::","", colnames(world_species_all_decontam))

world_species_all_decontam <- world_species_all_decontam %>%
  # remove poorly-preserved samples
  select(-matches(world_fail %>% pull(Sample) %>% str_c(collapse = "|")))

```

### Pacific and Euroepan tables together
```{r}
# read in MALT stats and add to the full table
malt_stats <- fread("./05-Documentation.backup/malt_stats_table_pac_world.tsv") %>%
  mutate(Pct_assigned = Assigned / (Mapped + Non_mapped) * 100)


metadata <- pacific_metadata %>%
  full_join(., world_metadata) %>%
  full_join(., malt_stats %>%
              mutate(Library_ID = str_replace_all(Library_ID, ".SG1","")), by = "Library_ID") %>%
  drop_na(Lab)

```


```{r pacific_world_all_decontam}
# combine raw tables and decontaminate with a list of all contaminants in all samples
pe_all_table <- pacific_species_raw_pass %>%
  filter(Library_ID %in% pacific_sample_list) %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  full_join(., mid_raw %>%
            # remove blanks
            select(-matches("EXB|LIB")) %>%
            full_join(., radcliffe_raw %>%
                      # remove blanks
                      select(-`CSD.unmapped`, -`CSL.unmapped`, -`CSN.unmapped`, -`CSS.unmapped`)) %>%
            full_join(., kilteasheen_raw) %>%
            full_join(., mw_raw) %>%
            full_join(., iberian_raw %>%
                      # remove blanks
                      select(-matches("EXB|LIB"))) %>%
            full_join(., ei2020_raw %>%
                      # remove blanks
                      select(-`SRR11176637.unmapped`, -`SRR11176643.unmapped`)) %>%
            full_join(., ot2021_raw %>%
              # remove blanks
              select(-`ERR5729658.unmapped`, -`ERR5729659.unmapped`, -`ERR5729660.unmapped`, -`ERR5729661.unmapped`)) %>%
            full_join(., de_raw %>%
                        select(-matches(nhp %>% str_c(collapse = "|")))) %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks, dentin and soil
            select(-matches("CSS|CSD|EXB|LIB|CSN|CSL")), by = "Species") %>%
  anti_join(.,  mid_contaminants %>%
            bind_rows(., rad_contaminants) %>%
            bind_rows(., kil_contaminants_freq) %>%
            bind_rows(., iber_contaminants) %>%
            bind_rows(., ei_contaminants) %>%
            bind_rows(., ot_contaminants) %>%
            bind_rows(., de_contaminants) %>%
            bind_rows(., mw_contaminants) %>%
            bind_rows(., cont_all_sp) %>% # pacific contaminants already combinded and de-duplicated
            distinct(), by = "Species") %>%
  replace(is.na(.), 0) %>%
  # remove any species that have no counts after removing the blanks
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

# remove the totals attribute so the column totals can be calculated again later
attr(pe_all_table, "totals") <- NULL

colnames(pe_all_table) <- gsub(".unmapped","", colnames(pe_all_table))
colnames(pe_all_table) <- gsub("MeganServer::","", colnames(pe_all_table))

pe_all_table <- pe_all_table %>%
  # remove poorly-preserved samples
  select(-matches(world_fail %>% pull(Sample) %>% str_c(collapse = "|"))) %>%
  # remove any species that have no counts after removing the poorly-preserved samples
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

# remove the totals attribute so the column totals can be calculated again later
attr(pe_all_table, "totals") <- NULL

```

```{r}
# combine decontaminated tables where all pacific samples were decontaminated together, and all European tables were decontaminated together

pe_each_table <- pacific_sp_all_filt %>%
  full_join(world_species_all_decontam) %>%
  replace(is.na(.), 0)

```

```{r}
# combine individually-decontaminated tables

pe_indiv_table <- world_species_indiv_decontam %>%
  full_join(., otago_sp, by = "Species") %>%
  full_join(., oklahoma_sp, by = "Species") %>%
  full_join(., jena_sp, by = "Species") %>%
  replace(is.na(.), 0)



```


Now plots
```{r pe_indiv_plot}

pe_indiv_table_1 <- pe_indiv_table %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_indiv_table_1_pca <- tune.pca(pe_indiv_table_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_indiv_table_1.pca <- pca(pe_indiv_table_1, ncomp = 3, logratio = 'CLR')

# plot the PCA
sp_indiv_lab_plot <- plot_pca(pe_indiv_table_1.pca, "PC1", "PC2", "Lab", "Lab")
sp_indiv_lab_plot

sp_indiv_study_plot <- plot_pca(pe_indiv_table_1.pca, "PC1", "PC2", "Study", "Lab")
sp_indiv_study_plot

sp_indiv_continent_plot <- plot_pca(pe_indiv_table_1.pca, "PC1", "PC2", "Continent", "Lab")
sp_indiv_continent_plot

# ggsave("./06-publication/main_figures/Figure_XX4/world_indiv.pdf", plot = sp_indiv_continent_plot, device = "pdf",
#        scale = 1, width = 8, height = 4.5, units = c("in"), dpi = 300)


# top 20 genera separating samples along PC1? (10+, 10-)
pe_indiv_table_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_indiv_table_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 30) %>%
  bind_rows(., pe_indiv_table_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_indiv_table_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 30))

pe_indiv_table_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_indiv_table_1.pca$loadings$X)) %>%
  select(Species, PC2) %>%
  arrange(PC2) %>%
  slice_max(PC2, n = 30) %>%
  bind_rows(., pe_indiv_table_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_indiv_table_1.pca$loadings$X)) %>%
            select(Species, PC2) %>%
            arrange(PC2) %>%
            slice_min(PC2, n = 30))

pe_indiv_table_1.pca$variates$X %>%
  as.data.frame(.)

```

```{r indiv_read_stats}
# average sequence length
plot_pca_cont(pe_indiv_table_1.pca, "PC1", "PC2", "seq_len_avg", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(pe_indiv_table_1.pca, "PC1", "PC2", "gc_avg", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(pe_indiv_table_1.pca, "PC1", "PC2", "No_reads_no_human", 3, "Total Reads")

# total number of reads
plot_pca_cont(pe_indiv_table_1.pca, "PC1", "PC2", "Age_mean_log", 3, "Log10(Mean Age (BP))")

```

```{r pe_each_plot}

pe_each_table_1 <- pe_each_table %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_each_table_1_pca <- tune.pca(pe_each_table_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_each_table_1.pca <- pca(pe_each_table_1, ncomp = 3, logratio = 'CLR')

# plot the PCA
sp_each_lab_plot <- plot_pca(pe_each_table_1.pca, "PC1", "PC2", "Lab", "Lab")
sp_each_lab_plot

sp_each_study_plot <- plot_pca(pe_each_table_1.pca, "PC1", "PC2", "Study", "Lab")
sp_each_study_plot

# top 20 genera separating samples along PC1? (10+, 10-)
pe_each_table_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_each_table_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 30) %>%
  bind_rows(., pe_each_table_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_each_table_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 30))

pe_each_table_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_each_table_1.pca$loadings$X)) %>%
  select(Species, PC2) %>%
  arrange(PC2) %>%
  slice_max(PC2, n = 30) %>%
  bind_rows(., pe_each_table_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_each_table_1.pca$loadings$X)) %>%
            select(Species, PC2) %>%
            arrange(PC2) %>%
            slice_min(PC2, n = 30))


```

```{r each_read_stats}
# average sequence length
plot_pca_cont(pe_each_table_1.pca, "PC1", "PC2", "seq_len_avg", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(pe_each_table_1.pca, "PC1", "PC2", "gc_avg", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(pe_each_table_1.pca, "PC1", "PC2", "No_reads_no_human", 3, "Total Reads")

# mean age
plot_pca_cont(pe_each_table_1.pca, "PC1", "PC2", "Age_mean_log", 3, "Log10(Mean Age (BP))")

```

```{r pe_all_plot}

pe_all_table_1 <- pe_all_table %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_table_1_pca <- tune.pca(pe_all_table_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_table_1.pca <- pca(pe_all_table_1, ncomp = 3, logratio = 'CLR')

# plot the PCA
sp_all_lab_plot <- plot_pca(pe_all_table_1.pca, "PC1", "PC2", "Lab", "Lab")
sp_all_lab_plot

sp_all_study_plot <- plot_pca(pe_all_table_1.pca, "PC1", "PC2", "Study", "Lab")
sp_all_study_plot

sp_all_continent_plot <- plot_pca(pe_all_table_1.pca, "PC1", "PC2", "Continent", "Lab")
sp_all_continent_plot

exp_var <- paste0(round(pe_all_table_1.pca$prop_expl_var$X * 100, 2), "%")

pe_all_table_1.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  inner_join(., fread("./05-Documentation.backup/humann3_GFs_world_sample_clustering.tsv"), by = "Library_ID") %>%
  mutate(Cluster = as.character(Cluster)) %>%
  ggplot(., aes(PC1, PC2)) +
         # geom_point(aes(fill = Continent, shape = Cluster), size = 2, stroke = 0.3) +
         ggstar::geom_star(aes(fill = Continent, starshape = Cluster), size = 1.5) + # , shape = 21
         ggstar::scale_starshape_manual(values = c(15,9)) +
         scale_fill_manual(values = Continent_colors) +
         xlab(paste("PC1 - ", exp_var[1])) +
         ylab(paste("PC2 - ", exp_var[2])) +
         theme_minimal(base_size = 14) +
         theme(legend.position = "right") +
         theme(legend.title = element_blank())  

# ggsave("./06-publication/main_figures/Figure_XX4/world_all.pdf", plot = sp_all_continent_plot, device = "pdf",
#        scale = 1, width = 8, height = 4.5, units = c("in"), dpi = 300)
# 
# ggsave("./06-publication/main_figures/Figure_XX4/world_all.png", plot = sp_all_continent_plot, device = "png",
#        scale = 1, width = 8, height = 4.5, units = c("in"), dpi = 300)

# top 20 genera separating samples along PC1? (10+, 10-)
pe_all_table_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_all_table_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 30) %>%
  bind_rows(., pe_all_table_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_all_table_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 30))

pe_all_table_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_all_table_1.pca$loadings$X)) %>%
  select(Species, PC2) %>%
  arrange(PC2) %>%
  slice_max(PC2, n = 30) %>%
  bind_rows(., pe_all_table_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_all_table_1.pca$loadings$X)) %>%
            select(Species, PC2) %>%
            arrange(PC2) %>%
            slice_min(PC2, n = 30))

sp_counts <- pe_all_table %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts > 0,1,0)) %>%
  pivot_wider(names_from = "Species", values_from = "Counts") %>%
  adorn_totals(where = "col") %>%
  rename(no_species = Total) %>%
  select(Library_ID, no_species) %>%
  as_data_frame()

# fwrite(sp_counts, "./05-Documentation.backup/pe_all_table_species_counts.tsv", quote = F, sep = "\t")

metadata <- metadata %>%
  full_join(., sp_counts, by = "Library_ID")

```

```{r all_read_stats}
# average sequence length
plot_pca_cont(pe_all_table_1.pca, "PC1", "PC2", "seq_len_avg", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(pe_all_table_1.pca, "PC1", "PC2", "gc_avg", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(pe_all_table_1.pca, "PC1", "PC2", "No_reads_no_human", 3, "Total Reads")

# malt % assigned reads
plot_pca_cont(pe_all_table_1.pca, "PC1", "PC2", "Pct_assigned", 3, "% reads assigned taxonomy")

# mean age
plot_pca_cont(pe_all_table_1.pca, "PC1", "PC2", "Age_mean_log", 3, "Log10(Mean Age (BP)")

# species counts
plot_pca_cont(pe_all_table_1.pca, "PC1", "PC2", "no_species", 3, "Number of species")

```

```{r}
# number of species
pe_all_table_1.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  select(Library_ID, PC1, PC2, no_species, Pct_assigned, Continent, Study, Lab) %>%
  ggplot(., aes(x = PC1, y = no_species)) + # Species
         geom_point(aes(fill = Continent, shape = Lab), size = 3, stroke = 0.3, shape = 21) +
         scale_fill_manual(values = Continent_colors) +
         theme_classic(base_size = 10) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         # ylab("Total SNPs/Genome") +
         ggtitle("")

# % of reads assigned taxonomy
pe_all_table_1.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  select(Library_ID, PC1, PC2, no_species, Pct_assigned, Continent, Study, Lab) %>%
  ggplot(., aes(x = PC1, y = Pct_assigned)) + # Species
         geom_point(aes(fill = Continent, shape = Lab), size = 3, stroke = 0.3, shape = 21) +
         scale_fill_manual(values = Continent_colors) +
         theme_classic(base_size = 10) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         # ylab("Total SNPs/Genome") +
         ggtitle("")

# % of reads assigned taxonomy
pe_all_table_1.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  select(Library_ID, PC1, PC2, no_species, Pct_assigned, Continent, Study, Lab) %>%
  ggplot(., aes(x = no_species, y = Pct_assigned)) + # Species
         geom_point(aes(fill = PC1, shape = Lab), size = 3, stroke = 0.3, shape = 21) +
         geom_smooth(method = glm, se = F, col = "black") +
         scale_fill_viridis_c(option = "C") +
         theme_classic(base_size = 10) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         ggpubr::stat_regline_equation(label.y = 80, aes(label = ..rr.label..)) +
         # ylab("Total SNPs/Genome") +
         ggtitle("")

# color number of species
pe_all_table_1.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  select(Library_ID, PC1, PC2, no_species, Pct_assigned, Continent, Study, Lab) %>%
  ggplot(., aes(x = PC1, y = Pct_assigned)) + # Species
         geom_point(aes(fill = no_species, shape = Lab), size = 3, stroke = 0.3, shape = 21) +
         geom_smooth(se = F, col = "black") +
         scale_fill_viridis_c(option = "C") +
         theme_classic(base_size = 10) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         ggpubr::stat_regline_equation(label.y = 80, aes(label = ..rr.label..)) +
         # ylab("Total SNPs/Genome") +
         ggtitle("")

# color Pct_assigned
plot_line <- pe_all_table_1.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  select(Library_ID, PC1, PC2, no_species, Pct_assigned, Continent, Study, Lab) %>%
  ggplot(., aes(x = PC1, y = no_species)) + # Species
         geom_point(aes(fill = Pct_assigned, shape = Lab), size = 1, stroke = 0.2, shape = 21) +
         geom_smooth(method = glm, se = F, col = "black", linewidth = 0.7) +
         scale_fill_viridis_c(option = "C") +
         theme_classic(base_size = 10) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         ggpubr::stat_regline_equation(label.y = 150, aes(label = ..rr.label..)) +
         ylab("Number of species") +
         ggtitle("")
plot_line

```


```{r}
plot_pca_bi(pe_all_table_1.pca, "PC1", "PC2", "Lab", PC1)

plot_pca_bi(pe_all_table_1.pca, "PC1", "PC2", "Lab", PC2)


# top 20 genera separating samples along PC1? (10+, 10-)
pe_all_table_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_all_table_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., pe_all_table_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_all_table_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))


```


Run a PERMANOVA with adonis2 on the abundance-filtered clr-transformed species matrix.
```{r adonis_pe_all}

pe_all_table_1 <- pe_all_table %>%
  select(-ERR5729638, -ERR5729643) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_table_1_pca <- tune.pca(pe_all_table_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_table_1.pca <- pca(pe_all_table_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
pe_all_table_1.euc <- vegdist(pe_all_table_1.pca$X, method = "euclidean", na.rm = F)
pe_all_table_1.euc.pcoa <- ape::pcoa(pe_all_table_1.euc)

pe_all_table_1.euc.pcoavectors <- as.data.frame(pe_all_table_1.euc.pcoa$vectors)
pe_all_table_1.euc.pcoavectors <- pe_all_table_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

pe_all_table_1.euc.vecmet <- left_join(pe_all_table_1.euc.pcoavectors, metadata, by = "Library_ID")

pe_all_table_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2)) +
    geom_point(aes(fill = Study), size = 3, stroke = 0.3, shape = 21) +
    scale_fill_manual(values = Study_colors) +
    scale_color_manual(values = Study_colors) +
    # scale_shape_manual(values = Lab_shapes) +
    stat_ellipse(aes(colour = Study, group = Study)) +
    theme_minimal(base_size = 7) +
    ggtitle("Study")

metadata_all <- metadata %>%
    right_join(., pe_all_table_1.euc.pcoavectors %>% select(Library_ID), by = "Library_ID")
  

# test for difference between different metadata categories 
adonis2(pe_all_table_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")

all_permanova <- adonis2(pe_all_table_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(adonis2(pe_all_table_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_table_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_table_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_table_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","**","***","***","**"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)


# all together
adonis2(pe_all_table_1.euc ~ Continent + Lab + Pct_assigned + no_species + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(pe_all_table_1.euc ~ Continent + Pct_assigned + no_species + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)


```

```{r beta_dispersion}
# Test whether there is more dispersion between any of the studies

# Study
groups_study <- pe_all_table %>%
  slice_head(n = 1) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  select(-Species, -Counts) %>%
  left_join(., 
            metadata_all %>%
            select(Library_ID, Study, Continent)) %>%
  # these are also already removed in the input table below
  drop_na(Continent) %>%
  pull(Continent)

## Calculate multivariate dispersions
study_betadisper <- betadisper(pe_all_table_1.euc, groups_study, bias.adjust =TRUE)
study_betadisper

## Perform test
anova(study_betadisper)

## Permutation test for F
permutest(study_betadisper, pairwise = TRUE, permutations = 99)

## Tukey's Honest Significant Differences
study_betadisper.HSD <- TukeyHSD(study_betadisper)
plot(study_betadisper.HSD)

## Plot the groups and distances to centroids on the
## first two PCoA axes
## with data ellipses instead of hulls
plot(study_betadisper, ellipse = TRUE, hull = FALSE) # 1 sd data ellipse

## can also specify which axes to plot, ordering respected
# plot(study_betadisper, axes = c(3,1), seg.col = "forestgreen", seg.lty = "dashed")

## Draw a boxplot of the distances to centroid for each group
boxplot(study_betadisper)

study_dist <- as.data.frame(study_betadisper$distances) %>%
  rownames_to_column("Library_ID") %>%
  rename(dist_to_centroid = 2)

```

```{r betadisper_raincloud}

raincloud <- study_dist %>%
  left_join(., metadata_all %>%
              select(Library_ID, Continent)) %>%
  mutate(Continent = fct_relevel(Continent, "Pacific","Asia","Africa","Europe","Caribbean", "North America")) %>%
  arrange(Continent) %>%
ggplot(., aes(x = Continent, y = dist_to_centroid, fill = Continent)) + 
  ## add half-violin from {ggdist} package
  ggdist::stat_halfeye(
    ## custom bandwidth
    adjust = .5, 
    ## adjust height
    width = .6, 
    ## move geom to the right
    justification = -.2, 
    ## remove slab interval
    .width = 0, 
    point_colour = NA
  ) + 
  geom_boxplot(
    width = .12, 
    ## remove outliers
    outlier.color = NA, ## `outlier.shape = NA` works as well
    color = "grey30"
  ) +
  ## add dot plots from {ggdist} package
  ggdist::stat_dots(
    ## orientation to the left
    side = "left", 
    ## move geom to the left
    justification = 1.1, 
    ## adjust grouping (binning) of observations 
    binwidth = 0.5
  ) + 
  ## remove white space on the left
  coord_cartesian(xlim = c(1.2, NA)) +
  theme_minimal(base_size = 8) +
  scale_fill_manual(values = Continent_colors) +
  theme(axis.text = element_text(size = 8)) +
  theme(legend.position = "none") +
  xlab("Study") +
  ylab("Distance to centroid") +
  scale_y_continuous(breaks=scales::pretty_breaks(n=8)) +
  theme(axis.title.x = element_text(vjust=-1.5)) +
  theme(axis.title.y = element_text(vjust=1.9))

raincloud


```

```{r}

study_dist %>%
  left_join(., metadata_all %>%
              select(Library_ID, Continent, Age_mean, Age_mean_log)) %>%
ggplot(., aes(x = Continent, y = Age_mean, fill = Continent)) + 
  ## add half-violin from {ggdist} package
  ggdist::stat_halfeye(
    ## custom bandwidth
    adjust = .5, 
    ## adjust height
    width = .6, 
    ## move geom to the right
    justification = -.2, 
    ## remove slab interval
    .width = 0, 
    point_colour = NA
  ) + 
  geom_boxplot(
    width = .12, 
    ## remove outliers
    outlier.color = NA, ## `outlier.shape = NA` works as well
    color = "grey30"
  ) +
  ## add dot plots from {ggdist} package
  ggdist::stat_dots(
    ## orientation to the left
    side = "left", 
    ## move geom to the left
    justification = 1.1, 
    ## adjust grouping (binning) of observations 
    binwidth = 5
  ) + 
  ## remove white space on the left
  coord_cartesian(ylim = c(NA, 16000)) +
  theme_minimal(base_size = 10) +
  scale_fill_manual(values = Continent_colors) +
  theme(axis.text = element_text(size = 10)) +
  theme(legend.position = "none") +
  xlab("Study") +
  ylab("Mean age (BP)") +
  scale_y_continuous(breaks=scales::pretty_breaks(n=8)) +
  theme(axis.title.x = element_text(vjust=-1.5)) +
  theme(axis.title.y = element_text(vjust=1.9))


```

```{r}

pcaplot_test <- (((sp_all_continent_plot / plot_line) + plot_layout(heights = c(4,1))) | raincloud) +
  plot_layout(widths = c(1,1.2))
pcaplot_test

# ggsave("./06-publication/main_figures/Figure_XX4/world_all.svg", plot = pcaplot_test, device = "svg",
#        scale = 1, width = 10, height = 4.5, units = c("in"), dpi = 300)
# 
# ggsave("./06-publication/main_figures/Figure_XX4/world_all.pdf", plot = pcaplot_test, device = "pdf",
#        scale = 1, width = 10, height = 4.5, units = c("in"), dpi = 300)

```

```{r}

pe_all_table_1.pca$X %>% 
  as.data.frame.matrix() %>%
  rownames_to_column("Library_ID") %>%
  as_tibble() %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "CLR") %>%
  filter(str_detect(Species, "Ottowia")) %>%
  select(-Species) %>%
  full_join(., pe_all_table %>%
            filter(str_detect(Species, "Ottowia")) %>%
            pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
            mutate(Counts = Counts + 1,
                   log_counts = log10(Counts)), by = "Library_ID") %>%
  select(-Species) %>%
  fwrite(., "./05-Documentation.backup/ottowia_sp_counts.tsv", quote = F, sep = "\t")

de_raw %>%
  filter(str_detect(`#Datasets`, "Ottowia")) %>%
  select(matches("Datasets|FUM003|GOY006|ZAF001|LIS002"))

pe_all_table_1.pca$X %>% 
  as.data.frame.matrix() %>%
  rownames_to_column("Library_ID") %>%
  as_tibble() %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "CLR") %>%
  filter(str_detect(Species, "Streptococcus")) %>%
  select(-Species) %>%
  full_join(., pe_all_table %>%
            filter(str_detect(Species, "Streptococcus")) %>%
            pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
            mutate(Counts = Counts + 1,
                   log_counts = log10(Counts)), by = "Library_ID") %>%
  group_by(Library_ID) %>%
  summarize(Strep_counts  = sum(Counts)) %>%
  ungroup() %>%
  rename(Counts = Strep_counts) %>%
  fwrite(., "./05-Documentation.backup/streptococcus_sp_counts.tsv", quote = F, sep = "\t")

pe_all_table_1.pca$X %>% 
  as.data.frame.matrix() %>%
  rownames_to_column("Library_ID") %>%
  as_tibble() %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "CLR") %>%
  filter(str_detect(Species, "Desulfobulbus")) %>%
  select(-Species) %>%
  full_join(., pe_all_table %>%
            filter(str_detect(Species, "Desulfobulbus")) %>%
            pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
            mutate(Counts = Counts + 1,
                   log_counts = log10(Counts)), by = "Library_ID") %>%
  select(-Species) %>%
  fwrite(., "./05-Documentation.backup/desulfobulbus_sp_counts.tsv", quote = F, sep = "\t")

```


```{r}

sp_lab_plot <- sp_indiv_lab_plot + sp_each_lab_plot + sp_all_lab_plot +
  plot_layout(nrow = 1)  + 
  plot_layout(guides = "collect")

sp_lab_plot


sp_study_plot <- sp_indiv_study_plot + sp_each_study_plot + sp_all_study_plot +
  plot_layout(nrow = 1)  + 
  plot_layout(guides = "collect")

sp_study_plot

```

### Filter out taxa in only 1 lab
```{r all_plot_1lab_filt}

# Taxa in only 1 group
pe_all_1_lab <- pe_all_table %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts == 0, 0, 1)) %>%
  full_join(., metadata %>%
              select(Library_ID, Lab)) %>%
  # get counts per lab for each species
  group_by(Species, Lab) %>%
  summarize(In_lab = sum(Counts)) %>%
  mutate(In_lab = ifelse(In_lab == 0,0,1)) %>%
  ungroup() %>%
  group_by(Species) %>%
  summarize(num_labs = sum(In_lab)) %>%
  arrange(num_labs) %>%
  filter(num_labs == 1)


pe_all_1lab_filt_1 <- pe_all_table %>%
  anti_join(., pe_all_1_lab) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_1lab_filt_1_pca <- tune.pca(pe_all_1lab_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_1lab_filt_1.pca <- pca(pe_all_1lab_filt_1, ncomp = 3, logratio = 'CLR')


# plot the PCA
pe_all_1lab_plot <- plot_pca(pe_all_1lab_filt_1.pca, "PC1", "PC2", "Study", "Lab")
pe_all_1lab_plot
plot_pca(pe_all_1lab_filt_1.pca, "PC1", "PC2", "Lab", "Lab")

# top 20 genera separating samples along PC1? (10+, 10-)
pe_all_1lab_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_all_1lab_filt_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., pe_all_1lab_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_all_1lab_filt_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r all_read_stats}
# average sequence length
plot_pca_cont(pe_all_1lab_filt_1.pca, "PC1", "PC2", "seq_len_avg", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(pe_all_1lab_filt_1.pca, "PC1", "PC2", "gc_avg", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(pe_all_1lab_filt_1.pca, "PC1", "PC2", "No_reads_no_human", 3, "Total Reads")

# mean age
plot_pca_cont(pe_all_1lab_filt_1.pca, "PC1", "PC2", "Age_mean_log", 3, "Log10(Mean Age (BP))")

```

### Filter out taxa in <= 1/3 of samples in each Lab
```{r all_plot_1lab_filt}
# Taxa in at least 1/3 of samples in at least 2 of the 3 labs
pe_all_filt_presence_more_30 <- pe_all_table %>%
  # mutate_if(is.numeric, ~1 * (. > 0)) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts == 0, 0, 1)) %>%
  left_join(., metadata %>%
              select(Library_ID, Lab), by = "Library_ID") %>%
  arrange(Species, Lab) %>%
  group_by(Species,Lab) %>%
  summarize(Percent_presence = sum(Counts)/length(Counts)*100) %>%
  # select the pathways that are present in at least 1/3 of at least 2 groups
  ungroup() %>%
  # pathways present in > 1/3 of samples per lab
  filter(Percent_presence >= 33.33) %>%
  group_by(Species) %>%
  count() %>%
  filter(n > 1) %>%
  # pull out these pathways to keep
  select(Species)

```

```{r}


pe_all_pres33_filt_1 <- pe_all_table %>%
  inner_join(., pe_all_filt_presence_more_30) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_pres33_filt_1_pca <- tune.pca(pe_all_pres33_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_pres33_filt_1.pca <- pca(pe_all_pres33_filt_1, ncomp = 3, logratio = 'CLR')

# plot the PCA
pe_all_pres33_plot <- plot_pca(pe_all_pres33_filt_1.pca, "PC1", "PC2", "Study", "Lab")
pe_all_pres33_plot

plot_pca(pe_all_pres33_filt_1.pca, "PC1", "PC2", "Lab", "Lab")

# top 20 genera separating samples along PC1? (10+, 10-)
pe_all_pres33_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_all_pres33_filt_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., pe_all_pres33_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_all_pres33_filt_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r all_read_stats}
# average sequence length
plot_pca_cont(pe_all_pres33_filt_1.pca, "PC1", "PC2", "seq_len_avg", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(pe_all_pres33_filt_1.pca, "PC1", "PC2", "gc_avg", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(pe_all_pres33_filt_1.pca, "PC1", "PC2", "No_reads_no_human", 3, "Total Reads")

# mean age
plot_pca_cont(pe_all_pres33_filt_1.pca, "PC1", "PC2", "Age_mean_log", 3, "Log10(Mean Age (BP))")

```

Run a PERMANOVA with adonis2 on the abundance-filtered clr-transformed species matrix.
```{r adonis_pe_all_pres33}

pe_all_pres33_filt_1 <- pe_all_table %>%
  # these have no age and need to be removed to deal with that
  select(-ERR5729638, -ERR5729643) %>%
  inner_join(., pe_all_filt_presence_more_30) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_pres33_filt_1_pca <- tune.pca(pe_all_pres33_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_pres33_filt_1.pca <- pca(pe_all_pres33_filt_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
pe_all_pres33_filt_1.euc <- vegdist(pe_all_pres33_filt_1.pca$X, method = "euclidean", na.rm = F)
pe_all_pres33_filt_1.euc.pcoa <- ape::pcoa(pe_all_pres33_filt_1.euc)

pe_all_pres33_filt_1.euc.pcoavectors <- as.data.frame(pe_all_pres33_filt_1.euc.pcoa$vectors)
pe_all_pres33_filt_1.euc.pcoavectors <- pe_all_pres33_filt_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

pe_all_pres33_filt_1.euc.vecmet <- left_join(pe_all_pres33_filt_1.euc.pcoavectors, metadata, by = "Library_ID")

pe_all_pres33_filt_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Lab)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Lab_colors) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Lab")

metadata_all <- metadata %>%
    right_join(., pe_all_pres33_filt_1.euc.pcoavectors %>% select(Library_ID), by = "Library_ID")
  


# test for difference between different metadata categories 
adonis2(pe_all_pres33_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")

all_permanova <- adonis2(pe_all_pres33_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(adonis2(pe_all_pres33_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_pres33_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_pres33_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_pres33_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","**","**","*"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)


# all together
adonis2(pe_all_pres33_filt_1.euc ~ Lab  + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(pe_all_pres33_filt_1.euc ~ Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)


```

```{r, eval = F}
# use this table for downstream analyses
fwrite(sp_all_pres33_filt_1, "./05-Documentation.backup/pacific_species_decontam_pres33_filt.tsv", quote = F, sep = "\t")
# fwrite(sp_all_pres33_filt_1, "../strep_clades/00-documentation/pacific_species_decontam_pres33_filt.tsv", quote = F, sep = "\t")

```


### Filter out taxa that are less than 0.005% relative abundane across all samples?
```{r, eval = F}
# let's look at the abundance of these taxa
pe_all_table %>%
  as_tibble(.) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  select(Species, Percent) %>%
  arrange(desc(Percent)) %>%
  mutate(Percent = round(Percent, 4))

```


```{r all_plot_abund005_filt}

# Taxa present at < 0.005% relative abundance
pe_abund005_lab <- pe_all_table %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent <= 0.0001) %>%
  # select(Species, Total, Percent)
  select(Species)

pe_all_abund005_filt_1 <- pe_all_table %>%
  anti_join(., pe_abund005_lab) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  # mutate(Counts = ifelse(Counts > 0, 1, 0)) %>%
  # pivot_wider(names_from = "Species", values_from = "Counts") %>%
  # adorn_totals(where = "col") %>%
  # select(Total)
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_abund005_filt_1_pca <- tune.pca(pe_all_abund005_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_abund005_filt_1.pca <- pca(pe_all_abund005_filt_1, ncomp = 3, logratio = 'CLR')

# plot the PCA
pe_all_abund005_plot <- plot_pca(pe_all_abund005_filt_1.pca, "PC1", "PC2", "Study", "Lab")
pe_all_abund005_plot

plot_pca(pe_all_abund005_filt_1.pca, "PC1", "PC2", "Lab", "Lab")

# top 20 genera separating samples along PC1? (10+, 10-)
pe_all_abund005_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_all_abund005_filt_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., pe_all_abund005_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_all_abund005_filt_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))

```

```{r}
# how many species per sample?

counts_table <- pe_all_table %>%
  anti_join(., pe_abund005_lab) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts > 0, 1, 0)) %>%
  pivot_wider(names_from = "Species", values_from = "Counts") %>%
  adorn_totals(where = "col") %>%
  select(Library_ID, Total) %>%
  as_tibble()

new_study_colors <- c("Pacific" = "#098BD9", 
                  "Velsko2022" = "#7472af", 
                  "Velsko2019" = "#cc79a7", 
                  "Mann2018" = "#f6865e", 
                  # "Iberia" = "#1ec1a7",
                  "Eisenhoffer2020" = "#86c73a",
                  "Ottoni2021" = "#616161",
                  "FellowsYates2021" = "#e1f2fe",
                  "Mann2018_world" = "7D3560")



raincloud <- counts_table %>%
  filter(!str_detect(Library_ID, nhp %>% str_c(collapse = "|"))) %>%
  left_join(., metadata %>%
              select(Library_ID, Study), by = "Library_ID") %>%
  mutate(Study = str_replace_all(Study, "Kilteasheen","Mann2018"),
         Study = str_replace_all(Study, "Middenbeemster","Velsko2022"),
         Study = str_replace_all(Study, "Iberia","Velsko2022"),
         Study = str_replace_all(Study, "Radcliffe","Velsko2019"),
         Study = replace_na(Study, "Pacific")) %>%
  mutate(Study = ifelse(str_detect(Library_ID, "SRR") & Study == "Mann2018", "Mann2018_world",Study)) %>%
  mutate(Study = fct_relevel(Study, "Pacific","Eisenhoffer2020","FellowsYates2021","Mann2018","Mann2018_world","Ottoni2021","Velsko2019","Velsko2022")) %>%
  ggplot(., aes(x = Study, y = Total, fill = Study)) + 
  ## add half-violin from {ggdist} package
  ggdist::stat_halfeye(
    ## custom bandwidth
    adjust = .5, 
    ## adjust height
    width = .6, 
    ## move geom to the right
    justification = -.2, 
    ## remove slab interval
    .width = 0, 
    point_colour = NA
  ) + 
  geom_boxplot(
    width = .12, 
    ## remove outliers
    outlier.color = NA, ## `outlier.shape = NA` works as well
    color = "grey30"
  ) +
  ## add dot plots from {ggdist} package
  ggdist::stat_dots(
    ## orientation to the left
    side = "left", 
    ## move geom to the left
    justification = 1.1, 
    ## adjust grouping (binning) of observations 
    binwidth = 2
  ) + 
  ## remove white space on the left
  coord_cartesian(xlim = c(1.2, NA)) +
  theme_minimal(base_size = 12) +
  scale_fill_manual(values = new_study_colors) +
  theme(axis.text.y = element_text(size = 12)) +
  theme(legend.position = "none") +
  xlab("") +
  ylab("No. species") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) 

  raincloud

# ggsave("./06-publication/supplemental_figures/Sup_fig_4/species_counts_raincloud.pdf", plot = raincloud, device = "pdf",
       # scale = 1, width = 8.5, height = 5, units = c("in"), dpi = 300)

```

```{r mpa3_data}
# what about from metaPhlAn3? Are species counts also lower in Pacific and Eisenhoffer2020 data than the rest?

anc_mpa3_raw <- fread("../strep_clades/05-results/mpa3_strep_clades_anc_calc_abundance_table.tsv") %>%
  select(-matches("CSS|CSD|01_S32"))

colnames(anc_mpa3_raw) <- gsub(".metaphlan.profile","", colnames(anc_mpa3_raw))


mod_mpa3_raw <- fread("../strep_clades/05-results/mpa3_strep_clades_mod_human_abundance_table.tsv")

colnames(mod_mpa3_raw) <- gsub(".metaphlan3","", colnames(mod_mpa3_raw))
colnames(mod_mpa3_raw) <- gsub(".SG1","", colnames(mod_mpa3_raw))


oral_mpa3_raw <- anc_mpa3_raw %>%
  full_join(., mod_mpa3_raw, by = "clade_name") %>%
  select(-clade_taxid, -NCBI_tax_id) %>%
  replace(is.na(.), 0)
  
  
```

```{r mpa3_species}

oral_mpa3_sp <- oral_mpa3_raw %>%
  filter(str_detect(clade_name, "s__")) %>%
  separate(clade_name, into = c("kingdom","phylum","class","order","family","genus","Species"), sep = "\\|") %>%
  select(-c("kingdom","phylum","class","order","family","genus")) %>%
  mutate(Species = str_replace_all(Species,"s__",""))


# number of species detected in each sample
oral_mpa3_species_counts <- oral_mpa3_sp %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts > 0,1,0)) %>%
  pivot_wider(names_from = "Species", values_from = "Counts") %>%
  adorn_totals(where = "col") %>%
  select(Library_ID,Total) %>%
  rename(Species_counts = Total) %>%
  as_tibble()

```

```{r}
# how many species per sample?
oral_mpa3_species_counts %>%
  filter(!str_detect(Library_ID, nhp %>% str_c(collapse = "|"))) %>%
  # remove blanks
  # filter(!str_detect(Library_ID, blanks %>% str_c(collapse = "|"))) %>%
  filter(Library_ID != "Blank") %>%
  inner_join(., metadata %>%
              select(Library_ID, Study), by = "Library_ID") %>%
  mutate(Study = str_replace_all(Study, "Kilteasheen","Mann2018"),
         Study = str_replace_all(Study, "Middenbeemster","Velsko2022"),
         Study = str_replace_all(Study, "Iberia","Velsko2022"),
         Study = str_replace_all(Study, "Radcliffe","Velsko2019"),
         Study = replace_na(Study, "Pacific")) %>%
  mutate(Study = ifelse(str_detect(Library_ID, "SRR") & Study == "Mann2018", "Mann2018_world",Study)) %>%
  mutate(Study = fct_relevel(Study, "Pacific","Eisenhoffer2020","FellowsYates2021","Mann2018","Mann2018_world","Ottoni2021","Velsko2019","Velsko2022")) %>%
  ggplot(., aes(x = Study, y = Species_counts, fill = Study)) + 
  ## add half-violin from {ggdist} package
  ggdist::stat_halfeye(
    ## custom bandwidth
    adjust = .5, 
    ## adjust height
    width = .6, 
    ## move geom to the right
    justification = -.2, 
    ## remove slab interval
    .width = 0, 
    point_colour = NA
  ) + 
  geom_boxplot(
    width = .12, 
    ## remove outliers
    outlier.color = NA, ## `outlier.shape = NA` works as well
    color = "grey30"
  ) +
  ## add dot plots from {ggdist} package
  ggdist::stat_dots(
    ## orientation to the left
    side = "left", 
    ## move geom to the left
    justification = 1.1, 
    ## adjust grouping (binning) of observations 
    binwidth = 2
  ) + 
  ## remove white space on the left
  coord_cartesian(xlim = c(1.2, NA)) +
  theme_minimal(base_size = 12) +
  scale_fill_manual(values = new_study_colors) +
  theme(axis.text.y = element_text(size = 12)) +
  theme(legend.position = "none") +
  xlab("") +
  ylab("No. species") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) 


```



```{r all_read_stats}
# average sequence length
plot_pca_cont(pe_all_abund005_filt_1.pca, "PC1", "PC2", "seq_len_avg", 3, "Avg. Read Length")

# average gc content
plot_pca_cont(pe_all_abund005_filt_1.pca, "PC1", "PC2", "gc_avg", 3, "Average GC (%)")

# total number of reads
plot_pca_cont(pe_all_abund005_filt_1.pca, "PC1", "PC2", "No_reads_no_human", 3, "Total Reads")

# mean age
plot_pca_cont(pe_all_abund005_filt_1.pca, "PC1", "PC2", "Age_mean_log", 3, "Log10(Mean Age (BP))")

# species count
exp_var <- paste0(round(pe_all_abund005_filt_1.pca$prop_expl_var$X * 100, 2), "%")
df_X <- pe_all_abund005_filt_1.pca$variates$X %>%
          as.data.frame() %>%
          rownames_to_column("Library_ID") %>%
          inner_join(metadata, by = "Library_ID") %>%
          left_join(., counts_table, by = "Library_ID")

df_X %>%
  ggplot(., aes(PC1, PC2)) +
    geom_point(aes(fill = Total), size = 3.5, stroke = 0.3, shape = 21) +
    scale_fill_viridis_c(option = "C") +
    xlab(paste("PC1 - ", exp_var[1])) +
    ylab(paste("PC2 - ", exp_var[2])) +
    theme_minimal(base_size = 14) +
    theme(legend.position = "right") +
    theme(legend.title = element_blank())


```

```{r pca_kegg_sample_clusters}
# now also plot with the clusters determined by hierarchical clustering of the KEGG orthologs
kegg_sample_clusters <- fread("./05-Documentation.backup/humann3_GFs_world_sample_clustering.tsv")

exp_var <- paste0(round(pe_all_abund005_filt_1.pca$prop_expl_var$X * 100, 2), "%")
df_X <- pe_all_abund005_filt_1.pca$variates$X %>%
          as.data.frame() %>%
          rownames_to_column("Library_ID") %>%
          inner_join(metadata, by = "Library_ID") %>%
          left_join(., kegg_sample_clusters %>%
                      select(Library_ID, Cluster), by = "Library_ID") %>%
  # there are samples left in the species PCA that were removed from KEGG Ortholog analysis b/c of too few orthologs identified
  # give a different symbol to these by adding a pseudo cluster number
  mutate(Cluster = replace_na(Cluster,3),
         Cluster = as.character(Cluster)) %>%
  mutate(Cluster = str_replace_all(Cluster, "1", "A"),
         Cluster = str_replace_all(Cluster, "2", "B"),
         Cluster = str_replace_all(Cluster, "3", "none"))

    
pca_kegg <- df_X %>%
 mutate(Cluster = as.character(Cluster)) %>% 
  ggplot(., aes(PC1, PC2)) +
    # geom_point(aes(fill = Study, shape = Cluster), size = 4, stroke = 0.3) +
    ggstar::geom_star(aes(fill = Continent, starshape = Cluster), size = 2.0) + # , shape = 21
    scale_fill_manual(values = Continent_colors) +
    ggstar::scale_starshape_manual(values = c(15,9, 12)) +
    # scale_shape_manual(values = c(21, 23, 25)) +
    xlab(paste("PC1 - ", exp_var[1])) +
    ylab(paste("PC2 - ", exp_var[2])) +
    theme_minimal(base_size = 14) +
    theme(legend.position = "right") +
    theme(legend.title = element_blank())
 pca_kegg

# ggsave("./06-publication/supplemental_figures/Sup_fig_7/species_pca_kegg_clusters.pdf", plot = pca_kegg, device = "pdf",
#        scale = 1, width = 8.5, height = 4, units = c("in"), dpi = 300)


```



```{r each_plot_abund005_filt}

# Taxa present at < 0.005% relative abundance
pe_abund005_each <- pe_all_table %>%
  # Pacific
  select(matches(c("Species",pacific_sample_list))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent > 0) %>%
  filter(Percent <= 0.001) %>%
  # select(Species, Total, Percent)
  select(Species) %>%
  full_join(., pe_all_table %>%
            # FellowsYates2021
            select(matches(c("Species",fy))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent > 0) %>%
            filter(Percent <= 0.001) %>%
            select(Species)) %>%
  full_join(., pe_all_table %>%
            # Ottoni2021
            select(matches(c("Species",ot))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent > 0) %>%
            filter(Percent <= 0.001) %>%
            select(Species)) %>%
  full_join(., pe_all_table %>%
            # Mann2018
            select(matches(c("Species",ma))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent > 0) %>%
            filter(Percent <= 0.001) %>%
            select(Species)) %>%
  full_join(., pe_all_table %>%
            # Eisenhoffer2020
            select(matches(c("Species",ei))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent > 0) %>%
            filter(Percent <= 0.001) %>%
            select(Species)) %>%
  full_join(., pe_all_table %>%
            # Velsko2019
            select(matches(c("Species",ve19))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent > 0) %>%
            filter(Percent <= 0.001) %>%
            select(Species)) %>%
  full_join(., pe_all_table %>%
            # Velsko2022
            select(matches(c("Species",ve22))) %>%
            adorn_totals(where = "col") %>%
            mutate(Percent = Total / sum(Total) * 100) %>%
            filter(Percent > 0) %>%
            filter(Percent <= 0.001) %>%
            select(Species))

pe_all_abund005_each_filt_1 <- pe_all_table %>%
  anti_join(., pe_abund005_each) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_abund005_each_filt_1_pca <- tune.pca(pe_all_abund005_each_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_abund005_each_filt_1.pca <- pca(pe_all_abund005_each_filt_1, ncomp = 3, logratio = 'CLR')

# plot the PCA
pe_all_abund005_each_plot <- plot_pca(pe_all_abund005_each_filt_1.pca, "PC1", "PC2", "Study", "Lab")
pe_all_abund005_each_plot

plot_pca(pe_all_abund005_each_filt_1.pca, "PC1", "PC2", "Lab", "Lab")

# top 20 genera separating samples along PC1? (10+, 10-)
pe_all_abund005_each_filt_1.pca$loadings$X %>%
  as_tibble(.) %>%
  mutate(Species = rownames(pe_all_abund005_each_filt_1.pca$loadings$X)) %>%
  select(Species, PC1) %>%
  arrange(PC1) %>%
  slice_max(PC1, n = 10) %>%
  bind_rows(., pe_all_abund005_each_filt_1.pca$loadings$X %>%
            as_tibble(.) %>%
            mutate(Species = rownames(pe_all_abund005_each_filt_1.pca$loadings$X)) %>%
            select(Species, PC1) %>%
            arrange(PC1) %>%
            slice_min(PC1, n = 10))


# pe_all_abund005_each_filt_1.pca$variates$X %>%
#   as_tibble(.) %>%
#   mutate(Species = rownames(pe_all_abund005_each_filt_1.pca$variates$X)) %>%
#   select(Species, PC1, PC2) %>%
#   filter(PC1 <= -25 | PC1 >= 25) %>%
#   filter(PC2 >= -20) %>%
#   filter(PC2 <= 20) %>%
#   fwrite(., "./05-Documentation.backup/pc_end_samples.tsv", quote = F, sep = "\t")

```

Run a PERMANOVA with adonis2 on the abundance-filtered clr-transformed species matrix.
```{r adonis_pe_all_pres33}

pe_all_abund005_all_filt_1 <- pe_all_table %>%
  # remove the species present at < 0.005% abundance
  anti_join(., pe_abund005_lab) %>%
  # these have no age and need to be removed to deal with that
  select(-ERR5729638, -ERR5729643) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_abund005_all_filt_1_pca <- tune.pca(pe_all_abund005_all_filt_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_abund005_all_filt_1.pca <- pca(pe_all_abund005_all_filt_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
pe_all_abund005_all_filt_1.euc <- vegdist(pe_all_abund005_all_filt_1.pca$X, method = "euclidean", na.rm = F)
pe_all_abund005_all_filt_1.euc.pcoa <- ape::pcoa(pe_all_abund005_all_filt_1.euc)

pe_all_abund005_all_filt_1.euc.pcoavectors <- as.data.frame(pe_all_abund005_all_filt_1.euc.pcoa$vectors)
pe_all_abund005_all_filt_1.euc.pcoavectors <- pe_all_abund005_all_filt_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

pe_all_abund005_all_filt_1.euc.vecmet <- left_join(pe_all_abund005_all_filt_1.euc.pcoavectors, metadata, by = "Library_ID")

pe_all_abund005_all_filt_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2, colour = Study)) +
  geom_point(size = 3) +
  scale_colour_manual(values = Study_colors) +
  # scale_shape_manual(values = Lab_shapes) +
  # stat_ellipse(aes(colour = Site_code, group = Site_code)) +
  theme_minimal(base_size = 7) +
  ggtitle("Lab")

metadata_all <- metadata %>%
    right_join(., pe_all_abund005_all_filt_1.euc.pcoavectors %>% select(Library_ID), by = "Library_ID") %>%
    left_join(., counts_table, by = "Library_ID")
  


# test for difference between different metadata categories 
adonis2(pe_all_abund005_all_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")

all_permanova <- adonis2(pe_all_abund005_all_filt_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(adonis2(pe_all_abund005_all_filt_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_abund005_all_filt_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_abund005_all_filt_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_abund005_all_filt_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_abund005_all_filt_1.euc ~ Total, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","***","**","**","*","***"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)


# all together
adonis2(pe_all_abund005_all_filt_1.euc ~ Total + Lab  + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(pe_all_abund005_all_filt_1.euc ~ Total + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)


```


### Plot together
patchwork example
```{r filtered_plots}

pe_plot <- sp_all_study_plot + pe_all_1lab_plot + pe_all_pres33_plot + pe_all_abund005_plot +
  plot_layout(nrow = 1)  + 
  plot_layout(guides = "collect")

pe_plot

```


```{r}

pe_all_abund005_all_filt_1.euc.mat <- as.matrix(pe_all_abund005_all_filt_1.euc)

selected_metadata <- world_metadata %>%
  full_join(., pacific_metadata) %>%
  select(Library_ID, Study, Lab, Age_mean) %>%
  # separate(Library_ID, into = "Library_ID", extra = "drop", sep = "\\.") %>%
  # mutate(Library_ID = str_replace_all(Library_ID, "-00-01","")) %>%
  # separate(Library_ID, into = "Library_ID", extra = "drop", sep = "_") %>%
  inner_join(., pe_all_abund005_all_filt_1.euc.mat %>%
               as.data.frame(.) %>%
               rownames_to_column("Library_ID") %>%
               select(Library_ID), by = "Library_ID") %>%
  # mutate(Library_ID = str_replace_all(Library_ID, "_"," ")) %>%
  column_to_rownames("Library_ID")

col_age = circlize::colorRamp2(c(0, 3000), c("#098BD9", "#7D3560"))
col_fun = circlize::colorRamp2(c(0, 50, 104), c("#ffffff", "#77c8f9", "#098BD9"))


row_ha <- ComplexHeatmap::HeatmapAnnotation(df = selected_metadata, col = list(Study = Study_colors,
                                                                               Lab = Lab_colors,
                                                                               Age_mean = col_age))

ComplexHeatmap::Heatmap(pe_all_abund005_all_filt_1.euc.mat, name = "EucDist", col = col_fun, clustering_method_rows = "complete", clustering_method_columns = "complete", row_dend_side = "left", top_annotation = row_ha, column_names_gp = grid::gpar(fontsize = 4), row_names_gp = grid::gpar(fontsize = 4))



```



Biplots
```{r unfilt_biplot}
# PC1
sp_unfilt_1_pc1 <- plot_pca_bi(pe_all_abund005_filt_1.pca, "PC1", "PC2", "Lab", PC1) # pe_all_pres33_filt_1.pca pe_all_1lab_filt_1.pca pe_all_abund005_filt_1.pca
sp_unfilt_1_pc1

# PC2
sp_unfilt_1_pc2 <- plot_pca_bi(pe_all_abund005_filt_1.pca, "PC1", "PC2", "Lab", PC2)
sp_unfilt_1_pc2

```










