---
title: "Species-level SourceTracker for Pacific calculus"
author: "Irina Velsko"
date: "`r format(Sys.time(), '%b %d, %Y')`"
output:
    github_document:
        fig_width: 8
        fig_height: 6
        toc: true
---


```{r load_libraries, echo = F, message = F}
library(knitr)
library(data.table)
library(mixOmics)
library(janitor)
library(tidyverse)
library(gplots)
library(viridis)
library(patchwork)
opts_chunk$set(echo=F, warning=F, message=F, dpi=150)
```

Set the working directory of the notebook to the parent directory of the 
notebook itself.  

```{r setup}
knitr::opts_knit$set(root.dir = normalizePath(".."))
```

# MALT read stats

```{r load_data, eval = F}

# load the taxonomy tables that have been decontaminated and the column headers are cleaned up
load("./05-results.backup/Taxonomy_tables.RData")


# load the metadata file
load("./05-results.backup/Metadata_tables.RData")

# pacific_metadata <- pacific_metadata %>%
#   mutate(`DNA_molecules_per_library_x10^6_log` = log10(`DNA_molecules_per_library_x10^6`))

outliers <- c("EXB059.A2501","LIB058.A0103","LIB058.A0106","CSS","CSD")
poor_samples <- fread("./05-results.backup/cuperdec_poor_samples.tsv") %>%
  pull(Sample)

```

```{r load_sourcetracker_out}
full_source_results <- fread("./05-Documentation.backup/sink_predictions_species.txt")

pacific_metadata <- fread("05-Documentation.backup/oceania_metadata_full.tsv")

mapfile <- fread("./05-Documentation.backup/source_tracker_mappingfile_20220722.tsv")

```

```{r poorly_preserved}
cfdp_fail <- fread("05-Documentation.backup/cfdp_fail_samples_20210413.txt") %>%
  rename(Library_ID = 1)

```

```{r set_colors}
full_colors <- c(modernCalc = "#7d3560", subPlaque = "#a1527f", supPlaque = "#a1527f", urbanGut = "#098bd9",  ruralGut = "#098bd9", skin = "#ffd5af", sediment = "#1ec1a6", archBone = "#148f77", Unknown = "#616161")

reduced_colors <- c(modernCalc = "#7d3560", Plaque = "#CC79A7", Gut = "#098bd9", skin = "#ffd5af", sediment = "#1ec1a6", archBone = "#148f77", Unknown = "#616161") # plaque a1527f

Env_colors <- c(`Anc. calculus` = "#4b1f39", `Mod. calculus` = "#7d3560", Plaque = "#CC79A7", Gut = "#098bd9", Skin = "#ffd5af", Sediment = "#1ec1a6", `Arch. bone` = "#148f77", `Pacific calculus` = "#098bd9", Blank = "#d0d0d0")

# ancient calculus "#642a4c" or "#702f56"
```


Get the samples ordered by decreasing amount of modern calculus
```{r}
samples_order_full <- full_source_results %>%
  rename(Library_ID = SampleID) %>%
  full_join(., pacific_metadata %>%
              select(Library_ID, Island, Type), by = "Library_ID") %>%
  filter(Type == "calculus") %>%
  arrange(Island, desc(modernCalculus)) %>%
  filter(Island != "Taiwan") %>%
  pull(Library_ID)


```


Plot bar charts
```{r}

st_full_plot_all <- full_source_results %>%
  # filter(!str_detect(SampleID, outliers %>% str_c(collapse = "|"))) %>%
  # filter(!str_detect(SampleID, poor_samples %>% str_c(collapse = "|"))) %>%
  pivot_longer(!SampleID, names_to = "Source", values_to = "Percent") %>%
  # combine plaque and combine gut 
  mutate(Source = str_replace_all(Source, "subPlaque","Plaque"),
         Source = str_replace_all(Source, "supPlaque","Plaque"),
         Source = str_replace_all(Source, "urbanGut","Gut"),
         Source = str_replace_all(Source, "ruralGut","Gut"),
         Source = str_replace_all(Source, "modernCalculus","modernCalc"))  %>%
  group_by(SampleID, Source) %>%
  summarize(Totals = sum(Percent)) %>%
  ungroup() %>%
  mutate(Percent = Totals * 100) %>%
  mutate(Source = fct_relevel(Source, "Unknown","archBone","sediment","skin","Gut","Plaque","modernCalc"),
         SampleID = fct_relevel(SampleID, samples_order_full)) %>%
  left_join(., pacific_metadata %>%
              select(Library_ID, Island) %>%
              rename(SampleID = Library_ID), by = "SampleID") %>%
  ggplot(., aes(x = SampleID, y = Percent, fill = Source)) +
         geom_bar(stat = "identity") + # , color = "black"
         theme_minimal(base_size = 14) +
         scale_fill_manual(values = reduced_colors) +
         theme(axis.text.x = element_text(angle = 90, hjust = 1),
               axis.text = element_text(size = 8)) +
         theme(panel.grid.major.x = element_blank()) +
         # geom_hline(yintercept=75, linetype="dotted") +
         # geom_hline(yintercept=90, linetype="dotted", color = "gray50") +
         # scale_y_continuous(breaks=scales::pretty_breaks(n=6)) +
         ylab("% attributed to source") +
         ggtitle("") +
         # facet_wrap(~Island, scales = "free_x", nrow = 2)
         facet_grid(~Island, scales = "free_x", space = 'free')
st_full_plot_all

# ggsave("./06-publication/prelim_figs/st_full_plot_all.svg", plot = st_full_plot_all, device = "svg",
#        scale = 1, width = 16, height = 5, units = c("in"), dpi = 300)

```


Are the sampls with poor preservation here the same as those identified with cuperdec?
```{r}
# samples with oral content under 50% - 25
oral_under_50 <- full_source_results %>%
  rename(Library_ID = SampleID) %>%
  full_join(., pacific_metadata %>%
              select(Library_ID, Site, Type), by = "Library_ID") %>%
  filter(Type == "calculus") %>%
  mutate(Oral = modernCalculus + subPlaque + supPlaque) %>%
  filter(Oral <= 0.5) %>%
  select(Library_ID)
  
# cuperdec failed samples
# genus - 38
cfdp_fail_gn <- fread("05-Documentation.backup/cfdp_fail_samples_20210413.txt") %>%
  rename(Library_ID = 1)

# species - 30
cfdp_fail_sp <- fread("05-Documentation.backup/oceania_cuperdec_species_poor_samples.tsv") %>%
  rename(Library_ID = 1)

# overlap between genus failed (Zandra's) and species failed
# all species fail are in genus fail
cfdp_fail_gn %>%
  inner_join(., cfdp_fail_sp)

# overlapping samples (all sourcetracker under 50 are in the cfpd fail list)
cfdp_fail_sp %>%
  inner_join(., oral_under_50)


# oral percent from sourcetracker for all cuperdec failed samples
full_source_results %>%
  rename(Library_ID = SampleID) %>%
  pivot_longer(!Library_ID, names_to = "Source", values_to = "Percent") %>%
  # combine plaque and combine gut 
  mutate(Source = str_replace_all(Source, "subPlaque","Plaque"),
         Source = str_replace_all(Source, "supPlaque","Plaque"),
         Source = str_replace_all(Source, "urbanGut","Gut"),
         Source = str_replace_all(Source, "ruralGut","Gut"),
         Source = str_replace_all(Source, "modernCalculus","modernCalc"))  %>%
  group_by(Library_ID, Source) %>%
  summarize(Totals = sum(Percent)) %>%
  ungroup() %>%
  mutate(Percent = Totals * 100) %>%
  select(-Totals) %>%
  pivot_wider(names_from = "Source", values_from = Percent, values_fill = 0) %>%
  inner_join(., cfdp_fail_sp) %>%
  arrange(desc(modernCalc))
  

```



Let's look at the species going into each source. Start with the reduced sources for simplicity
```{r, eval = F}

modernCalc_sources <- fread("~/archgen/microbiome_calculus/smoking_calculus/04-Analysis/sourcetracker/shotgun/species/shotgun_sourcetracker_species/full_results/sink_predictions_modernCalculus_contributions.txt") %>%
  rename(Library_ID = SampleID) %>%
  gather("Species","Percent",2:ncol(.)) %>%
  spread(Library_ID, Percent) %>%
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

plaque_sources <- fread("~/archgen/microbiome_calculus/smoking_calculus/04-Analysis/sourcetracker/shotgun/species/shotgun_sourcetracker_species/full_results/sink_predictions_plaque_contributions.txt") %>%
  rename(Library_ID = SampleID) %>%
  gather("Species","Percent",2:ncol(.)) %>%
  spread(Library_ID, Percent) %>%
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

stool_sources <- fread("~/archgen/microbiome_calculus/smoking_calculus/04-Analysis/sourcetracker/shotgun/species/shotgun_sourcetracker_species/full_results/sink_predictions_stool_contributions.txt") %>%
  rename(Library_ID = SampleID) %>%
  gather("Species","Percent",2:ncol(.)) %>%
  spread(Library_ID, Percent) %>%
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

skin_sources <- fread("~/archgen/microbiome_calculus/smoking_calculus/04-Analysis/sourcetracker/shotgun/species/shotgun_sourcetracker_species/full_results/sink_predictions_skin_contributions.txt") %>%
  rename(Library_ID = SampleID) %>%
  gather("Species","Percent",2:ncol(.)) %>%
  spread(Library_ID, Percent) %>%
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

sediment_sources <- fread("~/archgen/microbiome_calculus/smoking_calculus/04-Analysis/sourcetracker/shotgun/species/shotgun_sourcetracker_species/full_results/sink_predictions_sediment_contributions.txt") %>%
  rename(Library_ID = SampleID) %>%
  gather("Species","Percent",2:ncol(.)) %>%
  spread(Library_ID, Percent) %>%
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

bone_sources <- fread("~/archgen/microbiome_calculus/smoking_calculus/04-Analysis/sourcetracker/shotgun/species/shotgun_sourcetracker_species/full_results/sink_predictions_bone_contributions.txt") %>%
  rename(Library_ID = SampleID) %>%
  gather("Species","Percent",2:ncol(.)) %>%
  spread(Library_ID, Percent) %>%
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

unknown_sources <- fread("~/archgen/microbiome_calculus/smoking_calculus/04-Analysis/sourcetracker/shotgun/species/shotgun_sourcetracker_species/full_results/sink_predictions_Unknown_contributions.txt") %>%
  rename(Library_ID = SampleID) %>%
  gather("Species","Percent",2:ncol(.)) %>%
  spread(Library_ID, Percent) %>%
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

```

Overlap between modernCalculus and unknown?
```{r, eval = F}
# only in unknown sources, not also in modernCalculus
unknown_sources %>%
  select(Species) %>%
  anti_join(., modernCalc_sources %>%
              select(Species))

# in both unknown and modernCalculus
unknown_sources %>%
  select(Species) %>%
  inner_join(., modernCalc_sources %>%
              select(Species))


```

```{r, eval = F}
unknown_sources %>%
    filter(str_detect(Species, "Anaerovorax")) %>%
  as_tibble()

modernCalc_sources %>%
    filter(str_detect(Species, "Anaerovorax")) %>%
  as_tibble()


```


## Source PCA

```{r}
# read in full table for sourcetracker
source_table <- fread("05-Documentation.backup/oceania_sourcetracker_table_species.txt") %>%
  rename(Species = 1)

# read in rarefied table for sourcetracker
source_table_rarefied <- fread("05-Documentation.backup/oceania_sourcetracker_table_species_rarefied.txt") %>%
  rename(Species = 1)

# read in rarefied blanks table (note LIB027.A0136 was removed b/c it had too few reads)
blank_table_rarefied <- fread("05-Documentation.backup/oceania_sourcetracker_table_blanks_species_rarefied.tsv") %>%
  rename(Species = 1)

# read in poor samples from cuperdec species
poor_samples <- fread("05-Documentation.backup/oceania_cuperdec_species_poor_samples.tsv") %>%
  pull()

```

this has the blanks
```{r}
full_source_table <- fread("./05-Documentation.backup/oceania_malt_cRefSeq_summarized_species_20201105.txt") %>%
  rename(Species = 1) %>%
  full_join(., fread("./05-Documentation.backup/sources_MALT_cRefSeq_JFY_species_summarized.txt") %>%
              rename(Species = 1) %>%
              select(matches("Species|SRR|ERR|ARS")) %>%
              select(-matches("SRR059425|SRR1761692|SRR1761697|SRR1761710|SRR1761718")), by = "Species") %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_blanks_cRefSeq_species_summarized_20201130.txt") %>%
              rename(Species = 1), by = "Species") %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_bones_cRefSeq_species_summarized_20201202.txt") %>%
              rename(Species = 1), by = "Species") %>%
   full_join(., fread("./05-Documentation.backup/deep_evo_malt_refseq_comparison_species.tsv") %>%
              rename(Species = 1) %>%
              select(matches("Species|JAE006|JAE007|JAE008|JAE009|JAE012|VLC001|VLC002|VLC003|VLC004|VLC005")), by = "Species") %>%
 filter(!str_detect(Species, "Homo")) %>%
  # remove HPD samples 
  select(-matches("HPD"))

# clean the file names
colnames(full_source_table) <- gsub("MeganServer::", "", colnames(full_source_table)) 
colnames(full_source_table) <- gsub(".unmapped", "", colnames(full_source_table)) 
colnames(full_source_table) <- gsub(".SG1.1", "", colnames(full_source_table)) 
colnames(full_source_table) <- gsub(".SG1.2", "", colnames(full_source_table)) 
colnames(full_source_table) <- gsub("_L001_R1_001.fastq.combined.fq.prefixed.extractunmapped.bam", "", colnames(full_source_table)) 
colnames(full_source_table) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg19unmapped", "", colnames(full_source_table)) 
colnames(full_source_table) <- gsub("_S0_L001_R1_001.fastq.truncated.prefixed.extractunmapped.bam", "", colnames(full_source_table)) 
colnames(full_source_table) <- gsub("_S0_L001_R1_000.fastq.merged.prefixed.hg1", "", colnames(full_source_table)) 
colnames(full_source_table) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg1", "", colnames(full_source_table)) 
colnames(full_source_table) <- gsub(".SG1.2_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(full_source_table))  
colnames(full_source_table) <- gsub("_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(full_source_table)) 

# replace all NA with 0
full_source_table <- full_source_table %>% 
  replace(is.na(.), 0) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  arrange(Library_ID) %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts")

```

```{r}
# do we need to filter out bones?
bones <- pacific_metadata %>%
  filter(Type == "bone") %>%
  select(Library_ID, Lab, Type) %>%
  filter(!str_detect(Library_ID, "ARS|HPD"))

blanks <- pacific_metadata %>%
  filter(Type == "blank") %>%
  pull(Library_ID)

```

```{r}

full_source_table %>%
  select(matches(c("Species", blanks) %>% str_c(collapse = "|")))

```



Combine the input data table and the source table. This needs to be normalized
by sequencing depth
```{r}

# full_table_sourcetracker <- source_table 
# %>%
#   select(-matches(blanks %>% str_c(collapse = "|"))) 

map_file <- mapfile %>%
  mutate(Env = str_replace_all(Env, "calculus","Pacific calculus"),
         Env = str_replace_all(Env, "modernCalculus","Mod. calculus"),
         Env = str_replace_all(Env, "rural",""),
         Env = str_replace_all(Env, "urban",""),
         Env = str_replace_all(Env, "sediment","Sediment"),
         Env = str_replace_all(Env, "skin","Skin"),
         Env = str_replace_all(Env, "sub",""),
         Env = str_replace_all(Env, "sup",""),
         Env = str_replace_all(Env, "archBone","Arch. bone")) %>%
  rename(Library_ID = 1) %>%
  full_join(., pacific_metadata %>%
              filter(Type == "blank") %>%
              select(Library_ID, Type) %>%
              rename(Env = Type)) %>%
  mutate(Env = str_replace_all(Env, "blank","Blank"))

```

### Add ancient calculus
Load European contaminant tables - note these are only at the Species level (didn't do genus-level analysis)
```{r world_contaminants}
# contaminants
mid_contaminants <- fread("../smoking_calculus/05-results.backup/contaminant_species_mid.tsv", sep = "\t")
iber_contaminants <- fread("../smoking_calculus/05-results.backup/contaminant_species_iber.tsv", sep = "\t")
rad_contaminants <- fread("../smoking_calculus/05-results.backup/contaminant_species_rad.tsv", sep = "\t")
kil_contaminants_freq <- fread("../smoking_calculus/05-results.backup/contaminant_species_freq_kil.tsv", sep = "\t")
ei_contaminants <- fread("./05-Documentation.backup/contaminant_eisenhoffer_sp.tsv", sep = "\t")
ot_contaminants <- fread("./05-Documentation.backup/contaminant_ottoni_sp.tsv", sep = "\t")
de_contaminants <- fread("./05-Documentation.backup/contaminant_deepevo_sp.tsv", sep = "\t")
mw_contaminants <- fread("./05-Documentation.backup/contaminant_mw2018_sp.tsv")

# and the pacific contaminants
cont_all_sp <- fread("./05-Documentation.backup/contaminant_pacific_ooj_sp.tsv", sep = "\t", fill = T) %>%
  select(Species)

```

```{r world_poor}
# poorly-preserved samples
world_fail <- fread("./05-Documentation.backup/cuperdec_malt_ei_ot_de_poor_samples.tsv")


```


```{r world_tables}
# MALT tables
mid_raw <- fread("../smoking_calculus/05-results.backup/MID_malt_bactarchhomo2018_comparison-species.txt")
radcliffe_raw <- fread("../smoking_calculus/05-results.backup/Radcliffe_pfuT_comparison-species.txt")
kilteasheen_raw <- fread("../smoking_calculus/05-results.backup/Kilteasheen_comparison-species.txt")
iberian_raw <- fread("../smoking_calculus/05-results.backup/iberian_medieval_malt_species_raw.tsv")
# mod_raw <- fread("../smoking_calculus/05-results.backup/mod_malt_species_raw.tsv")
ei2020_raw <- fread("./05-Documentation.backup/eisenhoffer2020_malt_refseq_species.tsv")
ot2021_raw <- fread("./05-Documentation.backup/ottoni2021_malt_refseq_species.tsv")
de_raw <- fread("./05-Documentation.backup/deep_evo_malt_refseq_comparison_species.tsv")
mw_raw <- fread("./05-Documentation.backup/mann2018_world_malt_refseq_comparison_species.tsv")

```

```{r world_metadata}
calc_full_metadata <- fread("../smoking_calculus/00-documentation.backup/full_combined_metadata.tsv") %>%
  select(Library_ID, Site_code, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = str_replace_all(Site_code, "MID","Middenbeemster"),
         Study = str_replace_all(Study, "CMB","Middenbeemster"),
         Study = str_replace_all(Study, "KIL","Kilteasheen"),
         Study = str_replace_all(Study, "RAD","Radcliffe"),
         Study = str_replace_all(Study, "ELR","Iberia"),
         Study = str_replace_all(Study, "IVE","Iberia"),
         Lab = str_replace_all(Site_code, "MID","Jena"),
         Lab = str_replace_all(Lab, "CMB","Jena"),
         Lab = str_replace_all(Lab, "ELR","Jena"),
         Lab = str_replace_all(Lab, "IVE","Jena"),
         Lab = str_replace_all(Lab, "RAD","Oxford"),
         Lab = str_replace_all(Lab, "KIL","Tuebingen")) %>%
  mutate(Age_mean = ifelse(str_detect(Site_code, "MID|CMB|RAD"), 250,
                           ifelse(Site_code == "KIL", 700,
                                  ifelse(str_detect(Site_code, "ELR|IVE"), 620, 7)))) %>%
  full_join(., fread("./05-Documentation.backup/eisenhoffer2020_ottoni2021_deepevo_mann2018w_metadata.tsv") %>%
              mutate(Run_accession = ifelse(Study == "FellowsYates2021",Sample_alias,Run_accession)) %>%
              rename(Library_ID = Run_accession)) %>%
  mutate(Sample_type = ifelse(str_detect(Study, "EXB|LIB|DEN|SOL"),"blank",Sample_type)) %>%
  mutate(Sample_type = ifelse(is.na(Sample_type),"calculus",Sample_type)) %>%
  mutate(Sample_group = ifelse(Sample_type == "blank","blank",
                               ifelse(str_detect(Library_ID, "VLC|JAE"),"modern_calculus",Sample_group))) %>%
  mutate(Sample_group = ifelse(is.na(Sample_group),"ancient_calculus",Sample_group)) %>%
  mutate(Sample_alias = ifelse(is.na(Sample_alias),Library_ID,Sample_alias)) %>%
  mutate(Age_mean_log = log10(Age_mean))

world_metadata <- calc_full_metadata %>%
  # mutate(Sample_group = ifelse(is.na(Sample_group),"bone",Sample_group)) %>%
  filter(!str_detect(Sample_group, "primate|modern"))

# list the non-human primates and modern calculus to remove from the species table
nhp <- calc_full_metadata %>%
  filter(str_detect(Sample_group, "primate|modern")) %>%
  select(Sample_alias) %>%
  unique() %>%
  pull()

```


```{r plot_pca_fxn}

# add functions for:
# making the input table for PCA?

# plotting PCA with colored dots
plot_pca <- function(df, pc1, pc2, color_group) {
    metadata_group_colors <- get(paste(color_group, "_colors", sep = ""))

    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%")
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              full_join(map_file, by = "Library_ID") %>%
              # the woorld samples won't have an Env, so make one
              mutate(Env = ifelse(is.na(Env),"Anc. calculus",Env))

    color_group = df_X[[color_group]]

                       if (pc1 == 'PC1') {
                           pc1 <- df_X$PC1
                           exp_var_pc1 <- exp_var[1]
                           xaxis <- c("PC1")
                       }  else if (pc1 == 'PC2') {
                           pc1 <- df_X$PC2
                           exp_var_pc1 <- exp_var[2]
                           xaxis <- c("PC2")
                       } else if (pc1 == 'PC3') {
                           pc1 <- df_X$PC3
                           exp_var_pc1 <- exp_var[3]
                           xaxis <- c("PC3")
                       }

                       if (pc2 == 'PC1') {
                           pc2 <- df_X$PC1
                           exp_var_pc2 <- exp_var[1]
                           yaxis <- c("PC1")
                       }  else if (pc2 == 'PC2') {
                           pc2 <- df_X$PC2
                           exp_var_pc2 <- exp_var[2]
                           yaxis <- c("PC2")
                       } else if (pc2 == 'PC3') {
                           pc2 <- df_X$PC3
                           exp_var_pc2 <- exp_var[3]
                           yaxis <- c("PC3")
                       }

    pca_plot <- ggplot(df_X, aes(pc1, pc2)) +
                        geom_point(aes(fill = color_group, shape = shape_group), size = 4, stroke = 0.3, shape = 21) +
                        scale_fill_manual(values = metadata_group_colors) +
                        xlab(paste(xaxis, " - ", exp_var_pc1)) +
                        ylab(paste(yaxis, " - ", exp_var_pc2)) +
                        theme_minimal(base_size = 14) +
                        theme(legend.position = "right") +
                        theme(legend.title = element_blank())

    return(pca_plot)
}

```

```{r}
source_table_1 <- source_table %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0) %>%
  # filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pca(source_table_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
source_table_1.pca <- pca(source_table_1, ncomp = 3, logratio = 'CLR')

# now remove cuperdec failed samples
source_table_1_good <- source_table %>%
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0) %>%
  # filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pca(source_table_1_good, logratio = 'CLR')

# perform a PCA to see how the data cluster
source_table_1_good.pca <- pca(source_table_1_good, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(source_table_1.pca, "PC1", "PC2", "Env")
plot_pca(source_table_1_good.pca, "PC1", "PC2", "Env")
plot_pca(source_table_1.pca, "PC1", "PC3", "Env")
plot_pca(source_table_1_good.pca, "PC1", "PC3", "Env")
plot_pca(source_table_1.pca, "PC2", "PC3", "Env")
plot_pca(source_table_1_good.pca, "PC2", "PC3", "Env")

```


```{r}
source_table_rarefied_1 <- source_table_rarefied %>%
  full_join(., blank_table_rarefied, by = "Species") %>%
  replace(is.na(.), 0) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pca(source_table_rarefied_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
source_table_rarefied_1.pca <- pca(source_table_rarefied_1, ncomp = 3, logratio = 'CLR')

# now remove cuperdec failed samples
source_table_rarefied_1_good <- source_table_rarefied %>%
  full_join(., blank_table_rarefied, by = "Species") %>%
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  replace(is.na(.), 0) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pca(source_table_rarefied_1_good, logratio = 'CLR')

# perform a PCA to see how the data cluster
source_table_rarefied_1_good.pca <- pca(source_table_rarefied_1_good, ncomp = 3, logratio = 'CLR')

# plot the PCA
plot_pca(source_table_rarefied_1.pca, "PC1", "PC2", "Env")
plot_pca(source_table_rarefied_1_good.pca, "PC1", "PC2", "Env")
plot_pca(source_table_rarefied_1.pca, "PC1", "PC3", "Env")
plot_pca(source_table_rarefied_1_good.pca, "PC1", "PC3", "Env")
plot_pca(source_table_rarefied_1.pca, "PC2", "PC3", "Env")
plot_pca(source_table_rarefied_1_good.pca, "PC2", "PC3", "Env")

```

```{r}
full_source_table_1 <- full_source_table %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pca(full_source_table_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
full_source_table_1.pca <- pca(full_source_table_1, ncomp = 3, logratio = 'CLR')

# now remove cuperdec failed samples
full_source_table_1_good <- full_source_table %>%
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0) %>%
  filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pca(full_source_table_1_good, logratio = 'CLR')

# perform a PCA to see how the data cluster
full_source_table_1_good.pca <- pca(full_source_table_1_good, ncomp = 3, logratio = 'CLR')

# plot the PCA
fs_all_pc12 <- plot_pca(full_source_table_1.pca, "PC1", "PC2", "Env")
fs_all_pc12
fs_good_pc12 <- plot_pca(full_source_table_1_good.pca, "PC1", "PC2", "Env")
fs_good_pc12
fs_all_pc13 <- plot_pca(full_source_table_1.pca, "PC1", "PC3", "Env")
fs_all_pc13
fs_good_pc13 <- plot_pca(full_source_table_1_good.pca, "PC1", "PC3", "Env")
fs_good_pc13
plot_pca(full_source_table_1.pca, "PC2", "PC3", "Env")
plot_pca(full_source_table_1_good.pca, "PC2", "PC3", "Env")

```

### PCA with other ancient calculus
```{r pacific_world_all_decontam}
# combine raw tables and decontaminate with a list of all contaminants in all samples
pe_all_table <- full_source_table %>%
  full_join(., mid_raw %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks
            select(-matches("EXB|LIB")) %>%
  full_join(., radcliffe_raw %>%
             rename(Species = `#Datasets`) %>%
           # remove blanks
            select(-`CSD.unmapped`, -`CSL.unmapped`, -`CSN.unmapped`, -`CSS.unmapped`)) %>%
  full_join(., kilteasheen_raw %>%
            rename(Species = `#Datasets`)) %>%
  full_join(., mw_raw %>%
            rename(Species = `#Datasets`)) %>%
  full_join(., iberian_raw %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks
            select(-matches("EXB|LIB"))) %>%
  full_join(., ei2020_raw %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks
            select(-`SRR11176637.unmapped`, -`SRR11176643.unmapped`)) %>%
  full_join(., ot2021_raw %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks
            select(-`ERR5729658.unmapped`, -`ERR5729659.unmapped`, -`ERR5729660.unmapped`, -`ERR5729661.unmapped`)) %>%
  full_join(., de_raw %>%
            rename(Species = `#Datasets`) %>%
            select(-matches(nhp %>% str_c(collapse = "|")))) %>%
            # remove blanks, dentin and soil
            select(-matches("CSS|CSD|EXB|LIB|CSN|CSL")), by = "Species") %>%
  anti_join(.,  mid_contaminants %>%
            bind_rows(., rad_contaminants) %>%
            bind_rows(., kil_contaminants_freq) %>%
            bind_rows(., iber_contaminants) %>%
            bind_rows(., ei_contaminants) %>%
            bind_rows(., ot_contaminants) %>%
            bind_rows(., de_contaminants) %>%
            bind_rows(., mw_contaminants) %>%
            bind_rows(., cont_all_sp) %>% # pacific contaminants already combinded and de-duplicated
            distinct(), by = "Species") %>%
  replace(is.na(.), 0) %>%
  # remove any species that have no counts after removing the blanks
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

# remove the totals attribute so the column totals can be calculated again later
attr(pe_all_table, "totals") <- NULL

colnames(pe_all_table) <- gsub(".unmapped","", colnames(pe_all_table))
colnames(pe_all_table) <- gsub("MeganServer::","", colnames(pe_all_table))

pe_all_table <- pe_all_table %>%
  # remove poorly-preserved samples
  select(-matches(world_fail %>% pull(Sample) %>% str_c(collapse = "|"))) %>%
  # remove any species that have no counts after removing the poorly-preserved samples
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

# remove the totals attribute so the column totals can be calculated again later
attr(pe_all_table, "totals") <- NULL

```


```{r with_calculus}
# remove Gut samples
gut <- map_file %>%
  filter(Env == "Gut") %>%
  select(Library_ID) %>%
  pull()

full_source_table_1 <- pe_all_table %>%
  select(-matches(gut %>% str_c(collapse = "|"))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0) %>%
  # filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pca(full_source_table_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
full_source_table_1.pca <- pca(full_source_table_1, ncomp = 3, logratio = 'CLR')

# now remove cuperdec failed samples
full_source_table_1_good <- pe_all_table %>%
  select(-matches(gut %>% str_c(collapse = "|"))) %>%
  select(-matches(poor_samples %>% str_c(collapse = "|"))) %>%
  adorn_totals(where = "col") %>%
  mutate(Percent = Total / sum(Total) * 100) %>%
  filter(Percent >= 0) %>%
  # filter(Percent >= 0.005) %>%
  select(-Total, -Percent) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pca(full_source_table_1_good, logratio = 'CLR')

# perform a PCA to see how the data cluster
full_source_table_1_good.pca <- pca(full_source_table_1_good, ncomp = 3, logratio = 'CLR')

# plot the PCA
fs_all_pc12 <- plot_pca(full_source_table_1.pca, "PC1", "PC2", "Env")
fs_all_pc12
fs_good_pc12 <- plot_pca(full_source_table_1_good.pca, "PC1", "PC2", "Env")
fs_good_pc12
fs_all_pc13 <- plot_pca(full_source_table_1.pca, "PC1", "PC3", "Env")
fs_all_pc13
fs_good_pc13 <- plot_pca(full_source_table_1_good.pca, "PC1", "PC3", "Env")
fs_good_pc13
plot_pca(full_source_table_1.pca, "PC2", "PC3", "Env")
plot_pca(full_source_table_1_good.pca, "PC2", "PC3", "Env")

```

```{r}

pca_plot <- fs_all_pc12 + fs_good_pc12 + fs_all_pc13 + fs_good_pc13 +
  plot_layout(nrow = 2)  +
  # plot_annotation(tag_levels = 'A') +
  plot_layout(guides = "collect")

pca_plot

# ggsave("./06-publication/prelim_figs/source_pca_species.svg", plot = pca_plot, device = "svg",
#        scale = 1, width = 10, height = 7, units = c("in"), dpi = 300)


```








