---
title: "Pacific and world calculus humann3 gene family analysis"
author: "Irina Velsko"
date: "`r format(Sys.time(), '%b %d, %Y')`"
output:
    github_document:
        fig_width: 8
        fig_height: 6
        toc: true
---


```{r load_libraries, echo = F, message = F}
library(knitr)
# library(decontam)
library(data.table)
library(mixOmics)
library(FAVA)
library(janitor)
library(rstatix)
library(pander)
library(variancePartition)
library(vegan)
library(tidyverse)
library(gplots)
library(ggrepel)
library(pals)
library(viridis)
library(patchwork)
library(RColorBrewer)
library(microshades)
opts_chunk$set(echo=F, warning=F, message=F, dpi=150)
```

Set the working directory of the notebook to the parent directory of the 
notebook itself.  

```{r setup}
knitr::opts_knit$set(root.dir = normalizePath("../../.."))
```



```{r metadata}
# load the metadata file
pacific_metadata <- fread("05-Documentation.backup/oceania_metadata_full.tsv") %>%
  select(Library_ID, Lab, Type, Age_mean, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = "Pacific",
          Continent = "Pacific",
         Age_mean_log = log10(Age_mean))

pc_end_samples <- fread("./05-Documentation.backup/pc_end_samples.tsv") %>%
  pull(Species)


     
```

```{r}
metadata <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/00-documentation.backup/full_combined_metadata.tsv") %>%
  select(Library_ID, Site_code, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = str_replace_all(Site_code, "MID","Middenbeemster"),
         Study = str_replace_all(Study, "CMB","Middenbeemster"),
         Study = str_replace_all(Study, "KIL","Kilteasheen"),
         Study = str_replace_all(Study, "RAD","Radcliffe"),
         Study = str_replace_all(Study, "ELR","Iberia"),
         Study = str_replace_all(Study, "IVE","Iberia"),
         Lab = str_replace_all(Site_code, "MID","Jena"),
         Lab = str_replace_all(Lab, "CMB","Jena"),
         Lab = str_replace_all(Lab, "ELR","Jena"),
         Lab = str_replace_all(Lab, "IVE","Jena"),
         Lab = str_replace_all(Lab, "RAD","Oxford"),
         Lab = str_replace_all(Lab, "KIL","Tuebingen")) %>%
  mutate(Age_mean = ifelse(str_detect(Site_code, "MID|CMB|RAD"), 250,
                           ifelse(Site_code == "KIL", 700,
                                  ifelse(str_detect(Site_code, "ELR|IVE"), 620, 7)))) %>%
  mutate(Continent = "Europe") %>%
  full_join(., fread("./05-Documentation.backup/eisenhoffer2020_ottoni2021_deepevo_mann2018w_metadata.tsv") %>%
              mutate(Run_accession = ifelse(Study == "FellowsYates2021",Sample_alias,Run_accession)) %>%
              rename(Library_ID = Run_accession)) %>%
  mutate(Sample_type = ifelse(str_detect(Study, "EXB|LIB|DEN|SOL"),"blank",Sample_type)) %>%
  mutate(Sample_type = ifelse(is.na(Sample_type),"calculus",Sample_type)) %>%
  mutate(Sample_group = ifelse(Sample_type == "blank","blank",
                               ifelse(str_detect(Library_ID, "VLC|JAE"),"modern_calculus",Sample_group))) %>%
  mutate(Sample_group = ifelse(is.na(Sample_group),"ancient_calculus",Sample_group)) %>%
  mutate(Sample_alias = ifelse(is.na(Sample_alias),Library_ID,Sample_alias)) %>%
  mutate(Age_mean_log = log10(Age_mean))

world_metadata <- metadata %>%
  # mutate(Sample_group = ifelse(is.na(Sample_group),"bone",Sample_group)) %>%
  filter(!str_detect(Sample_group, "primate|modern"))

# list the non-human primates and modern calculus to remove from the species table
nhp <- metadata %>%
  filter(str_detect(Sample_group, "primate|modern")) %>%
  select(Sample_alias) %>%
  unique() %>%
  pull()

# and combine pacific and world
full_metadata <- pacific_metadata %>%
  bind_rows(., world_metadata)


```


```{r}
# make a list of blanks b/c they don't all have the same prefixes
blanks <- full_metadata %>%
  mutate(Type = ifelse(is.na(Type),Sample_type,Type)) %>%
  filter(str_detect(Type, "blank|bone")) %>%
  filter(!str_detect(Library_ID, "ARS")) %>%
  select(Library_ID) %>%
  pull()

```


```{r cuperdec_poor_samples}
poor_samples <- fread("./05-Documentation.backup/cfdp_fail_samples_20210413.txt") %>%
  pull(Sample)

world_fail <- fread("./05-Documentation.backup/cuperdec_malt_ei_ot_de_poor_samples.tsv") %>%
  bind_rows(., fread("05-Documentation.backup/oceania_cuperdec_mpa3_anc_params_poor_samples.tsv"))

```

```{r colors}
Island_colors = c("Tongatapu" = "#616161", 
                  "Flores" = "#f6865e", 
                  "Rapa Nui" = "#cc79a7", 
                  "Futuna" = "#7D3560", 
                  "Efate" = "#7472af", 
                  "Vao" = "#098BD9",
                  "Efate 3000 BP" = "#7DCCFF", 
                  "Uripiv" = "#e1f2fe", 
                  "Viti Levu" = "#1ec1a7", 
                  "Raiatea" = "#148f77", 
                  "Taumako" = "#4e7705", 
                  "Watom" = "#86c73a")


# Set microshades colors
# Set microshades colors
Lab_colors = c("Jena" = microshades_palette("micro_cvd_purple", 3, lightest = FALSE)[3],
                "Oklahoma" =  microshades_palette("micro_cvd_blue", 3, lightest = FALSE)[3],
                "Otago" = microshades_palette("micro_cvd_green", 3, lightest = FALSE)[3],
                "Oxford" = microshades_palette("micro_cvd_turquoise", 3, lightest = FALSE)[1],
                "Tuebingen" = microshades_palette("micro_cvd_purple", 3, lightest = FALSE)[1],
                "ACAD" = microshades_palette("micro_cvd_blue", 4, lightest = FALSE)[1],
                "Vienna" = microshades_palette("micro_cvd_green", 3, lightest = FALSE)[1])


Sample_type_colors <- c("Arch. Bone" = "#148F77", "blank" = "#8B8B8B", "Arch. calculus" = "black")

Study_colors <- c("Pacific" = "#098BD9", 
                  "Middenbeemster" = "#7472af", 
                  "Radcliffe" = "#cc79a7", 
                  "Kilteasheen" = "#f6865e", 
                  "Iberia" = "#1ec1a7",
                  "Eisenhoffer2020" = "#86c73a",
                  "Ottoni2021" = "#616161",
                  "FellowsYates2021" = "#e1f2fe",
                  "Mann2018" = "#f5c710ff") # 7D3560


Continent_colors <- c("Pacific" = "#098BD9", 
                  "Europe" = "#7472af", 
                  "Africa" = "#cc79a7", 
                  "Caribbean" = "#f6865e", 
                  "Asia" = "#86c73a",
                  "North America" = "#1ec1a7")


```

```{r}
Island_shapes = c("Tongatapu" = 21, 
                  "Flores" = 21, 
                  "Rapa Nui" = 21, 
                  "Futuna" = 21, 
                  "Efate" = 21, 
                  "Vao" = 21,
                  "Efate 3000 BP" = 21, 
                  "Uripiv" = 21, 
                  "Viti Levu" = 21, 
                  "Raiatea" = 21, 
                  "Taumako" = 21, 
                  "Watom" = 21)

Lab_shapes = c(21,21,21,21,21,21,21,21)
Lab_size = c(4,4,4,4,4,4,4)

Study_shapes = c(21,21,21,21,21,21,21,21,21,21,21)

Type_shapes = c(21,21,21,21,21,21,21,21)
Type_size = c(4,4,4,4,4,4,4,4)

```


```{r plot_pca_fxn}

# add functions for:
# making the input table for PCA?

# plotting PCA with colored dots
plot_pca <- function(df, pc1, pc2, color_group, shape_group) {
    metadata_group_colors <- get(paste(color_group, "_colors", sep = ""))
    metadata_group_shapes <- get(paste(shape_group, "_shapes", sep = ""))

    exp_var <- paste0(round(df$prop_expl_var$X * 100, 2), "%")
    df_X <- df$variates$X %>%
              as.data.frame() %>%
              rownames_to_column("Library_ID") %>%
              inner_join(full_metadata, by = "Library_ID")
    
    color_group = df_X[[color_group]]
    shape_group = df_X[[shape_group]]

                       if (pc1 == 'PC1') {
                           pc1 <- df_X$PC1
                           exp_var_pc1 <- exp_var[1]
                           xaxis <- c("PC1")
                       }  else if (pc1 == 'PC2') {
                           pc1 <- df_X$PC2
                           exp_var_pc1 <- exp_var[2]
                           xaxis <- c("PC2")
                       } else if (pc1 == 'PC3') {
                           pc1 <- df_X$PC3
                           exp_var_pc1 <- exp_var[3]
                           xaxis <- c("PC3")
                       }

                       if (pc2 == 'PC1') {
                           pc2 <- df_X$PC1
                           exp_var_pc2 <- exp_var[1]
                           yaxis <- c("PC1")
                       }  else if (pc2 == 'PC2') {
                           pc2 <- df_X$PC2
                           exp_var_pc2 <- exp_var[2]
                           yaxis <- c("PC2")
                       } else if (pc2 == 'PC3') {
                           pc2 <- df_X$PC3
                           exp_var_pc2 <- exp_var[3]
                           yaxis <- c("PC3")
                       }

    pca_plot <- ggplot(df_X, aes(pc1, pc2)) +
                        geom_point(aes(fill = color_group, shape = shape_group), size = 2.5, stroke = 0.3, shape = 21) +
                        scale_fill_manual(values = metadata_group_colors) +
                        xlab(paste(xaxis, " - ", exp_var_pc1)) +
                        ylab(paste(yaxis, " - ", exp_var_pc2)) +
                        theme_minimal(base_size = 10) +
                        theme(legend.position = "right") +
                        theme(legend.title = element_blank())

    return(pca_plot)
}

```









Pacific poorly preserved samples
```{r poorly_preserved}
cfdp_fail <- fread("05-Documentation.backup/oceania_cuperdec_species_poor_samples.tsv") %>%
  rename(Library_ID = 1)

```


```{r pacific_species_tables}
# read in the species table
pacific_rs_raw_sp <- fread("./05-Documentation.backup/oceania_malt_cRefSeq_summarized_species_20201105.txt") %>%
  rename(Species = 1) %>%
  # remove HPD samples
  select(-matches("HPD"))  %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_blanks_cRefSeq_species_summarized_20201130.txt") %>%
              rename(Species = 1), by = "Species") %>%
  full_join(., fread("./05-Documentation.backup/oceania_malt_bones_cRefSeq_species_summarized_20201202.txt") %>%
              rename(Species = 1), by = "Species") %>%
  filter(!str_detect(Species, "Homo"))

# clean the file names
colnames(pacific_rs_raw_sp) <- gsub("MeganServer::", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub(".unmapped", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub(".SG1.1", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub(".SG1.2", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_L001_R1_001.fastq.combined.fq.prefixed.extractunmapped.bam", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg19unmapped", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_S0_L001_R1_001.fastq.truncated.prefixed.extractunmapped.bam", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_S0_L001_R1_000.fastq.merged.prefixed.hg1", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub("_S0_L000_R1_000.fastq.merged.prefixed.hg1", "", colnames(pacific_rs_raw_sp)) 
colnames(pacific_rs_raw_sp) <- gsub(".SG1.2_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(pacific_rs_raw_sp))  
colnames(pacific_rs_raw_sp) <- gsub("_S0_L003_R1_001.fastq.combined.fq.prefixed.extrac.bam", "", colnames(pacific_rs_raw_sp)) 

# now do a pretend transpose to replace the NAs with 0s
pacific_rs_raw_sp <- pacific_rs_raw_sp %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = replace_na(Counts, 0)) %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts")

```


```{r pacific_outliers}
# HCLVMBCX2-3505-13-00-01_S13 appears to be a sample, although it's supposed to be a blank
# remove it from all further analyses

metadata_filt <- pacific_metadata %>%
  anti_join(., cfdp_fail) %>%
  filter(!str_detect(Library_ID, "01_S13"))

pacific_species_raw_pass <- pacific_rs_raw_sp %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  pivot_wider(names_from = "Species", values_from = "Counts") %>%
  anti_join(., cfdp_fail) %>%
  # remove blanks
  filter(!str_detect(Library_ID, "01_S13")) %>%
  # remove all taxa that have to entries anymore
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  adorn_totals(where = "col") %>%
  select(Total, everything()) %>%
  filter(Total != 0) %>%
  select(-Total) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  pivot_wider(names_from = "Species", values_from = "Counts")

```

## Prep Pacific data
Load Pacific contaminant tables 
```{r load_contam_lists}
# genus lists
cont_otago_gn <- fread("./05-Documentation.backup/contaminant_otago_gn.tsv")
cont_oklahoma_gn <- fread("./05-Documentation.backup/contaminant_oklahoma_gn.tsv")
cont_jena_gn <- fread("./05-Documentation.backup/contaminant_jena_gn.tsv")

cont_all_gn <- fread("./05-Documentation.backup/contaminant_pacific_ooj_gn.tsv")

# species lists
cont_otago_sp <- fread("./05-Documentation.backup/contaminant_otago_sp.tsv", sep = "\t")
cont_oklahoma_sp <- fread("./05-Documentation.backup/contaminant_oklahoma_sp.tsv", sep = "\t")
cont_jena_sp <- fread("./05-Documentation.backup/contaminant_jena_sp.tsv", sep = "\t")

cont_all_sp <- fread("./05-Documentation.backup/contaminant_pacific_ooj_sp.tsv", sep = "\t")

```

Load Pacific metadata
```{r}
pacific_metadata <- fread("05-Documentation.backup/oceania_metadata_full.tsv") %>%
  select(Library_ID, Lab, Type, Age_mean, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = "Pacific",
         Continent = "Pacific",
         Age_mean_log = log10(Age_mean))

```

Load European contaminant tables - note these are only at the Species level (didn't do genus-level analysis)
```{r}
# contaminants
mid_contaminants <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/05-results.backup/contaminant_species_mid.tsv", sep = "\t")
iber_contaminants <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/05-results.backup/contaminant_species_iber.tsv", sep = "\t")
rad_contaminants <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/05-results.backup/contaminant_species_rad.tsv", sep = "\t")
kil_contaminants_freq <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/05-results.backup/contaminant_species_kil.tsv", sep = "\t")
ei_contaminants <- fread("./05-Documentation.backup/contaminant_eisenhoffer_sp.tsv", sep = "\t")
ot_contaminants <- fread("./05-Documentation.backup/contaminant_ottoni_sp.tsv", sep = "\t")
de_contaminants <- fread("./05-Documentation.backup/contaminant_deepevo_sp.tsv", sep = "\t")
mw_contaminants <- fread("./05-Documentation.backup/contaminant_mw2018_sp.tsv", sep = "\t")

```

```{r}
# poorly-preserved samples
world_fail <- fread("./05-Documentation.backup/cuperdec_malt_ei_ot_de_poor_samples.tsv")


```


```{r world_tables}
# MALT tables
mid_raw <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/05-results.backup/MID_malt_bactarchhomo2018_comparison-species.txt")
radcliffe_raw <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/05-results.backup/Radcliffe_pfuT_comparison-species.txt")
kilteasheen_raw <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/05-results.backup/Kilteasheen_comparison-species.txt")
iberian_raw <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/05-results.backup/iberian_medieval_malt_species_raw.tsv")
ei2020_raw <- fread("./05-Documentation.backup/eisenhoffer2020_malt_refseq_species.tsv")
ot2021_raw <- fread("./05-Documentation.backup/ottoni2021_malt_refseq_species.tsv")
de_raw <- fread("./05-Documentation.backup/deep_evo_malt_refseq_comparison_species.tsv")
mw_raw <- fread("./05-Documentation.backup/mann2018_world_malt_refseq_comparison_species.tsv")

```

```{r}
metadataW <- fread("https://raw.githubusercontent.com/ivelsko/smoking_calculus/main/00-documentation.backup/full_combined_metadata.tsv") %>%
  select(Library_ID, Site_code, seq_len_avg, gc_avg, No_reads_no_human) %>%
  mutate(Study = str_replace_all(Site_code, "MID","Middenbeemster"),
         Study = str_replace_all(Study, "CMB","Middenbeemster"),
         Study = str_replace_all(Study, "KIL","Kilteasheen"),
         Study = str_replace_all(Study, "RAD","Radcliffe"),
         Study = str_replace_all(Study, "ELR","Iberia"),
         Study = str_replace_all(Study, "IVE","Iberia"),
         Lab = str_replace_all(Site_code, "MID","Jena"),
         Lab = str_replace_all(Lab, "CMB","Jena"),
         Lab = str_replace_all(Lab, "ELR","Jena"),
         Lab = str_replace_all(Lab, "IVE","Jena"),
         Lab = str_replace_all(Lab, "RAD","Oxford"),
         Lab = str_replace_all(Lab, "KIL","Tuebingen")) %>%
  mutate(Age_mean = ifelse(str_detect(Site_code, "MID|CMB|RAD"), 250,
                           ifelse(Site_code == "KIL", 700,
                                  ifelse(str_detect(Site_code, "ELR|IVE"), 620, 7)))) %>%
  mutate(Continent = "Europe") %>%
  full_join(., fread("./05-Documentation.backup/eisenhoffer2020_ottoni2021_deepevo_mann2018w_metadata.tsv") %>%
              mutate(Run_accession = ifelse(Study == "FellowsYates2021",Sample_alias,Run_accession)) %>%
              rename(Library_ID = Run_accession)) %>%
  mutate(Sample_type = ifelse(str_detect(Study, "EXB|LIB|DEN|SOL"),"blank",Sample_type)) %>%
  mutate(Sample_type = ifelse(is.na(Sample_type),"calculus",Sample_type)) %>%
  mutate(Sample_group = ifelse(Sample_type == "blank","blank",
                               ifelse(str_detect(Library_ID, "VLC|JAE"),"modern_calculus",Sample_group))) %>%
  mutate(Sample_group = ifelse(is.na(Sample_group),"ancient_calculus",Sample_group)) %>%
  mutate(Sample_alias = ifelse(is.na(Sample_alias),Library_ID,Sample_alias)) %>%
  mutate(Age_mean_log = log10(Age_mean))

world_metadata <- metadataW %>%
  # mutate(Sample_group = ifelse(is.na(Sample_group),"bone",Sample_group)) %>%
  filter(!str_detect(Sample_group, "primate|modern"))

# list the non-human primates and modern calculus to remove from the species table
nhp <- metadataW %>%
  filter(str_detect(Sample_group, "primate|modern")) %>%
  select(Sample_alias) %>%
  unique() %>%
  pull()

```

```{r world_all_decontam}
# remove all contaminants from all samples
world_species_all_decontam <- mid_raw %>%
  select(-matches("EXB|LIB")) %>%
  full_join(., radcliffe_raw %>%
            # remove blanks
            select(-`CSD.unmapped`, -`CSL.unmapped`, -`CSN.unmapped`, -`CSS.unmapped`)) %>%
  full_join(., kilteasheen_raw) %>%
  full_join(., mw_raw) %>%
  full_join(., iberian_raw %>%
              # remove blanks
              select(-matches("EXB|LIB"))) %>%
  full_join(., ei2020_raw %>%
              # remove blanks
              select(-`SRR11176637.unmapped`, -`SRR11176643.unmapped`)) %>%
  full_join(., ot2021_raw %>%
              # remove blanks
              select(-`ERR5729658.unmapped`, -`ERR5729659.unmapped`, -`ERR5729660.unmapped`, -`ERR5729661.unmapped`)) %>%
  full_join(., de_raw %>%
              select(-matches(nhp %>% str_c(collapse = "|")))) %>%
  rename(Species = `#Datasets`) %>%
  # remove blanks, dentin and soil
  select(-matches("CSS|CSD|EXB|LIB|CSN|CSL")) %>%
  # decontam all species
  anti_join(.,  mid_contaminants %>%
            bind_rows(., rad_contaminants) %>%
            bind_rows(., kil_contaminants_freq) %>%
            bind_rows(., iber_contaminants) %>%
            bind_rows(., ei_contaminants) %>%
            bind_rows(., ot_contaminants) %>%
            bind_rows(., mw_contaminants) %>%
            bind_rows(., de_contaminants) %>%
            distinct(), by = "Species") %>%
  filter(!str_detect(Species, "Homo sapiens|Not assigned")) %>%
  replace(is.na(.), 0) %>%
  # remove any species that have no counts after removing the blanks
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)
# %>%
#   full_join(., mod_raw)


# clean the file names
colnames(world_species_all_decontam) <- gsub(".unmapped","", colnames(world_species_all_decontam))
colnames(world_species_all_decontam) <- gsub("MeganServer::","", colnames(world_species_all_decontam))

world_species_all_decontam <- world_species_all_decontam %>%
  # remove poorly-preserved samples
  select(-matches(world_fail %>% pull(Sample) %>% str_c(collapse = "|")))

```

### Pacific and Euroepan tables together
```{r}
# read in MALT stats and add to the full table
malt_stats <- fread("./05-Documentation.backup/malt_stats_table_pac_world.tsv") %>%
  mutate(Pct_assigned = Assigned / (Mapped + Non_mapped) * 100)


metadata <- pacific_metadata %>%
  full_join(., world_metadata) %>%
  full_join(., malt_stats %>%
              mutate(Library_ID = str_replace_all(Library_ID, ".SG1","")), by = "Library_ID") %>%
  drop_na(Lab)

```

```{r}
# list all bones that are part of this study
# the ARS bones are a little different in the PCA, so we won't use them
bones <- pacific_metadata %>%
  filter(Type == "bone") %>%
  select(Library_ID, Lab, Type) %>%
  filter(!str_detect(Library_ID, "ARS|HPD"))

# list all samples and blanks from Otago
otago <- pacific_metadata %>%
  filter(Lab == "Otago") %>%
  select(Library_ID, Lab, Type) %>%
  filter(Library_ID != "HCLVMBCX2-3505-13-00-01_S13") %>%
  arrange(Library_ID) %>%
  filter(Type != "blank") %>%
  pull(Library_ID)

# list all sampmles and blanks from Oklahoma
oklahoma <- pacific_metadata %>%
  filter(Lab == "Oklahoma") %>%
  select(Library_ID, Lab, Type) %>%
  filter(Type != "bone") %>%
  filter(Type != "blank") %>%
  pull(Library_ID)

# list all samples and blanks from Jena
jena <- pacific_metadata %>%
  filter(Lab == "Jena") %>%
  select(Library_ID, Lab, Type) %>%
  filter(Type != "bone") %>%
  filter(Type != "blank") %>%
  filter(!str_detect(Library_ID, "HPD")) %>%
  pull(Library_ID)

# list all samples to include
pacific_sample_list <- c(otago,oklahoma,jena)

```

```{r pacific_world_all_decontam}
# combine raw tables and decontaminate with a list of all contaminants in all samples
pe_all_table <- pacific_species_raw_pass %>%
  filter(Library_ID %in% pacific_sample_list) %>%
  pivot_longer(!Library_ID, names_to = "Species", values_to = "Counts") %>%
  pivot_wider(names_from = "Library_ID", values_from = "Counts") %>%
  full_join(., mid_raw %>%
            # remove blanks
            select(-matches("EXB|LIB")) %>%
            full_join(., radcliffe_raw %>%
                      # remove blanks
                      select(-`CSD.unmapped`, -`CSL.unmapped`, -`CSN.unmapped`, -`CSS.unmapped`)) %>%
            full_join(., kilteasheen_raw) %>%
            full_join(., mw_raw) %>%
            full_join(., iberian_raw %>%
                      # remove blanks
                      select(-matches("EXB|LIB"))) %>%
            full_join(., ei2020_raw %>%
                      # remove blanks
                      select(-`SRR11176637.unmapped`, -`SRR11176643.unmapped`)) %>%
            full_join(., ot2021_raw %>%
              # remove blanks
              select(-`ERR5729658.unmapped`, -`ERR5729659.unmapped`, -`ERR5729660.unmapped`, -`ERR5729661.unmapped`)) %>%
            full_join(., de_raw %>%
                        select(-matches(nhp %>% str_c(collapse = "|")))) %>%
            rename(Species = `#Datasets`) %>%
            # remove blanks, dentin and soil
            select(-matches("CSS|CSD|EXB|LIB|CSN|CSL")), by = "Species") %>%
  anti_join(.,  mid_contaminants %>%
            bind_rows(., rad_contaminants) %>%
            bind_rows(., kil_contaminants_freq) %>%
            bind_rows(., iber_contaminants) %>%
            bind_rows(., ei_contaminants) %>%
            bind_rows(., ot_contaminants) %>%
            bind_rows(., de_contaminants) %>%
            bind_rows(., mw_contaminants) %>%
            bind_rows(., cont_all_sp) %>% # pacific contaminants already combinded and de-duplicated
            distinct(), by = "Species") %>%
  replace(is.na(.), 0) %>%
  # remove any species that have no counts after removing the blanks
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

# remove the totals attribute so the column totals can be calculated again later
attr(pe_all_table, "totals") <- NULL

colnames(pe_all_table) <- gsub(".unmapped","", colnames(pe_all_table))
colnames(pe_all_table) <- gsub("MeganServer::","", colnames(pe_all_table))

pe_all_table <- pe_all_table %>%
  # remove poorly-preserved samples
  select(-matches(world_fail %>% pull(Sample) %>% str_c(collapse = "|"))) %>%
  # remove any species that have no counts after removing the poorly-preserved samples
  adorn_totals(where = "col") %>%
  filter(Total > 0) %>%
  select(-Total)

# remove the totals attribute so the column totals can be calculated again later
attr(pe_all_table, "totals") <- NULL

```

```{r}
pe_all_table_1 <- pe_all_table %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_table_1_pca <- tune.pca(pe_all_table_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_table_1.pca <- pca(pe_all_table_1, ncomp = 3, logratio = 'CLR')

# plot the PCA
sp_all_continent_plot <- plot_pca(pe_all_table_1.pca, "PC1", "PC2", "Continent", "Lab")
sp_all_continent_plot

```

```{r}
sp_counts <- pe_all_table %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = ifelse(Counts > 0,1,0)) %>%
  pivot_wider(names_from = "Species", values_from = "Counts") %>%
  adorn_totals(where = "col") %>%
  rename(no_species = Total) %>%
  select(Library_ID, no_species) %>%
  as_data_frame()

# fwrite(sp_counts, "./05-Documentation.backup/pe_all_table_species_counts.tsv", quote = F, sep = "\t")

metadata <- metadata %>%
  full_join(., sp_counts, by = "Library_ID")

```

```{r no_ottowia}

pe_all_table_no_ott <- pe_all_table %>%
  filter(!str_detect(Species, "Ottowia")) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_table_no_ott_pca <- tune.pca(pe_all_table_no_ott, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_table_no_ott.pca <- pca(pe_all_table_no_ott, ncomp = 3, logratio = 'CLR')

# plot the PCA
sp_all_continent_no_ott_plot <- plot_pca(pe_all_table_no_ott.pca, "PC1", "PC2", "Continent", "Lab")
sp_all_continent_no_ott_plot

# ggsave("./06-publication/supplemental_figures/Sup_fig_7/sp_all_continent_no_ott_plot.pdf", plot = sp_all_continent_no_ott_plot, device = "pdf",
#        scale = 1, width = 8, height = 4.5, units = c("in"), dpi = 300)



```


```{r}
# color Pct_assigned
plot_line <- pe_all_table_1.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  select(Library_ID, PC1, PC2, no_species, Pct_assigned, Continent, Study, Lab) %>%
  ggplot(., aes(x = PC1, y = no_species)) + # Species
         geom_point(aes(fill = Pct_assigned, shape = Lab), size = 1, stroke = 0.2, shape = 21) +
         geom_smooth(method = glm, se = F, col = "black", linewidth = 0.7) +
         scale_fill_viridis_c(option = "C") +
         theme_classic(base_size = 10) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         ggpubr::stat_regline_equation(label.y = 140, aes(label = ..rr.label..)) +
         ylab("Number of species") +
         ggtitle("")
plot_line



plot_line <- pe_all_table_1.pca$variates$X %>%
  as.data.frame() %>%
  rownames_to_column("Library_ID") %>%
  inner_join(metadata, by = "Library_ID") %>%
  select(Library_ID, PC1, PC2, no_species, Pct_assigned, Continent, Study, Lab) %>%
  ggplot(., aes(x = PC1, y = no_species)) + # Species
         geom_point(aes(fill = Pct_assigned, shape = Lab), size = 1.2, stroke = 0.2, shape = 21) + # Pct_assigned
         geom_smooth(method = glm, se = F, col = "black", linewidth = 0.7) +
         scale_fill_viridis_c(option = "C") +
         # scale_fill_manual(values = c("gray50","gray50","gray50","gray50","gray50","gray50","gray50")) +
         theme_classic(base_size = 10) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
         ggpubr::stat_regline_equation(label.y = 140, aes(label = ..rr.label..)) +
         ylab("Number of species") +
         ggtitle("")
plot_line

pcaplot_test <- ((sp_all_continent_plot / plot_line) + plot_layout(heights = c(1.20,1)))
pcaplot_test


# ggsave("~/Desktop/pacific_color.png", plot = pcaplot_test, device = "png",
#        scale = 1, width = 10, height = 12, units = c("in"), dpi = 300)

```

```{r adonis_pe_all}

pe_all_table_1 <- pe_all_table %>%
  select(-ERR5729638, -ERR5729643) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  mutate(Counts = Counts + 1) %>%
  spread(Species,Counts) %>%
  column_to_rownames("Library_ID")

# check the number of components to retain by tuning the PCA
tune.pe_all_table_1_pca <- tune.pca(pe_all_table_1, logratio = 'CLR')

# perform a PCA to see how the data cluster
pe_all_table_1.pca <- pca(pe_all_table_1, ncomp = 3, logratio = 'CLR')


# Euclidean distance, uses the clr-transformed matrix from the mixOmics PCA above
pe_all_table_1.euc <- vegdist(pe_all_table_1.pca$X, method = "euclidean", na.rm = F)
pe_all_table_1.euc.pcoa <- ape::pcoa(pe_all_table_1.euc)

pe_all_table_1.euc.pcoavectors <- as.data.frame(pe_all_table_1.euc.pcoa$vectors)
pe_all_table_1.euc.pcoavectors <- pe_all_table_1.euc.pcoavectors %>%
  rownames_to_column("Library_ID")

pe_all_table_1.euc.vecmet <- left_join(pe_all_table_1.euc.pcoavectors, metadata, by = "Library_ID")

pe_all_table_1.euc.vecmet %>%
  ggplot(., aes(Axis.1, Axis.2)) +
    geom_point(aes(fill = Study), size = 3, stroke = 0.3, shape = 21) +
    scale_fill_manual(values = Study_colors) +
    scale_color_manual(values = Study_colors) +
    # scale_shape_manual(values = Lab_shapes) +
    stat_ellipse(aes(colour = Study, group = Study)) +
    theme_minimal(base_size = 7) +
    ggtitle("Study")

metadata_all <- metadata %>%
    right_join(., pe_all_table_1.euc.pcoavectors %>% select(Library_ID), by = "Library_ID")
  

# test for difference between different metadata categories 
adonis2(pe_all_table_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin")

all_permanova <- adonis2(pe_all_table_1.euc ~ Lab, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")
all_permanova  <- all_permanova %>% bind_rows(adonis2(pe_all_table_1.euc ~ No_reads_no_human, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_table_1.euc ~ Age_mean, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_table_1.euc ~ gc_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata")) %>%
  bind_rows(adonis2(pe_all_table_1.euc ~ seq_len_avg, metadata_all, permutations = 999, by = "margin") %>% rownames_to_column("Metadata"))


all_permanova %>%
  as_tibble() %>%
  drop_na(`F`) %>%
  arrange(desc(`F`)) %>%
  bind_cols(., as_tibble(c("***","**","***","***","**"))) %>%
  arrange(desc(R2))
  # fwrite(., "./05-results.backup/all_permanova.tsv", sep = "\t", quote =  F)


# all together
adonis2(pe_all_table_1.euc ~ Continent + Lab + Pct_assigned + no_species + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin")

# controlled for Lab
adonis2(pe_all_table_1.euc ~ Continent + Pct_assigned + no_species + Age_mean + gc_avg + seq_len_avg, metadata_all, permutations = 999, by = "margin", strata = metadata_all$Lab)


```

```{r beta_dispersion}
# Test whether there is more dispersion between any of the studies

# Study
groups_study <- pe_all_table %>%
  slice_head(n = 1) %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "Counts") %>%
  select(-Species, -Counts) %>%
  left_join(., 
            metadata_all %>%
            select(Library_ID, Study, Continent)) %>%
  # these are also already removed in the input table below
  drop_na(Continent) %>%
  pull(Continent)

## Calculate multivariate dispersions
study_betadisper <- betadisper(pe_all_table_1.euc, groups_study, bias.adjust =TRUE)
study_betadisper

## Perform test
anova(study_betadisper)

## Permutation test for F
permutest(study_betadisper, pairwise = TRUE, permutations = 99)

## Tukey's Honest Significant Differences
study_betadisper.HSD <- TukeyHSD(study_betadisper)
plot(study_betadisper.HSD)

## Plot the groups and distances to centroids on the
## first two PCoA axes
## with data ellipses instead of hulls
plot(study_betadisper, ellipse = TRUE, hull = FALSE) # 1 sd data ellipse

## can also specify which axes to plot, ordering respected
# plot(study_betadisper, axes = c(3,1), seg.col = "forestgreen", seg.lty = "dashed")

## Draw a boxplot of the distances to centroid for each group
boxplot(study_betadisper)

study_dist <- as.data.frame(study_betadisper$distances) %>%
  rownames_to_column("Library_ID") %>%
  rename(dist_to_centroid = 2)

```

```{r betadisper_raincloud}

raincloud <- study_dist %>%
  left_join(., metadata_all %>%
              select(Library_ID, Continent)) %>%
  mutate(Continent = fct_relevel(Continent, "Pacific","Asia","Africa","Europe","Caribbean", "North America")) %>%
  arrange(Continent) %>%
ggplot(., aes(x = Continent, y = dist_to_centroid, fill = Continent)) + 
  ## add half-violin from {ggdist} package
  ggdist::stat_halfeye(
    ## custom bandwidth
    adjust = .5, 
    ## adjust height
    width = .6, 
    ## move geom to the right
    justification = -.2, 
    ## remove slab interval
    .width = 0, 
    point_colour = NA
  ) + 
  geom_boxplot(
    width = .12, 
    ## remove outliers
    outlier.color = NA, ## `outlier.shape = NA` works as well
    color = "grey30"
  ) +
  ## add dot plots from {ggdist} package
  ggdist::stat_dots(
    ## orientation to the left
    side = "left", 
    ## move geom to the left
    justification = 1.1, 
    ## adjust grouping (binning) of observations 
    binwidth = 0.5
  ) + 
  ## remove white space on the left
  coord_cartesian(xlim = c(1.2, NA)) +
  theme_minimal(base_size = 10) +
  scale_fill_manual(values = Continent_colors) +
  theme(axis.text = element_text(size = 9)) +
  theme(legend.position = "none") +
  xlab("Study") +
  ylab("Distance to centroid") +
  scale_y_continuous(breaks=scales::pretty_breaks(n=8)) +
  theme(axis.title.x = element_text(vjust=-1.5)) +
  theme(axis.title.y = element_text(vjust=1.9))

raincloud


```

## FAVA 

```{r}
# format a table for FAVA
pe_all_table_ra <- pe_all_table %>%
  pivot_longer(!Species, names_to = "Library_ID", values_to = "counts") %>%
  pivot_wider(names_from = "Species", values_from = "counts") %>%
  inner_join(full_metadata %>%
               select(Library_ID, Continent, Study) %>%
               mutate(Study = ifelse(str_detect(Library_ID, "ELR|IVE"),"Middenbeemster",Study)), by = "Library_ID") %>%
  adorn_percentages(denominator = "row") %>%
  select(Continent, Study, Library_ID, everything())
  
```

```{r fava_world}

# FAVA for the whole group
f_world_all <- fava(pe_all_table_ra, K = 549, normalized = TRUE) %>%
  as_data_frame() %>%
  rename(FAVA = 1) %>%
  mutate(Continent = "All") %>%
  select(Continent, FAVA)

f_world_all

# try out FAVA with Continent as the variable
f_world_continent <- fava(pe_all_table_ra, group = "Continent", K = 549, normalized = TRUE)
f_world_continent


# try out FAVA with Study as the variable
fava(pe_all_table_ra, group = "Study", K = 549, normalized = TRUE)

```

```{r}

f_world_continent_plot <- f_world_continent %>%
  bind_rows(., f_world_all) %>%
  ggplot(., aes(x = Continent, y = FAVA, color = Continent)) +
         geom_line() +
         geom_point(size = 3) +
         scale_color_manual(values = Continent_colors) +
         theme_minimal(base_size = 10) +
         theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 12)) +
         theme(axis.text.y = element_text(size = 10)) +
         theme(panel.grid.minor = element_blank()) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=14)) +
        geom_hline(yintercept = 0, linetype = "dotted") +
        ylab("FAVA") +
        xlab("Continent")
f_world_continent_plot


# how many samples per Continent?
world_continent_plot <-pe_all_table_ra %>%
  group_by(Continent) %>%
  count() %>%
  ungroup() %>%
  ggplot(., aes(x = Continent, y = n, color = Continent)) +
         geom_line() +
         geom_point(size = 3) +
         scale_color_manual(values = Continent_colors) +
         theme_minimal(base_size = 12) +
         theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 12)) +
         theme(axis.text.y = element_text(size = 10)) +
         theme(panel.grid.minor = element_blank()) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=14)) +
        geom_hline(yintercept = 0, linetype = "dotted") +
        ylab("No. samples") +
        xlab("Continent")
world_continent_plot



fava_world_plot <- pe_all_table_ra %>%
  group_by(Continent) %>%
  count() %>%
  ungroup() %>%
  full_join(., f_world_continent, by = "Continent") %>%
  mutate(n = ifelse(Continent == "All",72,n)) %>%
  ggplot(., aes(x = n, y = FAVA, color = Continent)) +
         geom_line() +
         geom_point(size = 3) +
         scale_color_manual(values = Continent_colors) +
         theme_minimal(base_size = 8) +
         theme(axis.text.x = element_text(angle = 0, hjust = 0.5, size = 8)) +
         theme(axis.text.y = element_text(size = 7)) +
         theme(panel.grid.minor = element_blank()) +
         scale_y_continuous(breaks=scales::pretty_breaks(n=10)) +
         scale_x_continuous(breaks=scales::pretty_breaks(n=10)) +
        # add a line where all samples fall, not separated by continent
        geom_hline(yintercept = 0.0692, linetype = "dotted") +
        ylab("FAVA") +
        xlab("No. samples")
fava_world_plot

# ggsave("./06-publication/main_figures/Figure_3/fava_worldc_plot.pdf", plot = fava_world_plot, device = "pdf",
#        scale = 1, width = 8, height = 4.5, units = c("in"), dpi = 300)



# fava_world_plot <- world_continent_plot + f_world_continent_plot +
#   plot_layout(guides = "collect", widths = c(1,1))
# fava_world_plot


```

```{r}

pcaplot_test <- (((sp_all_continent_plot / plot_line) + plot_layout(heights = c(4,1))) | raincloud) +
  plot_layout(widths = c(1,1.2))
pcaplot_test

pcaplot_test <-  (((sp_all_continent_plot / plot_line) + plot_layout(heights = c(4,1))) | (raincloud / fava_world_plot + plot_layout(heights = c(1.2,1)))) +
  plot_layout(widths = c(1,1))
pcaplot_test


# ggsave("./06-publication/main_figures/Figure_3/world_all.svg", plot = pcaplot_test, device = "svg",
#        scale = 1, width = 10, height = 4.5, units = c("in"), dpi = 300)
# 
# ggsave("./06-publication/main_figures/Figure_3/world_all.pdf", plot = pcaplot_test, device = "pdf",
#        scale = 1, width = 10, height = 4.5, units = c("in"), dpi = 300)

```







